{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import FeatureLocation\n",
    "from Bio.Seq import UndefinedSequenceError\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing host annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_inphared</th>\n",
       "      <th>isolation_host_inphared</th>\n",
       "      <th>host_inphared</th>\n",
       "      <th>Colour (Host)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GU339467</td>\n",
       "      <td>Mycobacterium phage RedRock</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>#8C1292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MF417929</td>\n",
       "      <td>Uncultured Caudovirales phage clone 2F_1</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>#D3D3D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MH616963</td>\n",
       "      <td>crAssphage sp. isolate ctbg_1</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>#D3D3D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MH552500</td>\n",
       "      <td>crAssphage sp. isolate ctcc615</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>#D3D3D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BK010471</td>\n",
       "      <td>Carjivirus communis</td>\n",
       "      <td>Unspecified</td>\n",
       "      <td>#D3D3D3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20885</th>\n",
       "      <td>AY526908</td>\n",
       "      <td>Bordetella phage BMP-1</td>\n",
       "      <td>Bordetella</td>\n",
       "      <td>#1479C4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20886</th>\n",
       "      <td>AJ270057</td>\n",
       "      <td>Chlamydia phage 2</td>\n",
       "      <td>Chlamydia</td>\n",
       "      <td>#9A4AD6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20887</th>\n",
       "      <td>AB016282</td>\n",
       "      <td>Bacillus phage phi105</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>#74AF9B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20888</th>\n",
       "      <td>AF109874</td>\n",
       "      <td>Lactococcus phage Tuc2009</td>\n",
       "      <td>Lactococcus</td>\n",
       "      <td>#395C1F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20889</th>\n",
       "      <td>X58839</td>\n",
       "      <td>Acholeplasma phage MV-L51</td>\n",
       "      <td>Acholeplasma</td>\n",
       "      <td>#16CA19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20890 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      id_inphared                   isolation_host_inphared  host_inphared  \\\n",
       "0        GU339467               Mycobacterium phage RedRock  Mycobacterium   \n",
       "1        MF417929  Uncultured Caudovirales phage clone 2F_1    Unspecified   \n",
       "2        MH616963             crAssphage sp. isolate ctbg_1    Unspecified   \n",
       "3        MH552500            crAssphage sp. isolate ctcc615    Unspecified   \n",
       "4        BK010471                       Carjivirus communis    Unspecified   \n",
       "...           ...                                       ...            ...   \n",
       "20885    AY526908                    Bordetella phage BMP-1     Bordetella   \n",
       "20886    AJ270057                         Chlamydia phage 2      Chlamydia   \n",
       "20887    AB016282                     Bacillus phage phi105       Bacillus   \n",
       "20888    AF109874                 Lactococcus phage Tuc2009    Lactococcus   \n",
       "20889      X58839                 Acholeplasma phage MV-L51   Acholeplasma   \n",
       "\n",
       "      Colour (Host)  \n",
       "0           #8C1292  \n",
       "1           #D3D3D3  \n",
       "2           #D3D3D3  \n",
       "3           #D3D3D3  \n",
       "4           #D3D3D3  \n",
       "...             ...  \n",
       "20885       #1479C4  \n",
       "20886       #9A4AD6  \n",
       "20887       #74AF9B  \n",
       "20888       #395C1F  \n",
       "20889       #16CA19  \n",
       "\n",
       "[20890 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hosts = pd.read_csv(\"../data/raw/inphared_8Sep2023/8Sep2023_vConTACT2_host_annotations.tsv\", sep = \"\\t\")\n",
    "\n",
    "hosts = hosts.rename(columns={\n",
    "    \"Node_ID\": \"id_inphared\",\n",
    "    \"Description\": \"isolation_host_inphared\",\n",
    "    \"Host\": \"host_inphared\"\n",
    "})\n",
    "\n",
    "# Display the DataFrame to check if the changes are applied\n",
    "hosts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def engineer_features(folder):\n",
    "    # Read data from the TSV file\n",
    "    df = pd.read_csv(\"../data/raw/inphared_8Sep2023/8Sep2023_vConTACT2_host_annotations.tsv\", sep = \"\\t\")\n",
    "\n",
    "    df = df.rename(columns={\n",
    "        \"Node_ID\": \"id_inphared\",\n",
    "        \"Description\": \"isolation_host_inphared\",\n",
    "        \"Host\": \"host_inphared\"\n",
    "    })\n",
    "\n",
    "\n",
    "    # Initialize lists for storing data\n",
    "    ids = []\n",
    "    topologies = []\n",
    "    sequences = []\n",
    "\n",
    "    # Read the GenBank file for topology and sequences\n",
    "    genbank_file = f'{folder}/pharokka.gbk'\n",
    "    for record in SeqIO.parse(genbank_file, \"genbank\"):\n",
    "        ids.append(record.id)    \n",
    "        topologies.append(record.annotations.get('topology', 'N/A'))\n",
    "        sequences.append(str(record.seq))\n",
    "\n",
    "\n",
    "\n",
    "    # Create DataFrames from the topologies and sequences\n",
    "    topology_df = pd.DataFrame({'id_inphared': ids, 'topology': topologies})\n",
    "    sequence_df = pd.DataFrame({'id_inphared': ids, 'sequence': sequences})\n",
    "\n",
    "    # Merge the TSV data with the GenBank topology and sequence data\n",
    "    df = pd.merge(df, topology_df, on='id_inphared', how='left')\n",
    "    df = pd.merge(df, sequence_df, on='id_inphared', how='left')\n",
    "    del(topology_df)\n",
    "    del(sequence_df)\n",
    "\n",
    "    df = pd.get_dummies(df, columns=['topology'])\n",
    "\n",
    "\n",
    "    df_length_gc_cds_density = pd.read_csv(f\"{folder}/pharokka_length_gc_cds_density.tsv\", sep=\"\\t\")\n",
    "    df_length_gc_cds_density['contig'] = df_length_gc_cds_density['contig'].str.slice(0,8)\n",
    "    df_length_gc_cds_density\n",
    "\n",
    "\n",
    "    df = pd.merge(df, df_length_gc_cds_density, left_on='id_inphared', right_on='contig', how='outer')\n",
    "    del(df_length_gc_cds_density)\n",
    "\n",
    "    df_cds = pd.read_csv(f\"{folder}/pharokka_cds_functions.tsv\", sep=\"\\t\")\n",
    "    \n",
    "    # Define a dictionary of replacements\n",
    "    replacements = {\n",
    "        \"DNA, RNA and nucleotide metabolism\":\"nucleotide_metabolism\",\n",
    "        \"head and packaging\": \"head_packaging\",\n",
    "        \"moron, auxiliary metabolic gene and host takeover\": \"host_takeover\",\n",
    "        \"transcription regulation\": \"transcription\",\n",
    "        \"unknown function\": \"unkown_function\"\n",
    "    }\n",
    "\n",
    "    # Replace the values using the dictionary\n",
    "    df_cds['Description'] = df_cds['Description'].replace(replacements)\n",
    "\n",
    "    df_cds = df_cds.pivot(index='contig', columns='Description', values='Count').reset_index()\n",
    "    df_cds['contig'] = df_cds['contig'].str.slice(0,8)\n",
    "\n",
    "    df = pd.merge(df, df_cds, left_on='id_inphared', right_on='contig', how='outer')\n",
    "    del(df_cds)\n",
    "\n",
    "\n",
    "    df_frame = pd.read_csv(f\"{folder}/pharokka_cds_final_merged_output.tsv\", sep=\"\\t\",  low_memory=False)\n",
    "\n",
    "    frame_counts = df_frame.groupby('contig')['frame'].value_counts().unstack(fill_value=0)\n",
    "\n",
    "    # Ensure both '+' and '-' columns are present\n",
    "    frame_counts['+'] = frame_counts.get('+', 0)\n",
    "    frame_counts['-'] = frame_counts.get('-', 0)\n",
    "\n",
    "    # Rename columns explicitly\n",
    "    frame_counts = frame_counts.rename(columns={'+': 'frame_positive', '-': 'frame_negative'})\n",
    "\n",
    "    # Reset index if needed\n",
    "    frame_counts = frame_counts.reset_index()\n",
    "    frame_counts = frame_counts.rename_axis(None, axis=1).reset_index()\n",
    "    frame_counts['contig'] = frame_counts['contig'].str.slice(0,8)\n",
    "    df = pd.merge(df, frame_counts, left_on='id_inphared', right_on='contig', how='outer')\n",
    "\n",
    "    df['jumbophage'] = df['length'].apply(lambda x: x >= 200000)\n",
    "    df['jumbophage'] = df['jumbophage'].astype(int)  # Convert True/False to 1/0\n",
    "    \n",
    "\n",
    "    # Creating new manual features\n",
    "    df['positive_strand_%'] = round(df['frame_positive'] / df['CDS'] * 100,2)\n",
    "    df['negative_strand_%'] = round(df['frame_negative'] / df['CDS'] * 100,2)\n",
    "\n",
    "\n",
    "    df = df.rename(columns={'id_inphared': 'id'})\n",
    "    df['dummy_index'] = 0\n",
    "\n",
    "    # Group by the dummy index and aggregate using 'first'\n",
    "    df = df.groupby('dummy_index').first()\n",
    "\n",
    "    # Reset the index\n",
    "    df = df.reset_index(drop=True)\n",
    "\n",
    "    columns = ['id','host_inphared','isolation_host_inphared','length', 'jumbophage', 'gc_perc',\n",
    "        'CDS','frame_positive', 'frame_negative', 'cds_coding_density',  'positive_strand_%',\n",
    "        \"negative_strand_%\",\n",
    "        \n",
    "        'CARD_AMR_Genes',\n",
    "        'CRISPRs', 'VFDB_Virulence_Factors', 'connector',\n",
    "       'head_packaging', 'host_takeover', 'integration and excision', 'lysis',\n",
    "       'nucleotide_metabolism', 'other', 'tRNAs', 'tail', 'tmRNAs',\n",
    "       'transcription', 'unkown_function', 'transl_table', 'sequence']\n",
    "    \n",
    "    df = df[columns]\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def staining_feature(staining_df, features_df):\n",
    "    stain = pd.read_csv(staining_df, index_col = 0)\n",
    "    stain = stain[['Accession', 'staining']]\n",
    "    stain = stain.rename(columns={'Accession': 'id'})\n",
    "\n",
    "    features_df = pd.merge(features_df, stain, on='id', how='left')\n",
    "\n",
    "\n",
    "\n",
    "    return features_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-> Starting Feature Engineering ----------------------\n",
      "Output file already exists. Overwriting it...\n",
      "Processing Pharokka phages in directory: ../data/interim/pharokka/10_folders/\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16443.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16444.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16445.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16446.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16447.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16448.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16449.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16450.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16451.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16452.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16453.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16454.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16455.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16456.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16500.fasta.pharokka...\n",
      "Processing ../data/interim/pharokka/10_folders/sequence_16501.fasta.pharokka...\n",
      "Feature engineering completed and data stored to CSV in ../data/test/new_engineer_features.csv.\n"
     ]
    }
   ],
   "source": [
    "# Code to generate features of the phages in a given folder \n",
    "def process_pharokka_output(args_data, output_csv):\n",
    "    # Check if output CSV file exists, if not create it\n",
    "    if not os.path.exists(output_csv):\n",
    "        with open(output_csv, 'w') as f:\n",
    "            pass\n",
    "\n",
    "    # Iterate over each folder in the main folder\n",
    "    for folder in os.listdir(args_data):\n",
    "        folder_path = os.path.join(args_data, folder)\n",
    "\n",
    "        # Check if the path is a directory\n",
    "        if os.path.isdir(folder_path):\n",
    "            # Check if \"pharokka.gbk\" file exists in the folder\n",
    "            if \"pharokka.gbk\" in os.listdir(folder_path):\n",
    "                print(f\"Processing {folder_path}...\")\n",
    "                # Pass the full path of \"pharokka.gbk\" to your function and get the dataframe\n",
    "                df = engineer_features(folder_path)\n",
    "\n",
    "                # Append the dataframe to the CSV file\n",
    "                with open(output_csv, 'a') as f:\n",
    "                    df.to_csv(f, header=f.tell()==0, index=False)\n",
    "            else:\n",
    "                print(f\"Skipped {folder_path}: 'pharokka.gbk' file not found or not a directory\")\n",
    "\n",
    "args_data = '../data/interim/pharokka/10_folders/'  # Replace with your main folder path  # Replace with your main folder path\n",
    "output_csv = '../data/test/new_engineer_features.csv'\n",
    "\n",
    "print(\"-> Starting Feature Engineering ----------------------\")\n",
    "# Delete the output CSV file if it already exists\n",
    "if os.path.exists(output_csv):\n",
    "    print(\"Output file already exists. Overwriting it...\")\n",
    "    os.remove(output_csv)\n",
    "    \n",
    "if os.path.isdir(args_data):\n",
    "    print(f\"Processing Pharokka phages in directory: {args_data}\")\n",
    "    process_pharokka_output(args_data, output_csv)\n",
    "\n",
    "# Print completion message\n",
    "print(f\"Feature engineering completed and data stored to CSV in {output_csv}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id', 'host_inphared', 'isolation_host_inphared', 'length',\n",
       "       'jumbophage', 'gc_perc', 'CDS', 'frame_positive', 'frame_negative',\n",
       "       'cds_coding_density', 'positive_strand_%', 'negative_strand_%',\n",
       "       'CARD_AMR_Genes', 'CRISPRs', 'VFDB_Virulence_Factors', 'connector',\n",
       "       'head_packaging', 'host_takeover', 'integration and excision', 'lysis',\n",
       "       'nucleotide_metabolism', 'other', 'tRNAs', 'tail', 'tmRNAs',\n",
       "       'transcription', 'unkown_function', 'transl_table', 'sequence'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../data/test/new_engineer_features.csv')\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "phage",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
