{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/Alvaro/Library/Mobile Documents/com~apple~CloudDocs/DTU/Autumn 2023/Phage project/phage/notebooks'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import FeatureLocation\n",
    "from Bio.Seq import UndefinedSequenceError\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../data/interim/pharokka/tmp_pharokka/MZ079855__Klebsiella_phage_vB_Kpn_3.fna_pharokka/'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "code = \"MZ079855__Klebsiella_phage_vB_Kpn_3\" # different genome lengths annotated\n",
    "entry = f\"../data/interim/pharokka/tmp_pharokka/{code}.fna_pharokka/\"\n",
    "entry"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genbank file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def engineer_features(genbank_file):\n",
    "    # Lists to hold data\n",
    "    ids = []\n",
    "    genome_lengths = []\n",
    "    gc_contents = []\n",
    "    sequences = []\n",
    "    reverse_complements = []\n",
    "    cds_numbers = []\n",
    "    positive_strands = []\n",
    "    negative_strands = []\n",
    "    coding_capacities = []\n",
    "    molecule_types = []\n",
    "    topologies = []\n",
    "    trna_counts = []\n",
    "\n",
    "    \n",
    "    # Read the GenBank file\n",
    "    for record in SeqIO.parse(genbank_file, \"genbank\"):\n",
    "        try:\n",
    "            # Attempt to access the sequence, which may raise UndefinedSequenceError\n",
    "            sequence = str(record.seq)\n",
    "            # print(record.id)\n",
    "        except UndefinedSequenceError:\n",
    "            # print(f\"Skipping record {record.id} as sequence is undefined.\")\n",
    "            continue  # Skip this record\n",
    "\n",
    "        # Calculate genome length and GC content\n",
    "        total_length = len(sequence)\n",
    "        gc_content = round((sequence.count('C') + sequence.count('G')) / total_length * 100, 3)\n",
    "\n",
    "        # Initialize counters\n",
    "        plus = 0\n",
    "        minus = 0\n",
    "        coding_count = 0\n",
    "        trna_count = 0\n",
    "        seen = set()  # Store seen barcodes\n",
    "\n",
    "        for feature in record.features:\n",
    "            start = feature.location.start\n",
    "            end = feature.location.end\n",
    "            length = len(FeatureLocation(start, end))\n",
    "            barcode = f\"{start}_{end}_{length}\"\n",
    "\n",
    "            if feature.type != 'source' and barcode not in seen:\n",
    "                coding_count += length\n",
    "                seen.add(barcode)\n",
    "\n",
    "            if feature.type == 'CDS':\n",
    "                if feature.location.strand == 1:\n",
    "                    plus += 1\n",
    "                elif feature.location.strand == -1:\n",
    "                    minus += 1\n",
    "            elif feature.type == 'tRNA':\n",
    "                trna_count += 1\n",
    "\n",
    "        \n",
    "        # Calculate total number of CDS\n",
    "        total_CDS = plus + minus\n",
    "\n",
    "        # Calculate strand usage as a percentage\n",
    "        per_plus = round((plus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "        per_minus = round((minus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "\n",
    "        # Calculate coding capacity as a percentage\n",
    "        coding_capacity = (coding_count / total_length) * 100\n",
    "\n",
    "        # Extract molecule_type and topology\n",
    "        molecule_type = record.annotations.get('molecule_type', 'N/A')\n",
    "        topology = record.annotations.get('topology', 'N/A')\n",
    "\n",
    "        # Append data to lists\n",
    "        ids.append(record.id)\n",
    "        genome_lengths.append(total_length)\n",
    "        gc_contents.append(gc_content)\n",
    "        sequences.append(sequence)\n",
    "        reverse_complements.append(str(sequence[::-1]))\n",
    "        cds_numbers.append(total_CDS)\n",
    "        positive_strands.append(per_plus)\n",
    "        negative_strands.append(per_minus)\n",
    "        coding_capacities.append(coding_capacity)\n",
    "        molecule_types.append(molecule_type)\n",
    "        topologies.append(topology)\n",
    "        trna_counts.append(trna_count)\n",
    "    \n",
    "    print(\"Processing the entries...\")\n",
    "    # Convert lists to pandas DataFrame\n",
    "    df = pd.DataFrame({\n",
    "        'id': ids,\n",
    "        'genome_length': genome_lengths,\n",
    "        'gc_%': gc_contents,\n",
    "        'sequence': sequences,\n",
    "        'reverse_complement': reverse_complements,\n",
    "        'cds_number': cds_numbers,\n",
    "        'positive_strand_%': positive_strands,\n",
    "        'negative_strand_%': negative_strands,\n",
    "        'coding_capacity': coding_capacities,\n",
    "        'molecule_type': molecule_types,\n",
    "        'topology': topologies,\n",
    "        'trna_count': trna_counts\n",
    "    })\n",
    "\n",
    "\n",
    "    df['id'] = df['id'].str[:-2]\n",
    "\n",
    "    # Check for unexpected molecule types\n",
    "    expected_molecule_types = ['ss-DNA', 'DNA', 'RNA', 'ss-RNA']\n",
    "\n",
    "    # Check and correct 'cRNA' entries\n",
    "    cRNA_entries = df[df['molecule_type'] == 'cRNA']\n",
    "    if not cRNA_entries.empty:\n",
    "        for entry_id in cRNA_entries['id']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cRNA'. Changing it to 'RNA'.\")\n",
    "        df.loc[df['molecule_type'] == 'cRNA', 'molecule_type'] = 'RNA'\n",
    "\n",
    "    # Check and correct 'cDNA' entries\n",
    "    cDNA_entries = df[df['molecule_type'] == 'cDNA']\n",
    "    if not cDNA_entries.empty:\n",
    "        for entry_id in cDNA_entries['id']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cDNA'. Changing it to 'DNA'.\")\n",
    "        df.loc[df['molecule_type'] == 'cDNA', 'molecule_type'] = 'DNA'\n",
    "\n",
    "    unexpected_types = set(df['molecule_type']) - set(expected_molecule_types)\n",
    "\n",
    "    if unexpected_types:\n",
    "        for utype in unexpected_types:\n",
    "            # Get ids of entries with the unexpected molecule type\n",
    "            ids_to_exclude = df[df['molecule_type'] == utype]['id'].tolist()\n",
    "            for entry_id in ids_to_exclude:\n",
    "                print(f\"Warning: Entry with id '{entry_id}' has unrecognized molecule type '{utype}'. It will not be considered.\")\n",
    "            df = df[df['molecule_type'] != utype]\n",
    "            \n",
    "    df = pd.get_dummies(df, columns=['molecule_type'])\n",
    "\n",
    "    expected_columns = ['jumbophage', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']\n",
    "    for col in expected_columns:\n",
    "        if col not in df.columns:\n",
    "            df[col] = 0  # Filling with zeros\n",
    "        df[col] = df[col].astype(bool)  # Convert to boolean\n",
    "\n",
    "    df['jumbophage'] = df['genome_length'].apply(lambda x: x >= 200000)\n",
    "    df['jumbophage'] = df['jumbophage'].astype(int)  # Convert True/False to 1/0\n",
    "    df = pd.get_dummies(df, columns=['topology'])\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing the entries...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>gc_%</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reverse_complement</th>\n",
       "      <th>cds_number</th>\n",
       "      <th>positive_strand_%</th>\n",
       "      <th>negative_strand_%</th>\n",
       "      <th>coding_capacity</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>molecule_type_DNA</th>\n",
       "      <th>jumbophage</th>\n",
       "      <th>molecule_type_ss-DNA</th>\n",
       "      <th>molecule_type_RNA</th>\n",
       "      <th>molecule_type_ss-RNA</th>\n",
       "      <th>topology_linear</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn</td>\n",
       "      <td>112003</td>\n",
       "      <td>41.361</td>\n",
       "      <td>TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...</td>\n",
       "      <td>TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...</td>\n",
       "      <td>186</td>\n",
       "      <td>71.51</td>\n",
       "      <td>28.49</td>\n",
       "      <td>84.663804</td>\n",
       "      <td>17</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  id  genome_length    gc_%  \\\n",
       "0  MZ079855__Klebsiella_phage_vB_Kpn         112003  41.361   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...   \n",
       "\n",
       "                                  reverse_complement  cds_number  \\\n",
       "0  TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...         186   \n",
       "\n",
       "   positive_strand_%  negative_strand_%  coding_capacity  trna_count  \\\n",
       "0              71.51              28.49        84.663804          17   \n",
       "\n",
       "   molecule_type_DNA  jumbophage  molecule_type_ss-DNA  molecule_type_RNA  \\\n",
       "0               True           0                 False              False   \n",
       "\n",
       "   molecule_type_ss-RNA  topology_linear  \n",
       "0                 False             True  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = engineer_features(f\"{entry}pharokka.gbk\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CDS final merged output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['gene', 'start', 'stop', 'frame', 'contig', 'score', 'mmseqs_phrog',\n",
      "       'mmseqs_alnScore', 'mmseqs_seqIdentity', 'mmseqs_eVal',\n",
      "       'mmseqs_top_hit', 'pyhmmer_phrog', 'pyhmmer_bitscore', 'pyhmmer_evalue',\n",
      "       'custom_hmm_id', 'custom_hmm_bitscore', 'custom_hmm_evalue', 'phrog',\n",
      "       'Method', 'Region', 'color', 'annot', 'category', 'vfdb_hit',\n",
      "       'vfdb_alnScore', 'vfdb_seqIdentity', 'vfdb_eVal', 'vfdb_short_name',\n",
      "       'vfdb_description', 'vfdb_species', 'CARD_hit', 'CARD_alnScore',\n",
      "       'CARD_seqIdentity', 'CARD_eVal', 'CARD_species', 'ARO_Accession',\n",
      "       'CARD_short_name', 'Protein_Accession', 'DNA_Accession',\n",
      "       'AMR_Gene_Family', 'Drug_Class', 'Resistance_Mechanism',\n",
      "       'transl_table'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>frame</th>\n",
       "      <th>contig</th>\n",
       "      <th>score</th>\n",
       "      <th>mmseqs_phrog</th>\n",
       "      <th>mmseqs_alnScore</th>\n",
       "      <th>mmseqs_seqIdentity</th>\n",
       "      <th>mmseqs_eVal</th>\n",
       "      <th>...</th>\n",
       "      <th>CARD_eVal</th>\n",
       "      <th>CARD_species</th>\n",
       "      <th>ARO_Accession</th>\n",
       "      <th>CARD_short_name</th>\n",
       "      <th>Protein_Accession</th>\n",
       "      <th>DNA_Accession</th>\n",
       "      <th>AMR_Gene_Family</th>\n",
       "      <th>Drug_Class</th>\n",
       "      <th>Resistance_Mechanism</th>\n",
       "      <th>transl_table</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DHKIYHKJ_CDS_0001</td>\n",
       "      <td>175</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "      <td>4.2</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DHKIYHKJ_CDS_0002</td>\n",
       "      <td>730</td>\n",
       "      <td>930</td>\n",
       "      <td>+</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "      <td>4.5</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DHKIYHKJ_CDS_0003</td>\n",
       "      <td>1074</td>\n",
       "      <td>904</td>\n",
       "      <td>-</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "      <td>20.8</td>\n",
       "      <td>26718</td>\n",
       "      <td>148.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.179e-46</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DHKIYHKJ_CDS_0004</td>\n",
       "      <td>1187</td>\n",
       "      <td>1074</td>\n",
       "      <td>-</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "      <td>3.8</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>No_PHROG</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DHKIYHKJ_CDS_0005</td>\n",
       "      <td>1332</td>\n",
       "      <td>1670</td>\n",
       "      <td>+</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "      <td>45.5</td>\n",
       "      <td>24947</td>\n",
       "      <td>285.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.171e-92</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene  start  stop frame                               contig  \\\n",
       "0  DHKIYHKJ_CDS_0001    175     2     -  MZ079855__Klebsiella_phage_vB_Kpn_3   \n",
       "1  DHKIYHKJ_CDS_0002    730   930     +  MZ079855__Klebsiella_phage_vB_Kpn_3   \n",
       "2  DHKIYHKJ_CDS_0003   1074   904     -  MZ079855__Klebsiella_phage_vB_Kpn_3   \n",
       "3  DHKIYHKJ_CDS_0004   1187  1074     -  MZ079855__Klebsiella_phage_vB_Kpn_3   \n",
       "4  DHKIYHKJ_CDS_0005   1332  1670     +  MZ079855__Klebsiella_phage_vB_Kpn_3   \n",
       "\n",
       "   score mmseqs_phrog mmseqs_alnScore mmseqs_seqIdentity mmseqs_eVal  ...  \\\n",
       "0    4.2     No_PHROG        No_PHROG           No_PHROG    No_PHROG  ...   \n",
       "1    4.5     No_PHROG        No_PHROG           No_PHROG    No_PHROG  ...   \n",
       "2   20.8        26718           148.0                1.0   6.179e-46  ...   \n",
       "3    3.8     No_PHROG        No_PHROG           No_PHROG    No_PHROG  ...   \n",
       "4   45.5        24947           285.0                1.0   3.171e-92  ...   \n",
       "\n",
       "  CARD_eVal CARD_species ARO_Accession CARD_short_name Protein_Accession  \\\n",
       "0       NaN          NaN           NaN             NaN               NaN   \n",
       "1       NaN          NaN           NaN             NaN               NaN   \n",
       "2       NaN          NaN           NaN             NaN               NaN   \n",
       "3       NaN          NaN           NaN             NaN               NaN   \n",
       "4       NaN          NaN           NaN             NaN               NaN   \n",
       "\n",
       "  DNA_Accession AMR_Gene_Family Drug_Class Resistance_Mechanism transl_table  \n",
       "0           NaN             NaN        NaN                  NaN           11  \n",
       "1           NaN             NaN        NaN                  NaN           11  \n",
       "2           NaN             NaN        NaN                  NaN           11  \n",
       "3           NaN             NaN        NaN                  NaN           11  \n",
       "4           NaN             NaN        NaN                  NaN           11  \n",
       "\n",
       "[5 rows x 43 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(f\"{entry}pharokka_cds_final_merged_output.tsv\", sep=\"\\t\")\n",
    "df['gene'].value_counts().sum()\n",
    "print(df.columns)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>frame</th>\n",
       "      <th>Region</th>\n",
       "      <th>annot</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DHKIYHKJ_CDS_0001</td>\n",
       "      <td>175</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DHKIYHKJ_CDS_0002</td>\n",
       "      <td>730</td>\n",
       "      <td>930</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DHKIYHKJ_CDS_0003</td>\n",
       "      <td>1074</td>\n",
       "      <td>904</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DHKIYHKJ_CDS_0004</td>\n",
       "      <td>1187</td>\n",
       "      <td>1074</td>\n",
       "      <td>-</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DHKIYHKJ_CDS_0005</td>\n",
       "      <td>1332</td>\n",
       "      <td>1670</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>DHKIYHKJ_CDS_0182</td>\n",
       "      <td>108400</td>\n",
       "      <td>108798</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>DHKIYHKJ_CDS_0183</td>\n",
       "      <td>108854</td>\n",
       "      <td>109255</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>DHKIYHKJ_CDS_0184</td>\n",
       "      <td>109268</td>\n",
       "      <td>109777</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>DHKIYHKJ_CDS_0185</td>\n",
       "      <td>109790</td>\n",
       "      <td>110110</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>DHKIYHKJ_CDS_0186</td>\n",
       "      <td>110167</td>\n",
       "      <td>111030</td>\n",
       "      <td>+</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>unknown function</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>186 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene   start    stop frame Region                 annot  \\\n",
       "0    DHKIYHKJ_CDS_0001     175       2     -    CDS  hypothetical protein   \n",
       "1    DHKIYHKJ_CDS_0002     730     930     +    CDS  hypothetical protein   \n",
       "2    DHKIYHKJ_CDS_0003    1074     904     -    CDS  hypothetical protein   \n",
       "3    DHKIYHKJ_CDS_0004    1187    1074     -    CDS  hypothetical protein   \n",
       "4    DHKIYHKJ_CDS_0005    1332    1670     +    CDS  hypothetical protein   \n",
       "..                 ...     ...     ...   ...    ...                   ...   \n",
       "181  DHKIYHKJ_CDS_0182  108400  108798     +    CDS  hypothetical protein   \n",
       "182  DHKIYHKJ_CDS_0183  108854  109255     +    CDS  hypothetical protein   \n",
       "183  DHKIYHKJ_CDS_0184  109268  109777     +    CDS  hypothetical protein   \n",
       "184  DHKIYHKJ_CDS_0185  109790  110110     +    CDS  hypothetical protein   \n",
       "185  DHKIYHKJ_CDS_0186  110167  111030     +    CDS  hypothetical protein   \n",
       "\n",
       "             category  \n",
       "0    unknown function  \n",
       "1    unknown function  \n",
       "2    unknown function  \n",
       "3    unknown function  \n",
       "4    unknown function  \n",
       "..                ...  \n",
       "181  unknown function  \n",
       "182  unknown function  \n",
       "183  unknown function  \n",
       "184  unknown function  \n",
       "185  unknown function  \n",
       "\n",
       "[186 rows x 7 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['gene', 'start', 'stop', 'frame', 'Region', 'annot', 'category']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Region\n",
       "CDS    186\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"Region\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "annot\n",
       "hypothetical protein                                          97\n",
       "tail length tape measure protein                               4\n",
       "anaerobic ribonucleoside reductase large subunit               3\n",
       "DNA polymerase                                                 3\n",
       "replication origin binding                                     3\n",
       "NAD-dependent DNA ligase                                       3\n",
       "central tail fiber J                                           3\n",
       "NrdD-like anaerobic ribonucleotide reductase large subunit     2\n",
       "ribonucleoside diphosphate reductase small subunit             2\n",
       "endolysin                                                      2\n",
       "Rz-like spanin                                                 2\n",
       "DnaB-like replicative helicase                                 2\n",
       "DNA primase                                                    2\n",
       "NinI-like serine-threonine phosphatase                         2\n",
       "DNA helicase                                                   2\n",
       "DNA binding protein                                            2\n",
       "RusA-like Holliday junction resolvase                          2\n",
       "tail protein                                                   2\n",
       "straight fibre tail protein                                    2\n",
       "SbcC-like subunit of palindrome specific endonuclease          2\n",
       "terminase large subunit                                        2\n",
       "tail completion or Neck1 protein                               2\n",
       "membrane protein                                               2\n",
       "base plate tail tube protein                                   2\n",
       "DNA transfer protein                                           2\n",
       "receptor binding tail protein                                  2\n",
       "terminase small subunit                                        1\n",
       "PhoH-like phosphate starvation-inducible                       1\n",
       "ribonucleoside-diphosphate reductase large subunit             1\n",
       "ribonucleotide reductase large subunit                         1\n",
       "dihydrofolate reductase                                        1\n",
       "distal tail protein                                            1\n",
       "thymidylate synthase                                           1\n",
       "Sir2 (NAD-dependent deacetylase)                               1\n",
       "polynucleotide kinase                                          1\n",
       "deoxynucleoside monophosphate kinase                           1\n",
       "ATP-dependent protease                                         1\n",
       "holin                                                          1\n",
       "thioredoxin domain                                             1\n",
       "Rnase H                                                        1\n",
       "portal protein                                                 1\n",
       "tail assembly chaperone                                        1\n",
       "tail terminator                                                1\n",
       "Tail assembly chaperone                                        1\n",
       "collar tail protein for L-shaped tail fibre attachment         1\n",
       "tail fiber protein                                             1\n",
       "dUTPase                                                        1\n",
       "exonuclease                                                    1\n",
       "major tail protein                                             1\n",
       "SbcD-like subunit of palindrome specific endonuclease          1\n",
       "helicase                                                       1\n",
       "single strand DNA binding protein                              1\n",
       "head-tail adaptor                                              1\n",
       "major head protein                                             1\n",
       "head maturation protease                                       1\n",
       "transcriptional regulator, D5-like                             1\n",
       "structural protein with Ig domain                              1\n",
       "unknown function                                               1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"annot\"].value_counts().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "category\n",
       "unknown function                                     98\n",
       "DNA, RNA and nucleotide metabolism                   38\n",
       "tail                                                 21\n",
       "other                                                 8\n",
       "head and packaging                                    7\n",
       "lysis                                                 5\n",
       "moron, auxiliary metabolic gene and host takeover     4\n",
       "connector                                             4\n",
       "transcription regulation                              1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"category\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "frame\n",
       "+    133\n",
       "-     53\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"frame\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CDS functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>Count</th>\n",
       "      <th>contig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDS</td>\n",
       "      <td>186</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>connector</td>\n",
       "      <td>4</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNA, RNA and nucleotide metabolism</td>\n",
       "      <td>38</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>head and packaging</td>\n",
       "      <td>7</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>integration and excision</td>\n",
       "      <td>0</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>lysis</td>\n",
       "      <td>5</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>moron, auxiliary metabolic gene and host takeover</td>\n",
       "      <td>4</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>other</td>\n",
       "      <td>8</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tail</td>\n",
       "      <td>21</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>transcription regulation</td>\n",
       "      <td>1</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>unknown function</td>\n",
       "      <td>98</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tRNAs</td>\n",
       "      <td>23</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CRISPRs</td>\n",
       "      <td>0</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tmRNAs</td>\n",
       "      <td>0</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>VFDB_Virulence_Factors</td>\n",
       "      <td>0</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CARD_AMR_Genes</td>\n",
       "      <td>0</td>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Description  Count  \\\n",
       "0                                                 CDS    186   \n",
       "1                                           connector      4   \n",
       "2                  DNA, RNA and nucleotide metabolism     38   \n",
       "3                                  head and packaging      7   \n",
       "4                            integration and excision      0   \n",
       "5                                               lysis      5   \n",
       "6   moron, auxiliary metabolic gene and host takeover      4   \n",
       "7                                               other      8   \n",
       "8                                                tail     21   \n",
       "9                            transcription regulation      1   \n",
       "10                                   unknown function     98   \n",
       "11                                              tRNAs     23   \n",
       "12                                            CRISPRs      0   \n",
       "13                                             tmRNAs      0   \n",
       "14                             VFDB_Virulence_Factors      0   \n",
       "15                                     CARD_AMR_Genes      0   \n",
       "\n",
       "                                 contig  \n",
       "0   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "1   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "2   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "3   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "4   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "5   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "6   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "7   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "8   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "9   MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "10  MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "11  MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "12  MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "13  MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "14  MZ079855__Klebsiella_phage_vB_Kpn_3  \n",
       "15  MZ079855__Klebsiella_phage_vB_Kpn_3  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(f\"{entry}pharokka_cds_functions.tsv\", sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Length gc CDS density"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>length</th>\n",
       "      <th>gc_perc</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>cds_coding_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ223858__Enterobacter_phage_EV136_1</td>\n",
       "      <td>39701</td>\n",
       "      <td>52.07</td>\n",
       "      <td>11</td>\n",
       "      <td>91.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 contig  length  gc_perc  transl_table  \\\n",
       "0  MZ223858__Enterobacter_phage_EV136_1   39701    52.07            11   \n",
       "\n",
       "   cds_coding_density  \n",
       "0               91.05  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"../data/interim/pharokka/tmp_pharokka/MZ223858__Enterobacter_phage_EV136_1.fna_pharokka/pharokka_length_gc_cds_density.tsv\", sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top hits mash inphared"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>Accession</th>\n",
       "      <th>mash_distance</th>\n",
       "      <th>mash_pval</th>\n",
       "      <th>mash_matching_hashes</th>\n",
       "      <th>Description</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Genome_Length_(bp)</th>\n",
       "      <th>Jumbophage</th>\n",
       "      <th>molGC_(%)</th>\n",
       "      <th>...</th>\n",
       "      <th>Sub-family</th>\n",
       "      <th>Family</th>\n",
       "      <th>Order</th>\n",
       "      <th>Class</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>Realm</th>\n",
       "      <th>Baltimore_Group</th>\n",
       "      <th>Genbank_Division</th>\n",
       "      <th>Isolation_Host_(beware_inconsistent_and_nonsense_values)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn_3</td>\n",
       "      <td>MZ079855</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1000/1000</td>\n",
       "      <td>Klebsiella phage vB_Kpn_3</td>\n",
       "      <td>Klebsiella phage vB_Kpn_3 Sugarlandvirus Demer...</td>\n",
       "      <td>112003</td>\n",
       "      <td>False</td>\n",
       "      <td>41.361</td>\n",
       "      <td>...</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Demerecviridae</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Caudoviricetes</td>\n",
       "      <td>Uroviricota</td>\n",
       "      <td>Heunggongvirae</td>\n",
       "      <td>Duplodnaviria</td>\n",
       "      <td>Group I</td>\n",
       "      <td>PHG</td>\n",
       "      <td>MDR-Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                contig Accession  mash_distance  mash_pval  \\\n",
       "0  MZ079855__Klebsiella_phage_vB_Kpn_3  MZ079855            0.0        0.0   \n",
       "\n",
       "  mash_matching_hashes                Description  \\\n",
       "0            1000/1000  Klebsiella phage vB_Kpn_3   \n",
       "\n",
       "                                      Classification  Genome_Length_(bp)  \\\n",
       "0  Klebsiella phage vB_Kpn_3 Sugarlandvirus Demer...              112003   \n",
       "\n",
       "   Jumbophage  molGC_(%)  ...    Sub-family          Family         Order  \\\n",
       "0       False     41.361  ...  Unclassified  Demerecviridae  Unclassified   \n",
       "\n",
       "            Class       Phylum         Kingdom          Realm  \\\n",
       "0  Caudoviricetes  Uroviricota  Heunggongvirae  Duplodnaviria   \n",
       "\n",
       "   Baltimore_Group Genbank_Division  \\\n",
       "0          Group I              PHG   \n",
       "\n",
       "  Isolation_Host_(beware_inconsistent_and_nonsense_values)  \n",
       "0                          MDR-Klebsiella pneumoniae        \n",
       "\n",
       "[1 rows x 31 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(f\"{entry}pharokka_top_hits_mash_inphared.tsv\", sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Column Accession:  ['MZ079855']\n",
      "Column Genome_Length_(bp):  [112003]\n",
      "Column Jumbophage:  [False]\n",
      "Column molGC_(%):  [41.361]\n",
      "Column Molecule:  ['DNA']\n",
      "Column Modification_Date:  ['16-JUL-2021']\n",
      "Column Number_CDS:  [182]\n",
      "Column Positive_Strand_(%):  [70.8791208791209]\n",
      "Column Negative_Strand_(%):  [29.1208791208791]\n",
      "Column Coding_Capacity_(%):  [83.2638411471121]\n",
      "Column Low_Coding_Capacity_Warning:  [nan]\n",
      "Column tRNAs:  [19]\n"
     ]
    }
   ],
   "source": [
    "feature_columns = ['Accession',\n",
    "       'Genome_Length_(bp)', 'Jumbophage', 'molGC_(%)', 'Molecule',\n",
    "       'Modification_Date', 'Number_CDS', 'Positive_Strand_(%)',\n",
    "       'Negative_Strand_(%)', 'Coding_Capacity_(%)',\n",
    "       'Low_Coding_Capacity_Warning', 'tRNAs']\n",
    "for column in df.columns:\n",
    "    if column in feature_columns:\n",
    "        print(f\"Column {column}: \",df[column].to_list())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top hits card"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>gene</th>\n",
       "      <th>card_hit</th>\n",
       "      <th>card_alnScore</th>\n",
       "      <th>card_seqIdentity</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [contig, gene, card_hit, card_alnScore, card_seqIdentity, start, stop, frame]\n",
       "Index: []"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(f\"{entry}top_hits_card.tsv\", sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Top hits vfdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>gene</th>\n",
       "      <th>vfdb_hit</th>\n",
       "      <th>vfdb_alnScore</th>\n",
       "      <th>vfdb_seqIdentity</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>frame</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [contig, gene, vfdb_hit, vfdb_alnScore, vfdb_seqIdentity, start, stop, frame]\n",
       "Index: []"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(f\"{entry}top_hits_vfdb.tsv\", sep=\"\\t\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genbank file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>jumbophage</th>\n",
       "      <th>topology_linear</th>\n",
       "      <th>topology_circular</th>\n",
       "      <th>gc_%</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>cds_number</th>\n",
       "      <th>coding_capacity</th>\n",
       "      <th>positive_strand_%</th>\n",
       "      <th>negative_strand_%</th>\n",
       "      <th>molecule_type_ss-DNA</th>\n",
       "      <th>molecule_type_DNA</th>\n",
       "      <th>molecule_type_RNA</th>\n",
       "      <th>molecule_type_ss-RNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ079855__Klebsiella_phage_vB_Kpn</td>\n",
       "      <td>112003</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>41.361</td>\n",
       "      <td>17</td>\n",
       "      <td>186</td>\n",
       "      <td>84.663804</td>\n",
       "      <td>71.51</td>\n",
       "      <td>28.49</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  id  genome_length  jumbophage  \\\n",
       "0  MZ079855__Klebsiella_phage_vB_Kpn         112003           0   \n",
       "\n",
       "   topology_linear  topology_circular    gc_%  trna_count  cds_number  \\\n",
       "0             True              False  41.361          17         186   \n",
       "\n",
       "   coding_capacity  positive_strand_%  negative_strand_%  \\\n",
       "0        84.663804              71.51              28.49   \n",
       "\n",
       "   molecule_type_ss-DNA  molecule_type_DNA  molecule_type_RNA  \\\n",
       "0                 False               True              False   \n",
       "\n",
       "   molecule_type_ss-RNA  \n",
       "0                 False  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_file_path = f'{entry}pharokka.gbk'\n",
    "\n",
    "# Lists to hold data\n",
    "ids = []\n",
    "genome_lengths = []\n",
    "gc_contents = []\n",
    "sequences = []\n",
    "reverse_complements = []\n",
    "cds_numbers = []\n",
    "positive_strands = []\n",
    "negative_strands = []\n",
    "coding_capacities = []\n",
    "molecule_types = []\n",
    "topologies = []\n",
    "trna_counts = []\n",
    "\n",
    "# Read the GenBank file\n",
    "for record in SeqIO.parse(genbank_file_path, \"genbank\"):\n",
    "    try:\n",
    "        # Attempt to access the sequence, which may raise UndefinedSequenceError\n",
    "        sequence = str(record.seq)\n",
    "        # print(record.id)\n",
    "    except UndefinedSequenceError:\n",
    "        # print(f\"Skipping record {record.id} as sequence is undefined.\")\n",
    "        continue  # Skip this record\n",
    "\n",
    "    # Calculate genome length and GC content\n",
    "    total_length = len(sequence)\n",
    "    gc_content = round((sequence.count('C') + sequence.count('G')) / total_length * 100, 3)\n",
    "\n",
    "    # Initialize counters\n",
    "    plus = 0\n",
    "    minus = 0\n",
    "    coding_count = 0\n",
    "    trna_count = 0\n",
    "    seen = set()  # Store seen barcodes\n",
    "\n",
    "    for feature in record.features:\n",
    "        start = feature.location.start\n",
    "        end = feature.location.end\n",
    "        length = len(FeatureLocation(start, end))\n",
    "        barcode = f\"{start}_{end}_{length}\"\n",
    "\n",
    "        if feature.type != 'source' and barcode not in seen:\n",
    "            coding_count += length\n",
    "            seen.add(barcode)\n",
    "\n",
    "        if feature.type == 'CDS':\n",
    "            if feature.location.strand == 1:\n",
    "                plus += 1\n",
    "            elif feature.location.strand == -1:\n",
    "                minus += 1\n",
    "        elif feature.type == 'tRNA':\n",
    "            trna_count += 1\n",
    "\n",
    "\n",
    "    # Calculate total number of CDS\n",
    "    total_CDS = plus + minus\n",
    "\n",
    "    # Calculate strand usage as a percentage\n",
    "    per_plus = round((plus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "    per_minus = round((minus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "\n",
    "    # Calculate coding capacity as a percentage\n",
    "    coding_capacity = (coding_count / total_length) * 100\n",
    "\n",
    "    # Extract molecule_type and topology\n",
    "    molecule_type = record.annotations.get('molecule_type', 'N/A')\n",
    "    topology = record.annotations.get('topology', 'N/A')\n",
    "\n",
    "    # Append data to lists\n",
    "    ids.append(record.id)\n",
    "    genome_lengths.append(total_length)\n",
    "    gc_contents.append(gc_content)\n",
    "    sequences.append(sequence)\n",
    "    reverse_complements.append(str(sequence[::-1]))\n",
    "    cds_numbers.append(total_CDS)\n",
    "    positive_strands.append(per_plus)\n",
    "    negative_strands.append(per_minus)\n",
    "    coding_capacities.append(coding_capacity)\n",
    "    molecule_types.append(molecule_type)\n",
    "    topologies.append(topology)\n",
    "    trna_counts.append(trna_count)\n",
    "\n",
    "# Convert lists to pandas DataFrame\n",
    "df = pd.DataFrame({\n",
    "    'id': ids,\n",
    "    'genome_length': genome_lengths,\n",
    "    'gc_%': gc_contents,\n",
    "    'sequence': sequences,\n",
    "    'reverse_complement': reverse_complements,\n",
    "    'cds_number': cds_numbers,\n",
    "    'positive_strand_%': positive_strands,\n",
    "    'negative_strand_%': negative_strands,\n",
    "    'coding_capacity': coding_capacities,\n",
    "    'molecule_type': molecule_types,\n",
    "    'topology': topologies,\n",
    "    'trna_count': trna_counts\n",
    "})\n",
    "\n",
    "\n",
    "# Filter df based on previously known sequences with a staining class\n",
    "accession_values = pd.read_csv('../data/processed/model_data.csv')\n",
    "accession_values = accession_values['Accession'].tolist()\n",
    "df['id'] = df['id'].str[:-2]\n",
    "# df = df[df['id'].isin(accession_values)]\n",
    "\n",
    "\n",
    "# Check for unexpected molecule types\n",
    "expected_molecule_types = ['ss-DNA', 'DNA', 'RNA', 'ss-RNA']\n",
    "\n",
    "# Check and correct 'cRNA' entries\n",
    "cRNA_entries = df[df['molecule_type'] == 'cRNA']\n",
    "if not cRNA_entries.empty:\n",
    "    for entry_id in cRNA_entries['id']:\n",
    "        print(f\"Info: Entry with id '{entry_id}' has molecule type 'cRNA'. Changing it to 'RNA'.\")\n",
    "    df.loc[df['molecule_type'] == 'cRNA', 'molecule_type'] = 'RNA'\n",
    "\n",
    "# Check and correct 'cDNA' entries\n",
    "cDNA_entries = df[df['molecule_type'] == 'cDNA']\n",
    "if not cDNA_entries.empty:\n",
    "    for entry_id in cDNA_entries['id']:\n",
    "        print(f\"Info: Entry with id '{entry_id}' has molecule type 'cDNA'. Changing it to 'DNA'.\")\n",
    "    df.loc[df['molecule_type'] == 'cDNA', 'molecule_type'] = 'DNA'\n",
    "\n",
    "unexpected_types = set(df['molecule_type']) - set(expected_molecule_types)\n",
    "\n",
    "if unexpected_types:\n",
    "    for utype in unexpected_types:\n",
    "        # Get ids of entries with the unexpected molecule type\n",
    "        ids_to_exclude = df[df['molecule_type'] == utype]['id'].tolist()\n",
    "        for entry_id in ids_to_exclude:\n",
    "            print(f\"Warning: Entry with id '{entry_id}' has unrecognized molecule type '{utype}'. It will not be considered.\")\n",
    "        df = df[df['molecule_type'] != utype]\n",
    "\n",
    "        \n",
    "df = pd.get_dummies(df, columns=['molecule_type'])\n",
    "\n",
    "df = pd.get_dummies(df, columns=['topology'])\n",
    "\n",
    "expected_columns = ['topology_linear', 'topology_circular', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']\n",
    "for col in expected_columns:\n",
    "    if col not in df.columns:\n",
    "        df[col] = 0  # Filling with zeros\n",
    "    df[col] = df[col].astype(bool)  # Convert to boolean\n",
    "\n",
    "\n",
    "\n",
    "df['jumbophage'] = df['genome_length'].apply(lambda x: x >= 200000)\n",
    "df['jumbophage'] = df['jumbophage'].astype(int)  # Convert True/False to 1/0\n",
    "\n",
    "# Change order of entries\n",
    "df = df[['id', 'genome_length', 'jumbophage', 'topology_linear', 'topology_circular', 'gc_%', 'trna_count','cds_number', 'coding_capacity','positive_strand_%','negative_strand_%', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']]\n",
    "# df = df[['id', 'genome_length', 'jumbophage', 'molecule_type', 'topology', 'gc_%', 'trna_count','cds_number', 'coding_capacity','positive_strand_%','negative_strand_%' ,'sequence', 'reverse_complement']]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>molGC_(%)</th>\n",
       "      <th>Molecule</th>\n",
       "      <th>Number_CDS</th>\n",
       "      <th>Positive_Strand_(%)</th>\n",
       "      <th>Negative_Strand_(%)</th>\n",
       "      <th>Coding_Capacity_(%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>41.361</td>\n",
       "      <td>DNA</td>\n",
       "      <td>182</td>\n",
       "      <td>70.879121</td>\n",
       "      <td>29.120879</td>\n",
       "      <td>83.263841</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   molGC_(%) Molecule  Number_CDS  Positive_Strand_(%)  Negative_Strand_(%)  \\\n",
       "0     41.361      DNA         182            70.879121            29.120879   \n",
       "\n",
       "   Coding_Capacity_(%)  \n",
       "0            83.263841  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_inphared = pd.read_csv(f\"{entry}pharokka_top_hits_mash_inphared.tsv\", sep=\"\\t\")\n",
    "df_inphared[['molGC_(%)', 'Molecule', 'Number_CDS', 'Positive_Strand_(%)',\n",
    "       'Negative_Strand_(%)', 'Coding_Capacity_(%)']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Information gathered by INPHARED\n",
      "Column molGC_(%):  [41.361]\n",
      "Column Molecule:  ['DNA']\n",
      "Column Number_CDS:  [182]\n",
      "Column Positive_Strand_(%):  [70.8791208791209]\n",
      "Column Negative_Strand_(%):  [29.1208791208791]\n",
      "Column Coding_Capacity_(%):  [83.2638411471121]\n"
     ]
    }
   ],
   "source": [
    "feature_columns = ['molGC_(%)', 'Molecule', 'Number_CDS', 'Positive_Strand_(%)',\n",
    "       'Negative_Strand_(%)', 'Coding_Capacity_(%)']\n",
    "\n",
    "print(\"Number of CDS annotated by PHAROKKA\")\n",
    "\n",
    "\n",
    "print(\"Information gathered by INPHARED\")\n",
    "for column in df_inphared.columns:\n",
    "    if column in feature_columns:\n",
    "        print(f\"Column {column}: \",df_inphared[column].to_list())\n",
    "\n",
    "\n",
    "print"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "phage",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
