{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import FeatureLocation\n",
    "from Bio.Seq import UndefinedSequenceError\n",
    "import subprocess\n",
    "\n",
    "def unix_call(command):\n",
    "    job = subprocess.run(command.split()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dealing with GenBank record EF380009.1\n"
     ]
    }
   ],
   "source": [
    "gbk_filename =  \"/Users/Alvaro/Desktop/phage/sequence.gb\"\n",
    "faa_filename = \"NC_005213_converted.fna\"\n",
    "input_handle  = open(gbk_filename, \"r\")\n",
    "output_handle = open(faa_filename, \"w\")\n",
    "\n",
    "for seq_record in SeqIO.parse(input_handle, \"genbank\") :\n",
    "    print (\"Dealing with GenBank record %s\" % seq_record.id)\n",
    "    output_handle.write(\">%s %s\\n%s\\n\" % (\n",
    "           seq_record.id,\n",
    "           seq_record.description,\n",
    "           seq_record.seq))\n",
    "\n",
    "output_handle.close()\n",
    "input_handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calling one sequence as a test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_directory = \"../data/interim/pharokka/\"\n",
    "database_dir = \"../data/external/pharokka_v1.4.0_databases\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: EF380009.1\n",
      "Name: EF380009.1\n",
      "Description: EF380009.1 Enterobacteria phage phiX174 isolate AP100, complete genome\n",
      "Number of features: 8\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=30-OCT-2023\n",
      "/accessions=['EF380009']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=\n",
      "/organism=.\n",
      "/taxonomy=[]\n",
      "Seq('GAGTTTTATCGCTTCCATGACGCAGAAGTTAACACTTTCGGATATTTCTGATGA...GCA')\n",
      "ID: MN335248.1\n",
      "Name: MN335248.1\n",
      "Description: MN335248.1 prueba\n",
      "Number of features: 16\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=30-OCT-2023\n",
      "/accessions=['MN335248']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=\n",
      "/organism=.\n",
      "/taxonomy=[]\n",
      "Seq('AGTACCGCCCGAATTTCGCAGCAACCCAACCGACGCAAGCCCAACCAACAGAGA...CAC')\n",
      "Processing record: MN335248.1\n",
      "\n",
      "AGTACCGCCCGAATTTCGCAGCAACCCAACCGACGCAAGCCCAACCAACAGAGAGAACGAAGACCATGAGCAACGCACCGAAGATCACCATCAACAGCGCCGTCGAAACCCGCACTGTCACCACCTCGAAGGGCATGCCGAAGGCCATTTACAGCCAGCGCGCCACGCTCGAAACCGAAGCGATGCGTATCCAGATCGAAGTCGAATGCGATGGCCCGGACAAGGGCTATCCGGTCGGCACGGTCAAGGAATGGGATTTGGTCACCGATCTGGTGCCGGGCCGTTTCGGTGTCGAACTGGCGCGCCGCATGACGCTGGTCGATCCGCAGAGCGGCAAGCCGCAGCAGCGGCAGGCAGCGTAATCGATGGCCGTGCTCATCCCCGCCTGCCGGGAAGCCGATCTAGACGCGGCAGCGGGGACGTGCACGGCGGTCATCTGGATTCCTCAACCGTCACTCTTGCCGGAACTGGCGGTGAAGGATGCCCAGGCCATGGGCAGCGCTATCGCGTTCCTGTGGGCCACGGCGTACGTGTTCCGGCTCATCCGCAAGAAAATTCAACAGTCCTAGGAGGACATGCAATGCGCAAGCTCAAGACCCTGTTCAAGAACAAGACCGCTGCTCTGGCCGCAGTTGGTTCGGCCGCTCTCGTCTCCGCTCCGGCCTTCGCCACTGGCGGTGGTGGTGTGGATGTGGGTGACGTGGTGTCGGCCATCCAGGGCGCAGCAGGCCCGATTGCAGCCATCGGCGGTGCAGTGCTGACTGTCATGGTGGGCATCAAGGTCTACAAGTGGGTGCGCCGCGCCATGTAACGACCACCGGCGGACAGGGCCAACTCCCTCCCGCCGGTCTTTTATGGGGATAGGGCAGGGGAACGGGCGATGGAAGGGTGGATTTGGTTAGTCGCATGGCTGGTGGCCTGCGCGATTGTCTTCGTGGATTTCGAATAATGGGCTGGCTCGCACGCGTGTTTGCATCCGCGATTGCGCGTCGCCTCGCATACGTGCTTGTAGCGGCAACGCTCGCATGGTGTGGCATGGGCAAGGCGCACGCAGCGAATTGTGATGCAACCTCTGACATGTGCAGCGAGGGCGATGCCTGGAATGCGGCACGTCAGCTTGCAGATTCCCGTGGTCCTGACACCTGCAAACTGGTTGGCGGCAACAACGCCTCCTACACAGGCCCAAATATTGAGAAAGACCCGGGCAATTCTTCGAAGGACGTGGGTGCGCTCGCTGTGCGAGTTTCCTGTTCGAATGGGTCGCTGCCGTTCTTAGGCACCAAGTACTACGTGACCGCAAAATCGTGCGAACAAGAGCCTGGCTACACCGGCGGCGGTCCGTGGGGGACATATGTTGGCACCGCGCGCAACGGCAGTATCGGTTGTCGCAATGGCTGCGACGGTGTGTGGTTTGGCAATGGCGATGGCTCGATGACATGGGGCAGTACGGGCGCCGTGTGTCCGACTGATCCGGAGAAGACGTGCGATGCCATGGGCAAGGGGTACGGATGGAACGGCTACCTTGGCGTGTGTGAGCCGCCGCCGACTGAAGAGTGCCCAGAGGGTCAGGTGCCGGACGGGAAGGGTGGCTGTTCAACCAACAAATGCCCCGAAGGCATGCTATTGCAGGCTGACGGCACCTGCACACCGAAGAAGAACGATTGCCCTGCTGGGATGATCAAGTCGCCAACTGGTTCATGCCTTCCAGGCGACGGCCAGTGCGCCGCTGGTGAAGTGCGTGGTCCTGATGGAACGTGCAAGAAAGATGCAGATGGTGATGGTGAACCTGACGGGGACGGCGAGGGCGGCGAGGGCGAGGGCGATAGCAAGAGCGAATTTTCTGGCGGCGACAGCTGTGACTCTCCGCCTAGTTGCAGCGGTGACGCAATTATGTGCGGACAGGCGCGCATTCAATGGCGCATTGATTGCAACACCCGCCGCGATGTAAACATCACAGGTGGTTCGTGCGCGGCCATGCCTGTATGCGTTGGCAAGAACTGCAAGGCATTGGAGTATTCCCAGCTGTTGATGCAATGGCGTGCTGCATGTGCCTTGGAGAAGGCAGCGAACAACAGTGGCGGCGGTACTGGAAACAATGCAGACGTCAAGGCTATTCGCGATGCGATCACCGGCAATGGTTCTGCCGATATCGGTGCCGATGGCAAGCCTGCTGACGCGTTCTCTGATGAATCCGGATACGGCGAAGATGGCTACCCCACTGGCGAACTCGATACACAGGGCTTCGGCTATAGCCGCACGTGCCCAACGATTCCCGATGTAGCCGTGTTCGGCCAGACGCTGCATTTCGACACGTCCAAGTTCTGTCAGTGGATGGTGCTGGGCGGGCAGATTGTCTTGGTCATGGCGTCCTTGGTGTCCTTGCGTTTGATGAGTCAAGGAGGTAGCGCGTAATGCCCTGGTTAATCGCACAACTCGTCACCGCGCTTGCATGGCTGTTCAAGTCGCGTATCGGCTTATGGATCATGACCGCGCTTGTGTGGCTCGGTATCAATTTCGGCACGATCAAGATGGTGGTCGAGCCGGCAATCGATCTGCTCAAGGATTACGCCCAAGGCATGGGCAATGGCAACGGTCAGCTTGGCGCAGACGCCATGGCGTGGTTCGGCGTGTTGCAGTTCGATAAGGCGTTGACCATGGTCATCTCTGCCATCGCTGCCAAGCACGCCATCATGCAAGGTCGCCTGTTCCTGTTTAAGCGTGGATTCGGAGCGAAGCCGTAATGCCAATCGAGCTATACACCGGGCAACCCGGCAACGGTAAAACCGCTCTCATGATGGAGCGCCTGGTCGAAGAGTCGAAGCGCGCCGAGCGACCGATTTTCGCTGCGGGTATCGACGGCCTACAGGACGGCTTGGCGACTGTGCTCGATGACCCGCGCAAGTGGAACGATAAGGATGCAGACGGCAACTACGTTGTGCCCAATGGCTCGTTGATCTTCGTTGACGAGGCGTGGAAGTGGTATGGCCACCTGCACGACGCTACGCGCCAGCAGACGCCGAAACATGTGCTCGATCTTGCTGAGCATCGGCATCGCGGTCTGGATTTTGTGTGGACCACGCAGCAGCCGAACCAGCTGTATCCGTTCGTGCGTGGATTGATCGGTGCACATACGCATGTGGTGCGTCGGTTCAGCACGAAGATGATCGATGTGTTCCGCTGGGGCGAGTTGAACGAAGAAATCAAGTCGTCAGCCAAGCGCGATCTTGCCCAGCGCACCACGCGCCTTCTGCCGTCCTCGATCTTCGGCGCATACAAGTCGGCCGAGGTCCACACGATCAAGCCGCGTATTCCGTGGAAAGTGCTGGCGTTGCCGGGATTGGTCATCCTTGCCATCGCGCTTGGATGGCTCGCCTACACGATGCTCAAGCCTAGCGCGATGGCCGCAAAACTCGGAGATAAGGGGACGCAATCGGCGTTAGCCGATGCGGCCCCTGGCGGGTCGGCGAACTCCGCACGGCGTGATGGTCCGCGTTGGGAATCTCCCACCGAATATGCCAAGCAACACCTGCCCCGGTTCGGCACCATGCCGTGGACTGCACCGGTGTTCGATGTTGGGAATCTCCCACCGAATATGCCAAGCAACACCTGCCCCGGTTCGGCACCATGCCGTGGACTGCACCGGTGTTCGATGACCGCAGCATCACCGCCGATCCGATGCTGATCTGCATGTCGTCGCTCGCGGGCACGGACGCGCAGGGCAAGCACAAGGAAGCGTCTTGCACGTGCATGACAGAGCAGGGCACCGCGTACGACCTCGACCAGCCGCAGTGCCGCACCATTGCCAAGCGCGGTCCGGTGTACAACCCGTATCGCGAACGGCGTGAGAACGAGCAGCAGCCCGCGCAGCAGCAACAGGCGGTGCAGGGTGGGGCGGCTGCGCCTGGTCTGTCCGGCGTGGTCGTGCAGCGTGGCATGCGCACACAGGGCACCTTCCCGGAATCGAAGGGCTACAACACCAAGACCACCACGCCATCAACGTCGCTGGAGATGTGACATGACCAGCAGCGGCCGCGAGGCATTGAAGTGGATTGCGCTGGTGTTGATGACCGGCGATCACGTGGTAAAGGTGTTTTTCGGCGGCTATGTGCCGGTGCTATCCGAATTGGGGCGGATCGCGTTCCCGGTGTTCGCGTTGGTCATGGCGTACAACCTTGCCCAGCCACGGGCCGATTATGCGAAATCGGTGCTGCGCCTTGCGGGCTGGGGGCTGCTGGCGCAGCCATTTCACGCCTGGGCGTTCGGGTACTGGCTGCCGCTCAACGTCCTGCTGACTTTCGCGTTGGCGGCGCTGCTCGTCTGGACGGTCCACGCCCGGCATTGGCTCTATGCTGCTGTGTGCTGTCTCGTCCTGCCGCTCGTGGTCGATTACCAATGGTCCGGCGTCTGGCTCGTTCTGGCGGCGTGGTGCTGGTTCCGCACGGGCCGGTTGGAAGCCTTCGGTGGCGTCCTGGCCACCATGGCTGCGTTGTGCTGGTACAACGGCAATCTGTGGGCGCTGGGATCGCTCCCGGTGCTGGCCCTCGGCTACGTCTGGTGGCCGCTGCCGCGCCTGCGCTGGGCGTTCTACGGCTATTACGTTGGCCACCTCGGGCTGCTCGTGCTCATCGCGTCGTTGCCGGCGTTTCAGCAGCACGTCTCGTAGGTAGATGACGTTGTTCCTTGCAGAAATTCGGCTCCCTGGAACCCTTGTGGCAGTAGGCGTTTCAGGATTTGCCGATCGAGCATTGCGCTCGGTGGCCATCATCAATGACCACTCGCGAGCGATGCAGCAGGTTAGCGACCAGAACCGCATGTCGCACGGCTCGATGGTCTTGCCCTCAGGCGTAAAGAAGTGGTGGCCCTGGAAACCAAAACCGGCCCAAGGGCCGGCAAGGTCTACGCGGTCGTAGGTGTCGATCACGCTGCGATTTCCGTATCAGAAGGACCATGGGACGGTAGCAGCGAGGTAAGCCAGAGCCATAGCCAGCGCGCCCGCGATTGCAGCGCCTGCCGCACACGATTTCGCATAATGTATAGAGAGTTAGCGTTATGTTCGGTCGAAAAGTGACTTTGATCACTGATTTTCGATGCCTCAGCCCGGCCCGGCAGGGACACGCCAAGTAGCAGCAAGACGCCCAGCGCGCTTGCAGCTATGCGTTGAAGGATTCCTTTGTGTTTGATGTTCTCCGGTGAGTGCATGCGCACTTCCGATGCTGTCTTGATGATTGGCAATTTCGCTAGCGCTGCAATTTCAGCTGCTTGTTCGAAGTTCGGTACAGAAGTGCCGTCTAACCACCTGATCACCGCGACGTGCGACACCTTGGCTATGGTGGCGAACTCTCGAATGCTCTTTGCATTCGTTTCTTTCATCGCCAGCTCAACCAACTCTTTTGCGTCCATCGTCATTTCCAGTGGAACGGGAGGTTGACAGCGTATCTGAACTGCTGGTAACGTCCCTACTAGAAGCTACCACTGCCCACCCGCCGGCCCCGCCCGGTGCCGGTTGGGCGGGTTCTACCGGGCACCGGGCAGGGGATACACCTTGAAAGCACTTATCCAGTTTTGCCGTGAATTCGGCCTGCTTGTCTGCATGGCATGCGCTACCGGGTTTGGCATTGGCGGACTCGTTGGCTTTATCGCTGGTCAGGGCGCTGCAAGTAGCGTTGAACACCAGATTCAGCGCGATGAGGCGCTCCAGCAGCTTTGCATCGACGGTAACGAAGGCGCCTGCCGCGTTATGGAGTTCCGCAGATGAGCCGCCCAACCGACACCGAACGCGGCGCACGCATTGCGCTCGACTACGTTGAATCAAAGCTTATTCAACGCGATTTATTTCCAACCCGCCGCGCGCCGTCTTTAAAGTTCTGGCGCGAAATAAAGGCGATTGCGACGGAACACCTTGCCGAATGCAAGGCACTCCGCGAGGCCCGCGCATGAACACGGGCGACCTGAAATTCTGGCAGCTGATCGAGCTGAAAAAGAAAGCCGAAACCGACGACGAACGCGCCGAGCTAGACCGACTGATAGAGGAACGCATCAGGTGGGTAGGCTCGGGGCAGGGCGATGAGCCATGAAGACGCAGCCGTCCAGGCTAGCCCATTACCCGAACAGCCCCTGCTATCAGTGCGGGGGAGCGCAGTTTCAGACGTTGAACGCGTGGGATGCGAAGTTGACGGTCTGCACCGATTGCGGCGTGCTGATCTCGAAGCGCAGGGATATGCAGAGTTCTTACAGCGGATGCAGTGGCAACAGTTCTGGACACTCACATTTCGCATCGAAGAAGCCGGCCGAACCGGTGGTGTCCACCCAGAAAAGGCTGATAAAGCGTTCCGATTCTTCGCCAGCTGTATCAACCGGCAAATATACGGCCGTGAGTGGTCAAAGCGTCCGCATGGCGGCATCCAGTGGGCACGGGGGCAAGAGTTCCACAAAGACGGTCGATTGCATTTCCACGCCGTCGCAGCTGCACCTACCGATGATTTAAACCGGCTTATGAGCCGCTATGAGTGGCATGAGTTTTGGTTCAAGGAATTCGGACGTAATCGCATAGAGGCACCACGCAGCCAGCTCGATATAACCGGCTACGTGTCGAAGTACGTAACGAAGGGGGGAGTGGTGGACGTGTCGAAGAACTTCGGCGCCTGGACGCCACCACCGATCGACTACACGCGCCGCCCAGTGCAGGCCGAGTTCGATCAAACAACGCGTAAGGGGAGCATCGAATCACTGGACCGGGGTGTAGGGGCAGCGCCCCTACGGACTGGTCAACGAAACCGCCGGTGATCGCTCACCGGGACATGCAACACCGCCCCCGGTCTGGGAGGCACCGAAGCCGCAGCCTGTTCGCCTACGCGCACCAGCACGGCAACGCCAGGAATCCACCCCTGAAGACCCGCCTTCGCTCGCAGGCGACCTAAAGCGACCGGTAGGCAGACACCGTGACAGTCAGCCCCGAGCGGCAGGTGGCAAGCCCCGCGCACGGTCATCGCTAAGCCTCACCGCACCCCCCGCATGGGGGGTAAGGGGGGCCTTAGCTTGACCCCAC\n"
     ]
    }
   ],
   "source": [
    "genbank_file_path2 =  \"../data/raw/inphared_8Sep2023/8Sep2023_phages_downloaded_from_genbank.gb\"\n",
    "genbank_file_path2 =  \"../data/interim/genbank_engineering/sequence.gb\"\n",
    "genbank_file_path2 = \"/Users/Alvaro/Desktop/prueba_pharokka2/pharokka.gbk\"\n",
    "# Read the GenBank file\n",
    "for record in SeqIO.parse(genbank_file_path2, \"genbank\"):\n",
    "\n",
    "    \n",
    "    if record.id == \"MN335248.1\":\n",
    "        print(f\"Processing record: {record.id}\\n\")\n",
    "        \n",
    "        sequence = str(record.seq)\n",
    "        print(sequence)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-10-30 10:08:57.331 | INFO     | input_commands:instantiate_dirs:160 - Removing output directory /Users/Alvaro/Desktop/prueba_pharokka as -f or --force was specified.\n",
      "2023-10-30 10:08:57.339 | INFO     | __main__:main:84 - Starting Pharokka v1.4.1\n",
      "2023-10-30 10:08:57.340 | INFO     | __main__:main:85 - Command executed: Namespace(citation=False, coding_table='11', custom_hmm='', database='/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/', dnaapler=False, evalue='1E-05', fast=False, force=True, genbank=False, gene_predictor='phanotate', infile='../data/interim/genbank_engineering/sequence.fna', locustag='Default', meta=False, meta_hmm=False, mmseqs2_only=False, outdir='/Users/Alvaro/Desktop/prueba_pharokka', prefix='Default', split=False, terminase=False, terminase_start='nothing', terminase_strand='nothing', threads='2')\n",
      "2023-10-30 10:08:57.340 | INFO     | __main__:main:86 - Repository homepage is https://github.com/gbouras13/pharokka\n",
      "2023-10-30 10:08:57.340 | INFO     | __main__:main:87 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
      "2023-10-30 10:08:57.340 | INFO     | __main__:main:89 - Checking database installation in /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/.\n",
      "2023-10-30 10:08:57.341 | INFO     | __main__:main:92 - All databases have been successfully checked.\n",
      "2023-10-30 10:08:57.341 | INFO     | __main__:main:108 - Checking dependencies.\n",
      "2023-10-30 10:08:57.861 | INFO     | input_commands:check_dependencies:322 - Phanotate version found is v1.5.1\n",
      "2023-10-30 10:08:57.863 | INFO     | input_commands:check_dependencies:331 - Phanotate version is ok.\n",
      "2023-10-30 10:08:57.906 | INFO     | input_commands:check_dependencies:353 - MMseqs2 version found is v13.45111\n",
      "2023-10-30 10:08:57.909 | INFO     | input_commands:check_dependencies:362 - MMseqs2 version is ok.\n",
      "2023-10-30 10:08:58.245 | INFO     | input_commands:check_dependencies:386 - tRNAscan-SE version found is v2.0.12\n",
      "2023-10-30 10:08:58.247 | INFO     | input_commands:check_dependencies:397 - tRNAscan-SE version is ok.\n",
      "2023-10-30 10:08:58.547 | INFO     | input_commands:check_dependencies:421 - MinCED version found is v0.4.2\n",
      "2023-10-30 10:08:58.550 | INFO     | input_commands:check_dependencies:432 - MinCED version is ok.\n",
      "2023-10-30 10:08:58.564 | INFO     | input_commands:check_dependencies:458 - ARAGORN version found is v1.2.41\n",
      "2023-10-30 10:08:58.566 | INFO     | input_commands:check_dependencies:469 - ARAGORN version is ok.\n",
      "2023-10-30 10:08:58.589 | INFO     | input_commands:check_dependencies:485 - mash version found is v2.3\n",
      "2023-10-30 10:08:58.591 | INFO     | input_commands:check_dependencies:492 - mash version is ok.\n",
      "2023-10-30 10:09:00.943 | INFO     | input_commands:check_dependencies:511 - Dnaapler version found is v0.3.2\n",
      "2023-10-30 10:09:00.944 | INFO     | input_commands:check_dependencies:518 - Dnaapler version is ok.\n",
      "2023-10-30 10:09:00.945 | INFO     | input_commands:check_dependencies:529 - Pyrodigal version is v3.0.1\n",
      "2023-10-30 10:09:00.945 | INFO     | input_commands:check_dependencies:530 - Pyrodigal version is ok.\n",
      "2023-10-30 10:09:00.971 | INFO     | input_commands:validate_fasta:205 - Checking Input FASTA.\n",
      "2023-10-30 10:09:00.972 | INFO     | input_commands:validate_fasta:207 - FASTA checked.\n",
      "2023-10-30 10:09:00.972 | INFO     | input_commands:validate_gene_predictor:214 - Phanotate will be used for gene prediction.\n",
      "2023-10-30 10:09:00.973 | INFO     | __main__:main:268 - Running Phanotate.\n",
      "2023-10-30 10:09:00.976 | INFO     | external_tools:run:50 - Started running phanotate.py -f fasta -o /Users/Alvaro/Desktop/prueba_pharokka/phanotate_out_tmp.fasta ../data/interim/genbank_engineering/sequence.fna ...\n",
      "2023-10-30 10:09:04.628 | INFO     | external_tools:run:52 - Done running phanotate.py -f fasta -o /Users/Alvaro/Desktop/prueba_pharokka/phanotate_out_tmp.fasta ../data/interim/genbank_engineering/sequence.fna\n",
      "2023-10-30 10:09:04.630 | INFO     | external_tools:run:50 - Started running phanotate.py -f tabular -o /Users/Alvaro/Desktop/prueba_pharokka/phanotate_out.txt ../data/interim/genbank_engineering/sequence.fna ...\n",
      "2023-10-30 10:09:08.803 | INFO     | external_tools:run:52 - Done running phanotate.py -f tabular -o /Users/Alvaro/Desktop/prueba_pharokka/phanotate_out.txt ../data/interim/genbank_engineering/sequence.fna\n",
      "2023-10-30 10:09:08.823 | INFO     | __main__:main:291 - Starting tRNA-scanSE.\n",
      "2023-10-30 10:09:08.824 | INFO     | external_tools:run:50 - Started running tRNAscan-SE --thread 2 -G -Q -j /Users/Alvaro/Desktop/prueba_pharokka/trnascan_out.gff ../data/interim/genbank_engineering/sequence.fna ...\n",
      "2023-10-30 10:09:09.169 | INFO     | external_tools:run:52 - Done running tRNAscan-SE --thread 2 -G -Q -j /Users/Alvaro/Desktop/prueba_pharokka/trnascan_out.gff ../data/interim/genbank_engineering/sequence.fna\n",
      "2023-10-30 10:09:09.180 | INFO     | processes:run_minced:694 - Running MinCED.\n",
      "2023-10-30 10:09:09.186 | INFO     | external_tools:run:50 - Started running minced ../data/interim/genbank_engineering/sequence.fna /Users/Alvaro/Desktop/prueba_pharokka/pharokka_minced_spacers.txt /Users/Alvaro/Desktop/prueba_pharokka/pharokka_minced.gff ...\n",
      "2023-10-30 10:09:09.570 | INFO     | external_tools:run:52 - Done running minced ../data/interim/genbank_engineering/sequence.fna /Users/Alvaro/Desktop/prueba_pharokka/pharokka_minced_spacers.txt /Users/Alvaro/Desktop/prueba_pharokka/pharokka_minced.gff\n",
      "2023-10-30 10:09:09.573 | INFO     | processes:run_aragorn:723 - Running Aragorn.\n",
      "2023-10-30 10:09:09.574 | INFO     | external_tools:run:50 - Started running aragorn -l -gcbact -w -m -o /Users/Alvaro/Desktop/prueba_pharokka/pharokka_aragorn.txt ../data/interim/genbank_engineering/sequence.fna ...\n",
      "2023-10-30 10:09:09.596 | INFO     | external_tools:run:52 - Done running aragorn -l -gcbact -w -m -o /Users/Alvaro/Desktop/prueba_pharokka/pharokka_aragorn.txt ../data/interim/genbank_engineering/sequence.fna\n",
      "2023-10-30 10:09:09.598 | INFO     | __main__:main:299 - Starting MMseqs2.\n",
      "2023-10-30 10:09:09.599 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on PHROG Database.\n",
      "2023-10-30 10:09:09.604 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs ...\n",
      "2023-10-30 10:09:10.083 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs\n",
      "2023-10-30 10:09:10.090 | INFO     | external_tools:run:50 - Started running mmseqs search -e 1E-05 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/tmp_dir/ -s 8.5 --threads 2 ...\n",
      "2023-10-30 10:24:08.907 | INFO     | external_tools:run:52 - Done running mmseqs search -e 1E-05 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/tmp_dir/ -s 8.5 --threads 2\n",
      "2023-10-30 10:24:08.911 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs_results.tsv --full-header --threads 2 ...\n",
      "2023-10-30 10:24:08.979 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs_results.tsv --full-header --threads 2\n",
      "2023-10-30 10:24:08.982 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on CARD Database.\n",
      "2023-10-30 10:24:08.984 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs ...\n",
      "2023-10-30 10:24:09.094 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs\n",
      "2023-10-30 10:24:09.099 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/CARD_dir/ -s 8.5 --threads 2 ...\n",
      "2023-10-30 10:26:28.181 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/CARD_dir/ -s 8.5 --threads 2\n",
      "2023-10-30 10:26:28.184 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/CARD_results.tsv --full-header --threads 2 ...\n",
      "2023-10-30 10:26:28.212 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/CARD_results.tsv --full-header --threads 2\n",
      "2023-10-30 10:26:28.216 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on VFDB Database.\n",
      "2023-10-30 10:26:28.218 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs ...\n",
      "2023-10-30 10:26:28.531 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs\n",
      "2023-10-30 10:26:28.536 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/vfdb /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB_dir/ -s 8.5 --threads 2 ...\n",
      "2023-10-30 10:30:22.885 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/vfdb /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB_dir/ -s 8.5 --threads 2\n",
      "2023-10-30 10:30:22.890 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/vfdb /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/vfdb_results.tsv --full-header --threads 2 ...\n",
      "2023-10-30 10:30:22.923 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/vfdb /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/vfdb_results.tsv --full-header --threads 2\n",
      "2023-10-30 10:30:22.930 | INFO     | __main__:main:330 - Running PyHMMER on PHROGs.\n",
      "2023-10-30 10:30:43.325 | INFO     | __main__:main:346 - Post Processing Output.\n",
      "2023-10-30 10:30:43.366 | INFO     | post_processing:create_mmseqs_tophits:1920 - Processing mmseqs2 outputs.\n",
      "2023-10-30 10:30:43.366 | INFO     | post_processing:create_mmseqs_tophits:1921 - Processing PHROGs output.\n",
      "2023-10-30 10:30:43.578 | INFO     | post_processing:process_vfdb_results:2134 - Processing VFDB output.\n",
      "2023-10-30 10:30:43.595 | INFO     | post_processing:process_vfdb_results:2224 - 0 VFDB virulence factors identified.\n",
      "2023-10-30 10:30:43.597 | INFO     | post_processing:process_card_results:2241 - Processing CARD output.\n",
      "2023-10-30 10:30:43.609 | INFO     | post_processing:process_card_results:2367 - 0 CARD AMR genes identified.\n",
      "2023-10-30 10:30:43.993 | INFO     | __main__:main:404 - Converting gff to genbank.\n",
      "2023-10-30 10:30:44.060 | INFO     | __main__:main:416 - Finding the closest match for each contig in INPHARED using mash.\n",
      "2023-10-30 10:30:44.061 | INFO     | external_tools:run:50 - Started running mash sketch ../data/interim/genbank_engineering/sequence.fna -o /Users/Alvaro/Desktop/prueba_pharokka/input_mash_sketch.msh -i ...\n",
      "2023-10-30 10:30:44.097 | INFO     | external_tools:run:52 - Done running mash sketch ../data/interim/genbank_engineering/sequence.fna -o /Users/Alvaro/Desktop/prueba_pharokka/input_mash_sketch.msh -i\n",
      "2023-10-30 10:30:44.100 | INFO     | external_tools:run_to_stdout:59 - Started running mash dist /Users/Alvaro/Desktop/prueba_pharokka/input_mash_sketch.msh /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/1Aug2023_genomes.fa.msh -d 0.2 -i ...\n",
      "2023-10-30 10:30:45.617 | INFO     | external_tools:run_to_stdout:61 - Done running mash dist /Users/Alvaro/Desktop/prueba_pharokka/input_mash_sketch.msh /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/1Aug2023_genomes.fa.msh -d 0.2 -i\n",
      "2023-10-30 10:30:45.974 | INFO     | __main__:main:431 - Pharokka has finished.\n",
      "2023-10-30 10:30:45.974 | INFO     | __main__:main:432 - Elapsed time: 1308.64 seconds\n",
      "2023-10-30 10:30:45.975 | INFO     | __main__:main:434 - If you use Pharokka in your research, please cite:\n",
      "2023-10-30 10:30:45.975 | INFO     | __main__:main:435 - George Bouras, Roshan Nepal, Ghais Houtak, Alkis James Psaltis, Peter-John Wormald, Sarah Vreugde.\n",
      "2023-10-30 10:30:45.975 | INFO     | __main__:main:438 - Pharokka: a fast scalable bacteriophage annotation tool.\n",
      "2023-10-30 10:30:45.976 | INFO     | __main__:main:439 - Bioinformatics, Volume 39, Issue 1, January 2023, btac776.\n",
      "2023-10-30 10:30:45.976 | INFO     | __main__:main:440 - https://doi.org/10.1093/bioinformatics/btac776.\n",
      "2023-10-30 10:30:45.979 | INFO     | __main__:main:441 - You should also cite the full list of tools Pharokka uses, which can be found at https://github.com/gbouras13/pharokka#citation.\n"
     ]
    }
   ],
   "source": [
    "# database_dir = \"/Users/Alvaro/Downloads/pharokka_v1.4.0_databases\"\n",
    "database_dir = \"/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/\"\n",
    "sequence = \"../data/interim/genbank_engineering/double_sequence.fna\"\n",
    "sequence = \"../data/interim/genbank_engineering/sequence.fna\"\n",
    "output_dir = \"/Users/Alvaro/Desktop/prueba_pharokka\"\n",
    "unix_call(f'pharokka.py -i {sequence} -o {output_dir} -f -t 2 -d {database_dir}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-10-30 11:12:31.387 | INFO     | input_commands:instantiate_dirs:160 - Removing output directory /Users/Alvaro/Desktop/full_pharokka as -f or --force was specified.\n",
      "2023-10-30 11:12:31.396 | INFO     | __main__:main:84 - Starting Pharokka v1.4.1\n",
      "2023-10-30 11:12:31.397 | INFO     | __main__:main:85 - Command executed: Namespace(citation=False, coding_table='11', custom_hmm='', database='/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/', dnaapler=False, evalue='1E-05', fast=False, force=True, genbank=False, gene_predictor='phanotate', infile='../data/interim/genbank_engineering/double_sequence.fna', locustag='Default', meta=False, meta_hmm=False, mmseqs2_only=False, outdir='/Users/Alvaro/Desktop/full_pharokka', prefix='Default', split=False, terminase=False, terminase_start='nothing', terminase_strand='nothing', threads='10')\n",
      "2023-10-30 11:12:31.397 | INFO     | __main__:main:86 - Repository homepage is https://github.com/gbouras13/pharokka\n",
      "2023-10-30 11:12:31.398 | INFO     | __main__:main:87 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
      "2023-10-30 11:12:31.399 | INFO     | __main__:main:89 - Checking database installation in /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/.\n",
      "2023-10-30 11:12:31.401 | INFO     | __main__:main:92 - All databases have been successfully checked.\n",
      "2023-10-30 11:12:31.402 | INFO     | __main__:main:108 - Checking dependencies.\n",
      "2023-10-30 11:12:32.424 | INFO     | input_commands:check_dependencies:322 - Phanotate version found is v1.5.1\n",
      "2023-10-30 11:12:32.427 | INFO     | input_commands:check_dependencies:331 - Phanotate version is ok.\n",
      "2023-10-30 11:12:32.505 | INFO     | input_commands:check_dependencies:353 - MMseqs2 version found is v13.45111\n",
      "2023-10-30 11:12:32.508 | INFO     | input_commands:check_dependencies:362 - MMseqs2 version is ok.\n",
      "2023-10-30 11:12:32.957 | INFO     | input_commands:check_dependencies:386 - tRNAscan-SE version found is v2.0.12\n",
      "2023-10-30 11:12:32.959 | INFO     | input_commands:check_dependencies:397 - tRNAscan-SE version is ok.\n",
      "2023-10-30 11:12:33.385 | INFO     | input_commands:check_dependencies:421 - MinCED version found is v0.4.2\n",
      "2023-10-30 11:12:33.387 | INFO     | input_commands:check_dependencies:432 - MinCED version is ok.\n",
      "2023-10-30 11:12:33.403 | INFO     | input_commands:check_dependencies:458 - ARAGORN version found is v1.2.41\n",
      "2023-10-30 11:12:33.405 | INFO     | input_commands:check_dependencies:469 - ARAGORN version is ok.\n",
      "2023-10-30 11:12:33.437 | INFO     | input_commands:check_dependencies:485 - mash version found is v2.3\n",
      "2023-10-30 11:12:33.439 | INFO     | input_commands:check_dependencies:492 - mash version is ok.\n",
      "2023-10-30 11:12:39.011 | INFO     | input_commands:check_dependencies:511 - Dnaapler version found is v0.3.2\n",
      "2023-10-30 11:12:39.013 | INFO     | input_commands:check_dependencies:518 - Dnaapler version is ok.\n",
      "2023-10-30 11:12:39.013 | INFO     | input_commands:check_dependencies:529 - Pyrodigal version is v3.0.1\n",
      "2023-10-30 11:12:39.013 | INFO     | input_commands:check_dependencies:530 - Pyrodigal version is ok.\n",
      "2023-10-30 11:12:39.110 | INFO     | input_commands:validate_fasta:205 - Checking Input FASTA.\n",
      "2023-10-30 11:12:39.111 | INFO     | input_commands:validate_fasta:207 - FASTA checked.\n",
      "2023-10-30 11:12:39.112 | INFO     | input_commands:validate_gene_predictor:214 - Phanotate will be used for gene prediction.\n",
      "2023-10-30 11:12:39.115 | INFO     | input_commands:validate_meta:245 - More than one contig detected in the input file. Re-running pharokka with -m meta mode is recommended unless this is a fragmented isolate genome. Continuing.\n",
      "2023-10-30 11:12:39.116 | INFO     | __main__:main:268 - Running Phanotate.\n",
      "2023-10-30 11:12:39.119 | INFO     | external_tools:run:50 - Started running phanotate.py -f fasta -o /Users/Alvaro/Desktop/full_pharokka/phanotate_out_tmp.fasta ../data/interim/genbank_engineering/double_sequence.fna ...\n",
      "2023-10-30 11:13:00.901 | INFO     | external_tools:run:52 - Done running phanotate.py -f fasta -o /Users/Alvaro/Desktop/full_pharokka/phanotate_out_tmp.fasta ../data/interim/genbank_engineering/double_sequence.fna\n",
      "2023-10-30 11:13:00.904 | INFO     | external_tools:run:50 - Started running phanotate.py -f tabular -o /Users/Alvaro/Desktop/full_pharokka/phanotate_out.txt ../data/interim/genbank_engineering/double_sequence.fna ...\n",
      "2023-10-30 11:13:20.402 | INFO     | external_tools:run:52 - Done running phanotate.py -f tabular -o /Users/Alvaro/Desktop/full_pharokka/phanotate_out.txt ../data/interim/genbank_engineering/double_sequence.fna\n",
      "2023-10-30 11:13:20.440 | INFO     | __main__:main:291 - Starting tRNA-scanSE.\n",
      "2023-10-30 11:13:20.442 | INFO     | external_tools:run:50 - Started running tRNAscan-SE --thread 10 -G -Q -j /Users/Alvaro/Desktop/full_pharokka/trnascan_out.gff ../data/interim/genbank_engineering/double_sequence.fna ...\n",
      "2023-10-30 11:13:21.287 | INFO     | external_tools:run:52 - Done running tRNAscan-SE --thread 10 -G -Q -j /Users/Alvaro/Desktop/full_pharokka/trnascan_out.gff ../data/interim/genbank_engineering/double_sequence.fna\n",
      "2023-10-30 11:13:21.290 | INFO     | processes:run_minced:694 - Running MinCED.\n",
      "2023-10-30 11:13:21.292 | INFO     | external_tools:run:50 - Started running minced ../data/interim/genbank_engineering/double_sequence.fna /Users/Alvaro/Desktop/full_pharokka/pharokka_minced_spacers.txt /Users/Alvaro/Desktop/full_pharokka/pharokka_minced.gff ...\n",
      "2023-10-30 11:13:21.608 | INFO     | external_tools:run:52 - Done running minced ../data/interim/genbank_engineering/double_sequence.fna /Users/Alvaro/Desktop/full_pharokka/pharokka_minced_spacers.txt /Users/Alvaro/Desktop/full_pharokka/pharokka_minced.gff\n",
      "2023-10-30 11:13:21.611 | INFO     | processes:run_aragorn:723 - Running Aragorn.\n",
      "2023-10-30 11:13:21.614 | INFO     | external_tools:run:50 - Started running aragorn -l -gcbact -w -m -o /Users/Alvaro/Desktop/full_pharokka/pharokka_aragorn.txt ../data/interim/genbank_engineering/double_sequence.fna ...\n",
      "2023-10-30 11:13:21.638 | INFO     | external_tools:run:52 - Done running aragorn -l -gcbact -w -m -o /Users/Alvaro/Desktop/full_pharokka/pharokka_aragorn.txt ../data/interim/genbank_engineering/double_sequence.fna\n",
      "2023-10-30 11:13:21.641 | INFO     | __main__:main:299 - Starting MMseqs2.\n",
      "2023-10-30 11:13:21.642 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on PHROG Database.\n",
      "2023-10-30 11:13:21.644 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/full_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs ...\n",
      "2023-10-30 11:13:21.923 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/full_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs\n",
      "2023-10-30 11:13:21.929 | INFO     | external_tools:run:50 - Started running mmseqs search -e 1E-05 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/tmp_dir/ -s 8.5 --threads 10 ...\n",
      "2023-10-30 11:32:59.444 | INFO     | external_tools:run:52 - Done running mmseqs search -e 1E-05 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/tmp_dir/ -s 8.5 --threads 10\n",
      "2023-10-30 11:32:59.450 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/mmseqs_results.tsv --full-header --threads 10 ...\n",
      "2023-10-30 11:33:00.118 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/mmseqs_results.tsv --full-header --threads 10\n",
      "2023-10-30 11:33:00.128 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on CARD Database.\n",
      "2023-10-30 11:33:00.130 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/full_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/full_pharokka/CARD_target_dir/target_seqs ...\n",
      "2023-10-30 11:33:00.365 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/full_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/full_pharokka/CARD_target_dir/target_seqs\n",
      "2023-10-30 11:33:00.371 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/full_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/CARD_dir/ -s 8.5 --threads 10 ...\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/Alvaro/Library/Mobile Documents/com~apple~CloudDocs/DTU/Autumn 2023/Phage project/phage/notebooks/05_feature_engineering.ipynb Cell 8\u001b[0m line \u001b[0;36m5\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m sequence \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m../data/interim/genbank_engineering/double_sequence.fna\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m output_dir \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m/Users/Alvaro/Desktop/full_pharokka\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m unix_call(\u001b[39mf\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mpharokka.py -i \u001b[39;49m\u001b[39m{\u001b[39;49;00msequence\u001b[39m}\u001b[39;49;00m\u001b[39m -o \u001b[39;49m\u001b[39m{\u001b[39;49;00moutput_dir\u001b[39m}\u001b[39;49;00m\u001b[39m -f -t 10 -d \u001b[39;49m\u001b[39m{\u001b[39;49;00mdatabase_dir\u001b[39m}\u001b[39;49;00m\u001b[39m'\u001b[39;49m)\n",
      "\u001b[1;32m/Users/Alvaro/Library/Mobile Documents/com~apple~CloudDocs/DTU/Autumn 2023/Phage project/phage/notebooks/05_feature_engineering.ipynb Cell 8\u001b[0m line \u001b[0;36m9\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39munix_call\u001b[39m(command):\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m     job \u001b[39m=\u001b[39m subprocess\u001b[39m.\u001b[39;49mrun(command\u001b[39m.\u001b[39;49msplit())\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:495\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    493\u001b[0m \u001b[39mwith\u001b[39;00m Popen(\u001b[39m*\u001b[39mpopenargs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs) \u001b[39mas\u001b[39;00m process:\n\u001b[1;32m    494\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 495\u001b[0m         stdout, stderr \u001b[39m=\u001b[39m process\u001b[39m.\u001b[39;49mcommunicate(\u001b[39minput\u001b[39;49m, timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[1;32m    496\u001b[0m     \u001b[39mexcept\u001b[39;00m TimeoutExpired \u001b[39mas\u001b[39;00m exc:\n\u001b[1;32m    497\u001b[0m         process\u001b[39m.\u001b[39mkill()\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:1020\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1018\u001b[0m         stderr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mread()\n\u001b[1;32m   1019\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mclose()\n\u001b[0;32m-> 1020\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mwait()\n\u001b[1;32m   1021\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   1022\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:1083\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1081\u001b[0m     endtime \u001b[39m=\u001b[39m _time() \u001b[39m+\u001b[39m timeout\n\u001b[1;32m   1082\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1083\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_wait(timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[1;32m   1084\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1085\u001b[0m     \u001b[39m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1086\u001b[0m     \u001b[39m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1087\u001b[0m     \u001b[39m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1088\u001b[0m     \u001b[39m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1089\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:1822\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1820\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mreturncode \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m   1821\u001b[0m     \u001b[39mbreak\u001b[39;00m  \u001b[39m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 1822\u001b[0m (pid, sts) \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_try_wait(\u001b[39m0\u001b[39;49m)\n\u001b[1;32m   1823\u001b[0m \u001b[39m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   1824\u001b[0m \u001b[39m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   1825\u001b[0m \u001b[39m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   1826\u001b[0m \u001b[39mif\u001b[39;00m pid \u001b[39m==\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpid:\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:1780\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1778\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   1779\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1780\u001b[0m     (pid, sts) \u001b[39m=\u001b[39m os\u001b[39m.\u001b[39;49mwaitpid(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mpid, wait_flags)\n\u001b[1;32m   1781\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mChildProcessError\u001b[39;00m:\n\u001b[1;32m   1782\u001b[0m     \u001b[39m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   1783\u001b[0m     \u001b[39m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   1784\u001b[0m     \u001b[39m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   1785\u001b[0m     pid \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpid\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# database_dir = \"/Users/Alvaro/Downloads/pharokka_v1.4.0_databases\"\n",
    "database_dir = \"/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/\"\n",
    "sequence = \"../data/interim/genbank_engineering/double_sequence.fna\"\n",
    "output_dir = \"/Users/Alvaro/Desktop/full_pharokka\"\n",
    "unix_call(f'pharokka.py -i {sequence} -o {output_dir} -f -t 10 -d {database_dir}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# After pharokka: processing genbank files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>jumbophage</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>topology</th>\n",
       "      <th>gc_%</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>cds_number</th>\n",
       "      <th>coding_capacity</th>\n",
       "      <th>positive_strand_%</th>\n",
       "      <th>negative_strand_%</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reverse_complement</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GU339467</td>\n",
       "      <td>53332</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>64.530</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>92.857946</td>\n",
       "      <td>38.95</td>\n",
       "      <td>61.05</td>\n",
       "      <td>TGCGGCTGCCCCATCCTGTACGGGTTTCCAAGTCGATCTCGGGGGC...</td>\n",
       "      <td>AATGGCTGGCCAGTCCTTCTGTGGGGGACCCCCATAGGGGGGACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KT462701</td>\n",
       "      <td>3580</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.296</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>99.748603</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGAGACAGATATGACTAATCTGACATTCGCTTTCACTAATGAATGG...</td>\n",
       "      <td>ATCTCGGAAGACACGCCTACTAGGCCGTGGACTGAATAGTACCAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KT462700</td>\n",
       "      <td>3609</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>41.701</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>94.846218</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ACGTCTGTCGTCGCACGATTATTAAGGAGGTGCTATATGGAAATTT...</td>\n",
       "      <td>AGTACAGTCCCTTCTTGCTGTTCTATGTAAGACAGTTCTTATTAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KT462698</td>\n",
       "      <td>3776</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>44.147</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>98.013771</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CAGTAATCGCAATCCAAGGAAAGGGAAATTTACTTATCCCATTCTT...</td>\n",
       "      <td>GAACGGGGGCCGAAACCGCATTTCGGACGGTCGTTGATAAGCTAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MF417929</td>\n",
       "      <td>32618</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>39.218</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>88.635109</td>\n",
       "      <td>16.67</td>\n",
       "      <td>83.33</td>\n",
       "      <td>ATGTTGTCTAGTCCATCATAGGTGCAACGGATATACGAGCATTTTT...</td>\n",
       "      <td>AATTTTTATTAATTATTATTAGTTTTTCAATTATATTTTTTATATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MH616963</td>\n",
       "      <td>94878</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>28.507</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>91.829507</td>\n",
       "      <td>55.56</td>\n",
       "      <td>44.44</td>\n",
       "      <td>ATGTCGAGTATGAGGAATAAGAAACGCTTAATGACGGGGGAGTTCA...</td>\n",
       "      <td>TTAAAAAAGACTCTACTTAAAATTTGGCTACCACGCCGGCGACGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MH552500</td>\n",
       "      <td>90023</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>29.180</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>92.209769</td>\n",
       "      <td>50.00</td>\n",
       "      <td>50.00</td>\n",
       "      <td>ATGAGTTGTAATCTTTCAATAGACAAGAACATTATTCTTGATAAGA...</td>\n",
       "      <td>TTATATTTTCAAAAATATTAAATAAAAACAAACTAAAGTCTGATAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BK010471</td>\n",
       "      <td>97065</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>29.274</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>194.522227</td>\n",
       "      <td>50.00</td>\n",
       "      <td>50.00</td>\n",
       "      <td>TAGTAAGAGTGAAAATAATGTATATTGGTGTTTTGGTGAGGCTAGT...</td>\n",
       "      <td>AATGGTAATAGTAATGGTAATAATAGAAATAAAAGTAGAAGTAATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NR_074910</td>\n",
       "      <td>1544</td>\n",
       "      <td>0</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>54.663</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>CGAAGAAATTCCATTCCTCCACTAGGTTGGCGTCCAAGGGGATGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AC171169.</td>\n",
       "      <td>104373</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>38.783</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGTGCTATACAAAAGCGGGGGGCACGAGTCCACCCCGTTTTTCAC...</td>\n",
       "      <td>CCCTCACGTAAGGCGGTTATGTAAAAAAAATTCTCACCCTTTTTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BK014221</td>\n",
       "      <td>3524</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>57.577</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>89.131669</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GAGGGCTCTGCCCTCCTCTTCTAGATCATCCACTACCACAAGTGGG...</td>\n",
       "      <td>CCCCTCCTACGTTGCGTAGAAGGTGGAACCCGTTTAACCCGGGTAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BK014220</td>\n",
       "      <td>3326</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>53.428</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>95.971137</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTTTCCCAGTCACGATAAGCCACAAACCAACTTTGCGAGGTAGCTA...</td>\n",
       "      <td>CAAAGGGTCAGTGCTATGAGGAGGCCGATCAATCCATTGGTGGCTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BK014219</td>\n",
       "      <td>3642</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>44.783</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>97.391543</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTTTCCCAGTCACGATAACCCCATGGTGTAGGTGTCTTTGGGAGGC...</td>\n",
       "      <td>CAAAGGGTCAGTGCTATGTCAATGCTAAATTTAGGAAGTGCGCGGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>BK014218</td>\n",
       "      <td>3441</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>47.632</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>92.850915</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTCCCGAAAGGGACCTCTAGAACCTATGGTTCTCTCCGTCACGCAT...</td>\n",
       "      <td>GAGTCAAAATTTTCCCCCATCTTGGGGAGAGGTCTACTGATTTTTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>BK014217</td>\n",
       "      <td>3499</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.558</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>91.311803</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCGGCGCCCAGTGCCCTTCAGACGGAGTCAACTACACATGCACAAG...</td>\n",
       "      <td>GGGAGGAGAGGCTTGGTCGATGTTCAATCGACCGCCCGAGGAGTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BK014216</td>\n",
       "      <td>3533</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.259</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96.688367</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCCTTCCTGTCCTATCCATTCCGGTCCCTATTCTTCAGTTGGAAG...</td>\n",
       "      <td>GGCGGAGGATGCATAATTGCATCTGGGAGCAGGCGGAAAAGGAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>BK014215</td>\n",
       "      <td>3727</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.697</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>90.796888</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TCTCAGGGGCCTTTCTTGTAACCGTGTACTTTTCAGGCACGGCTGT...</td>\n",
       "      <td>CCCCCAACTCGTGTCGTTAGTTGGGGTACAATACTGAAAGCAGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>BK014214</td>\n",
       "      <td>3753</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.346</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>94.084732</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AAAGCTCCGGAACATCCTTTCCCGGAACTATCGCTATGGCTAACCT...</td>\n",
       "      <td>AGACGTGACCCTTCGCCCCGTATTCTCGGGTCCCCTTGTAGAAACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BK014213</td>\n",
       "      <td>3849</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.922</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>93.141076</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AACGGCTTCCAATCAATGGCCCTCGTTTAGGACGTGAGTATAATCA...</td>\n",
       "      <td>GAGACGAAGACGAAGCGATTCCTGGCAGGGTACATTTTACCAAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>BK014212</td>\n",
       "      <td>3592</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.782</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>92.455457</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCGAAGGCGTGGATAACCGTTCCATGGAGTCCCTGCATGAAATCCT...</td>\n",
       "      <td>CAAAAATATGACTTCGGGGTCGCAGCTCTCAAGGAATGAATCCACG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>BK014211</td>\n",
       "      <td>4061</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.018</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>95.666092</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CGCCACTTTACAAACCCAACGGTTATGAAAGGAAGGTTAATGACTT...</td>\n",
       "      <td>GAAACGCAGTCCCCCATCGGGTGCTCCCCGCTAACGATGTAGTTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>BK014210</td>\n",
       "      <td>3831</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>49.256</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96.162882</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GAGAGCACTTTGCTCTCCCTTATACGTATACTAATACGTATTGGCT...</td>\n",
       "      <td>CGGACGAGGTTCTTTGACACCTTCCAGGAGGAGATTTCCTTCCTCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>BK014209</td>\n",
       "      <td>3933</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.259</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96.643783</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGGATTGTCCCCCTCCTGCTCATCCTTGAGAACTGCTATGCCGAC...</td>\n",
       "      <td>TAGGCTACACTACGAATAGATAGAGCCATTACTAGGCTCTAAAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>BK014208</td>\n",
       "      <td>4032</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>53.596</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>90.401786</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGGCCCAGGCGTCATCCACTGGGACCGCTGTTGCGTCGCTCAGGG...</td>\n",
       "      <td>GAGACGAAATGTCCCATTAATGGGGTGGTCGGAAAGCCGCGAAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>BK014207</td>\n",
       "      <td>3512</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.804</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>94.988610</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CATGGGGGTTCCTTCTGCATGGATGGACAGGTATTTCAGTCAGTGA...</td>\n",
       "      <td>ATTGCGACCAATTGTCGCAAGGGGGCTCCGAGTTTTGCGGAGCAGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>BK014206</td>\n",
       "      <td>3526</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>39.989</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>96.596710</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TACTATGTCAGCAAAAATTCTTATTATCATTATATACGTTATCATG...</td>\n",
       "      <td>GCCGATAGGGTTATCCAAAGGTGTGGATGCCCTGGCTTAGAGTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BK014205</td>\n",
       "      <td>3646</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.962</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>89.687329</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCCAGGCATGCCTGGAGAGCCTCGGTGCATCGATTCCTCGATCAC...</td>\n",
       "      <td>TAGGTGAACCCCACCACGGGGATGCTGGGTTTGAACCCGGAGGTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BK014204</td>\n",
       "      <td>3681</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.757</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>90.627547</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCAGTCTATAGCCTGCGGTTCCACGACGTGTGGATTACGCAGGTA...</td>\n",
       "      <td>AAAGTAAGATAGGGGGCGCTCCCCCGAGGTGGCCACAAAGTGGCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>BK014203</td>\n",
       "      <td>3756</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>49.760</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>93.929712</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCTGGCTCCGCTTATCCATTAGCAGAGCTGCGGCGGGCTCATTGGC...</td>\n",
       "      <td>GACTCGTAGTACGAAGTAGTGTGGTCCGCCGAGGAATAGCTCCTCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>BK014202</td>\n",
       "      <td>3763</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.564</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>88.493223</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGATTGGATCTACTGCGAACGCTGTGAAGCGTAAGAGTAGCTCGA...</td>\n",
       "      <td>GCGCGGGGGTGGGTGAGGGCGATCCCTCCCGGCGGCCGACAGTGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>BK014201</td>\n",
       "      <td>3764</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>47.556</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>95.164718</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCCGCTATCAATTTGAGGAGTAACAAACCATGATTTATATTCACC...</td>\n",
       "      <td>GAAGGTTGTCTAATGCTAGTCACTAGTTGGGAGGCCCGTGAATTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>BK014200</td>\n",
       "      <td>3766</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.460</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>93.520977</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCCCATAATGGGAGGCTCTCGAGCGATAGCTTGCCGAATCCCTCG...</td>\n",
       "      <td>GAATGCGGCGGGAAGGAGACTTTCTTCTCGAACTTCTTAATAGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>BK014199</td>\n",
       "      <td>3809</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>50.223</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>94.197952</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTCACCTTGTTGGTGACACCGTATATTGAGATTCTCTCTGTATCGG...</td>\n",
       "      <td>GCGGTACGACAATGTAGTTGGTTGCACAAAGGCTGTGGAGGGTCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BK014198</td>\n",
       "      <td>3888</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.415</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96.373457</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCTCGCGGGGCCCTCGGCGTCGTCGACGCTGTTGTATTTTTCTACT...</td>\n",
       "      <td>GCCGTAAGATGACAAATACCAGTTACTGGGGGGTAGCTTGTGATCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>BK014197</td>\n",
       "      <td>3901</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>46.116</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>95.206357</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTTCCCCCTCTCTTTAGTGTCATACTTATTGTATGACGCGACTCAG...</td>\n",
       "      <td>TGGCGCACATATGGGTGGAAACTGACGAGGTTCATCATCAGATGGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BK014196</td>\n",
       "      <td>3911</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>54.641</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>92.354896</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGACGACAAGTGTCGTCCCTCCAGGCGAAAGCCTGATCCTTTTTA...</td>\n",
       "      <td>TTTGAGAGGCGCGGAGTCCTCTATATCTAACCCTTGTCGATTCGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BK014195</td>\n",
       "      <td>3966</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>50.429</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>98.310640</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGTCCGACTATGTCGGTGGGCAATCGCCCGGGATCCGGGTAAACCG...</td>\n",
       "      <td>GCCCGACCGAAATAGTCACGTTTATCTGTGGAGACCTTTAGGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>BK014194</td>\n",
       "      <td>4025</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>50.882</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>90.857143</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCCCTCTCGGGGAGCACCGTTTCAACAGCGGTGACTCTGTCTGCAG...</td>\n",
       "      <td>CCCTCAGAGAGCGTAGAGGACGGCAAGAGAATAGATACCAACCGAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>BK014193</td>\n",
       "      <td>4031</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>50.360</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>86.703051</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GAGACCCCGAGTGGGGCCTCCCGTAGAGTAATCTCTATCGATTTCA...</td>\n",
       "      <td>GCGGTGGGTTCATCAGAAACGAGCACTCTCGAGTCTTGATCACCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BK014192</td>\n",
       "      <td>4073</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.394</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>92.290695</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCGGGGACCCCGAAACTAGACTACCTGCTAGTGTAGGTGTAAGTTC...</td>\n",
       "      <td>GATGAATGAATTCCATCATTGGTGCCACGATCTGGTCGTGGCGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>BK014191</td>\n",
       "      <td>3477</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>47.800</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>97.382801</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCTAGCTACCTTGGTAGCTGTTTGCTCAAAAGAGCAGACCTGGGC...</td>\n",
       "      <td>TGGCTTAACCACGAATGGTTGTGGGGAATTTCTTTCCTCGGGGAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>BK014190</td>\n",
       "      <td>3553</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.112</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>92.288207</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CGTAAGGGCTGCCTTCCTTCATGCTAAGTTGGCACAAAGGAGTGAG...</td>\n",
       "      <td>CGAATACGAGGTAGGGTTCCACCTACTACTTAGGTGCGCGGGTCGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>BK014189</td>\n",
       "      <td>3863</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>47.010</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>94.123738</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTGGCCTCCTTCACTACTTCTCAGAAGACTCTTGTCTACTGGCAGA...</td>\n",
       "      <td>CAGCGAATGGTATAACCCCCCCCCAGCCGTGACGTTCACGTGCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>BK014188</td>\n",
       "      <td>3626</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.592</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>93.739658</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCCCATTGGGGGGCCCTGCATGCACTTGAGGAGGAGACACCATTAT...</td>\n",
       "      <td>ACGCTTACCGACGAGCCCCTTCGAACGACGTAACACGTCGAATCGA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           id  genome_length  jumbophage molecule_type  topology    gc_%  \\\n",
       "0    GU339467          53332           0           DNA    linear  64.530   \n",
       "1    KT462701           3580           0           RNA    linear  48.296   \n",
       "2    KT462700           3609           0           RNA    linear  41.701   \n",
       "3    KT462698           3776           0           RNA    linear  44.147   \n",
       "4    MF417929          32618           0           DNA    linear  39.218   \n",
       "5    MH616963          94878           0           DNA  circular  28.507   \n",
       "6    MH552500          90023           0           DNA  circular  29.180   \n",
       "7    BK010471          97065           0           DNA  circular  29.274   \n",
       "8   NR_074910           1544           0          rRNA    linear  54.663   \n",
       "9   AC171169.         104373           0           DNA    linear  38.783   \n",
       "10   BK014221           3524           0           RNA    linear  57.577   \n",
       "11   BK014220           3326           0           RNA    linear  53.428   \n",
       "12   BK014219           3642           0           RNA    linear  44.783   \n",
       "13   BK014218           3441           0           RNA    linear  47.632   \n",
       "14   BK014217           3499           0           RNA    linear  52.558   \n",
       "15   BK014216           3533           0           RNA    linear  48.259   \n",
       "16   BK014215           3727           0           RNA    linear  52.697   \n",
       "17   BK014214           3753           0           RNA    linear  51.346   \n",
       "18   BK014213           3849           0           RNA    linear  48.922   \n",
       "19   BK014212           3592           0           RNA    linear  51.782   \n",
       "20   BK014211           4061           0           RNA    linear  48.018   \n",
       "21   BK014210           3831           0           RNA    linear  49.256   \n",
       "22   BK014209           3933           0           RNA    linear  51.259   \n",
       "23   BK014208           4032           0           RNA    linear  53.596   \n",
       "24   BK014207           3512           0           RNA    linear  48.804   \n",
       "25   BK014206           3526           0           RNA    linear  39.989   \n",
       "26   BK014205           3646           0           RNA    linear  52.962   \n",
       "27   BK014204           3681           0           RNA    linear  52.757   \n",
       "28   BK014203           3756           0           RNA    linear  49.760   \n",
       "29   BK014202           3763           0           RNA    linear  52.564   \n",
       "30   BK014201           3764           0           RNA    linear  47.556   \n",
       "31   BK014200           3766           0           RNA    linear  51.460   \n",
       "32   BK014199           3809           0           RNA    linear  50.223   \n",
       "33   BK014198           3888           0           RNA    linear  51.415   \n",
       "34   BK014197           3901           0           RNA    linear  46.116   \n",
       "35   BK014196           3911           0           RNA    linear  54.641   \n",
       "36   BK014195           3966           0           RNA    linear  50.429   \n",
       "37   BK014194           4025           0           RNA    linear  50.882   \n",
       "38   BK014193           4031           0           RNA    linear  50.360   \n",
       "39   BK014192           4073           0           RNA    linear  52.394   \n",
       "40   BK014191           3477           0           RNA    linear  47.800   \n",
       "41   BK014190           3553           0           RNA    linear  51.112   \n",
       "42   BK014189           3863           0           RNA    linear  47.010   \n",
       "43   BK014188           3626           0           RNA    linear  52.592   \n",
       "\n",
       "    trna_count  cds_number  coding_capacity  positive_strand_%  \\\n",
       "0            1          95        92.857946              38.95   \n",
       "1            0           1        99.748603             100.00   \n",
       "2            0           3        94.846218             100.00   \n",
       "3            0           3        98.013771             100.00   \n",
       "4            0          42        88.635109              16.67   \n",
       "5            0          81        91.829507              55.56   \n",
       "6            0          80        92.209769              50.00   \n",
       "7            0          88       194.522227              50.00   \n",
       "8            0           0       100.000000               0.00   \n",
       "9            0           0         0.000000               0.00   \n",
       "10           0           3        89.131669             100.00   \n",
       "11           0           3        95.971137             100.00   \n",
       "12           0           3        97.391543             100.00   \n",
       "13           0           3        92.850915             100.00   \n",
       "14           0           3        91.311803             100.00   \n",
       "15           0           3        96.688367             100.00   \n",
       "16           0           3        90.796888             100.00   \n",
       "17           0           3        94.084732             100.00   \n",
       "18           0           3        93.141076             100.00   \n",
       "19           0           3        92.455457             100.00   \n",
       "20           0           4        95.666092             100.00   \n",
       "21           0           3        96.162882             100.00   \n",
       "22           0           3        96.643783             100.00   \n",
       "23           0           3        90.401786             100.00   \n",
       "24           0           4        94.988610             100.00   \n",
       "25           0           4        96.596710             100.00   \n",
       "26           0           3        89.687329             100.00   \n",
       "27           0           3        90.627547             100.00   \n",
       "28           0           3        93.929712             100.00   \n",
       "29           0           3        88.493223             100.00   \n",
       "30           0           4        95.164718             100.00   \n",
       "31           0           3        93.520977             100.00   \n",
       "32           0           4        94.197952             100.00   \n",
       "33           0           3        96.373457             100.00   \n",
       "34           0           4        95.206357             100.00   \n",
       "35           0           3        92.354896             100.00   \n",
       "36           0           3        98.310640             100.00   \n",
       "37           0           3        90.857143             100.00   \n",
       "38           0           3        86.703051             100.00   \n",
       "39           0           4        92.290695             100.00   \n",
       "40           0           3        97.382801             100.00   \n",
       "41           0           4        92.288207             100.00   \n",
       "42           0           3        94.123738             100.00   \n",
       "43           0           3        93.739658             100.00   \n",
       "\n",
       "    negative_strand_%                                           sequence  \\\n",
       "0               61.05  TGCGGCTGCCCCATCCTGTACGGGTTTCCAAGTCGATCTCGGGGGC...   \n",
       "1                0.00  GGAGACAGATATGACTAATCTGACATTCGCTTTCACTAATGAATGG...   \n",
       "2                0.00  ACGTCTGTCGTCGCACGATTATTAAGGAGGTGCTATATGGAAATTT...   \n",
       "3                0.00  CAGTAATCGCAATCCAAGGAAAGGGAAATTTACTTATCCCATTCTT...   \n",
       "4               83.33  ATGTTGTCTAGTCCATCATAGGTGCAACGGATATACGAGCATTTTT...   \n",
       "5               44.44  ATGTCGAGTATGAGGAATAAGAAACGCTTAATGACGGGGGAGTTCA...   \n",
       "6               50.00  ATGAGTTGTAATCTTTCAATAGACAAGAACATTATTCTTGATAAGA...   \n",
       "7               50.00  TAGTAAGAGTGAAAATAATGTATATTGGTGTTTTGGTGAGGCTAGT...   \n",
       "8                0.00  GTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "9                0.00  GGGTGCTATACAAAAGCGGGGGGCACGAGTCCACCCCGTTTTTCAC...   \n",
       "10               0.00  GAGGGCTCTGCCCTCCTCTTCTAGATCATCCACTACCACAAGTGGG...   \n",
       "11               0.00  GTTTCCCAGTCACGATAAGCCACAAACCAACTTTGCGAGGTAGCTA...   \n",
       "12               0.00  GTTTCCCAGTCACGATAACCCCATGGTGTAGGTGTCTTTGGGAGGC...   \n",
       "13               0.00  GTCCCGAAAGGGACCTCTAGAACCTATGGTTCTCTCCGTCACGCAT...   \n",
       "14               0.00  CCGGCGCCCAGTGCCCTTCAGACGGAGTCAACTACACATGCACAAG...   \n",
       "15               0.00  CTCCTTCCTGTCCTATCCATTCCGGTCCCTATTCTTCAGTTGGAAG...   \n",
       "16               0.00  TCTCAGGGGCCTTTCTTGTAACCGTGTACTTTTCAGGCACGGCTGT...   \n",
       "17               0.00  AAAGCTCCGGAACATCCTTTCCCGGAACTATCGCTATGGCTAACCT...   \n",
       "18               0.00  AACGGCTTCCAATCAATGGCCCTCGTTTAGGACGTGAGTATAATCA...   \n",
       "19               0.00  CCGAAGGCGTGGATAACCGTTCCATGGAGTCCCTGCATGAAATCCT...   \n",
       "20               0.00  CGCCACTTTACAAACCCAACGGTTATGAAAGGAAGGTTAATGACTT...   \n",
       "21               0.00  GAGAGCACTTTGCTCTCCCTTATACGTATACTAATACGTATTGGCT...   \n",
       "22               0.00  GGGGATTGTCCCCCTCCTGCTCATCCTTGAGAACTGCTATGCCGAC...   \n",
       "23               0.00  GGGGCCCAGGCGTCATCCACTGGGACCGCTGTTGCGTCGCTCAGGG...   \n",
       "24               0.00  CATGGGGGTTCCTTCTGCATGGATGGACAGGTATTTCAGTCAGTGA...   \n",
       "25               0.00  TACTATGTCAGCAAAAATTCTTATTATCATTATATACGTTATCATG...   \n",
       "26               0.00  CTCCAGGCATGCCTGGAGAGCCTCGGTGCATCGATTCCTCGATCAC...   \n",
       "27               0.00  CTCAGTCTATAGCCTGCGGTTCCACGACGTGTGGATTACGCAGGTA...   \n",
       "28               0.00  GCTGGCTCCGCTTATCCATTAGCAGAGCTGCGGCGGGCTCATTGGC...   \n",
       "29               0.00  GGGATTGGATCTACTGCGAACGCTGTGAAGCGTAAGAGTAGCTCGA...   \n",
       "30               0.00  CTCCGCTATCAATTTGAGGAGTAACAAACCATGATTTATATTCACC...   \n",
       "31               0.00  CTCCCATAATGGGAGGCTCTCGAGCGATAGCTTGCCGAATCCCTCG...   \n",
       "32               0.00  GTCACCTTGTTGGTGACACCGTATATTGAGATTCTCTCTGTATCGG...   \n",
       "33               0.00  CCTCGCGGGGCCCTCGGCGTCGTCGACGCTGTTGTATTTTTCTACT...   \n",
       "34               0.00  GTTCCCCCTCTCTTTAGTGTCATACTTATTGTATGACGCGACTCAG...   \n",
       "35               0.00  GGGACGACAAGTGTCGTCCCTCCAGGCGAAAGCCTGATCCTTTTTA...   \n",
       "36               0.00  GGTCCGACTATGTCGGTGGGCAATCGCCCGGGATCCGGGTAAACCG...   \n",
       "37               0.00  CCCCTCTCGGGGAGCACCGTTTCAACAGCGGTGACTCTGTCTGCAG...   \n",
       "38               0.00  GAGACCCCGAGTGGGGCCTCCCGTAGAGTAATCTCTATCGATTTCA...   \n",
       "39               0.00  GCGGGGACCCCGAAACTAGACTACCTGCTAGTGTAGGTGTAAGTTC...   \n",
       "40               0.00  CTCTAGCTACCTTGGTAGCTGTTTGCTCAAAAGAGCAGACCTGGGC...   \n",
       "41               0.00  CGTAAGGGCTGCCTTCCTTCATGCTAAGTTGGCACAAAGGAGTGAG...   \n",
       "42               0.00  CTGGCCTCCTTCACTACTTCTCAGAAGACTCTTGTCTACTGGCAGA...   \n",
       "43               0.00  CCCCATTGGGGGGCCCTGCATGCACTTGAGGAGGAGACACCATTAT...   \n",
       "\n",
       "                                   reverse_complement  \n",
       "0   AATGGCTGGCCAGTCCTTCTGTGGGGGACCCCCATAGGGGGGACCC...  \n",
       "1   ATCTCGGAAGACACGCCTACTAGGCCGTGGACTGAATAGTACCAGT...  \n",
       "2   AGTACAGTCCCTTCTTGCTGTTCTATGTAAGACAGTTCTTATTAGA...  \n",
       "3   GAACGGGGGCCGAAACCGCATTTCGGACGGTCGTTGATAAGCTAGT...  \n",
       "4   AATTTTTATTAATTATTATTAGTTTTTCAATTATATTTTTTATATT...  \n",
       "5   TTAAAAAAGACTCTACTTAAAATTTGGCTACCACGCCGGCGACGTG...  \n",
       "6   TTATATTTTCAAAAATATTAAATAAAAACAAACTAAAGTCTGATAA...  \n",
       "7   AATGGTAATAGTAATGGTAATAATAGAAATAAAAGTAGAAGTAATA...  \n",
       "8   CGAAGAAATTCCATTCCTCCACTAGGTTGGCGTCCAAGGGGATGCC...  \n",
       "9   CCCTCACGTAAGGCGGTTATGTAAAAAAAATTCTCACCCTTTTTAT...  \n",
       "10  CCCCTCCTACGTTGCGTAGAAGGTGGAACCCGTTTAACCCGGGTAG...  \n",
       "11  CAAAGGGTCAGTGCTATGAGGAGGCCGATCAATCCATTGGTGGCTG...  \n",
       "12  CAAAGGGTCAGTGCTATGTCAATGCTAAATTTAGGAAGTGCGCGGA...  \n",
       "13  GAGTCAAAATTTTCCCCCATCTTGGGGAGAGGTCTACTGATTTTTC...  \n",
       "14  GGGAGGAGAGGCTTGGTCGATGTTCAATCGACCGCCCGAGGAGTGA...  \n",
       "15  GGCGGAGGATGCATAATTGCATCTGGGAGCAGGCGGAAAAGGAATT...  \n",
       "16  CCCCCAACTCGTGTCGTTAGTTGGGGTACAATACTGAAAGCAGGAG...  \n",
       "17  AGACGTGACCCTTCGCCCCGTATTCTCGGGTCCCCTTGTAGAAACC...  \n",
       "18  GAGACGAAGACGAAGCGATTCCTGGCAGGGTACATTTTACCAAGCT...  \n",
       "19  CAAAAATATGACTTCGGGGTCGCAGCTCTCAAGGAATGAATCCACG...  \n",
       "20  GAAACGCAGTCCCCCATCGGGTGCTCCCCGCTAACGATGTAGTTGC...  \n",
       "21  CGGACGAGGTTCTTTGACACCTTCCAGGAGGAGATTTCCTTCCTCG...  \n",
       "22  TAGGCTACACTACGAATAGATAGAGCCATTACTAGGCTCTAAAGAT...  \n",
       "23  GAGACGAAATGTCCCATTAATGGGGTGGTCGGAAAGCCGCGAAGAT...  \n",
       "24  ATTGCGACCAATTGTCGCAAGGGGGCTCCGAGTTTTGCGGAGCAGT...  \n",
       "25  GCCGATAGGGTTATCCAAAGGTGTGGATGCCCTGGCTTAGAGTAAA...  \n",
       "26  TAGGTGAACCCCACCACGGGGATGCTGGGTTTGAACCCGGAGGTGA...  \n",
       "27  AAAGTAAGATAGGGGGCGCTCCCCCGAGGTGGCCACAAAGTGGCCC...  \n",
       "28  GACTCGTAGTACGAAGTAGTGTGGTCCGCCGAGGAATAGCTCCTCG...  \n",
       "29  GCGCGGGGGTGGGTGAGGGCGATCCCTCCCGGCGGCCGACAGTGCG...  \n",
       "30  GAAGGTTGTCTAATGCTAGTCACTAGTTGGGAGGCCCGTGAATTGG...  \n",
       "31  GAATGCGGCGGGAAGGAGACTTTCTTCTCGAACTTCTTAATAGAAG...  \n",
       "32  GCGGTACGACAATGTAGTTGGTTGCACAAAGGCTGTGGAGGGTCAA...  \n",
       "33  GCCGTAAGATGACAAATACCAGTTACTGGGGGGTAGCTTGTGATCG...  \n",
       "34  TGGCGCACATATGGGTGGAAACTGACGAGGTTCATCATCAGATGGT...  \n",
       "35  TTTGAGAGGCGCGGAGTCCTCTATATCTAACCCTTGTCGATTCGAC...  \n",
       "36  GCCCGACCGAAATAGTCACGTTTATCTGTGGAGACCTTTAGGCCCC...  \n",
       "37  CCCTCAGAGAGCGTAGAGGACGGCAAGAGAATAGATACCAACCGAC...  \n",
       "38  GCGGTGGGTTCATCAGAAACGAGCACTCTCGAGTCTTGATCACCTT...  \n",
       "39  GATGAATGAATTCCATCATTGGTGCCACGATCTGGTCGTGGCGGTG...  \n",
       "40  TGGCTTAACCACGAATGGTTGTGGGGAATTTCTTTCCTCGGGGAAT...  \n",
       "41  CGAATACGAGGTAGGGTTCCACCTACTACTTAGGTGCGCGGGTCGC...  \n",
       "42  CAGCGAATGGTATAACCCCCCCCCAGCCGTGACGTTCACGTGCTGG...  \n",
       "43  ACGCTTACCGACGAGCCCCTTCGAACGACGTAACACGTCGAATCGA...  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_file_path = \"../data/interim/genbank_engineering/sequences.gb\"\n",
    "# genbank_file_path = \"../data/interim/pharokka/pharokka.gbk\"\n",
    "# Lists to hold data\n",
    "ids = []\n",
    "genome_lengths = []\n",
    "gc_contents = []\n",
    "sequences = []\n",
    "reverse_complements = []\n",
    "cds_numbers = []\n",
    "positive_strands = []\n",
    "negative_strands = []\n",
    "coding_capacities = []\n",
    "molecule_types = []\n",
    "topologies = []\n",
    "trna_counts = []\n",
    "\n",
    "# Read the GenBank file\n",
    "for record in SeqIO.parse(genbank_file_path, \"genbank\"):\n",
    "    try:\n",
    "        # Attempt to access the sequence, which may raise UndefinedSequenceError\n",
    "        sequence = str(record.seq)\n",
    "        # print(record.id)\n",
    "    except UndefinedSequenceError:\n",
    "        # print(f\"Skipping record {record.id} as sequence is undefined.\")\n",
    "        continue  # Skip this record\n",
    "\n",
    "    # Calculate genome length and GC content\n",
    "    total_length = len(sequence)\n",
    "    gc_content = round((sequence.count('C') + sequence.count('G')) / total_length * 100, 3)\n",
    "\n",
    "    # Initialize counters\n",
    "    plus = 0\n",
    "    minus = 0\n",
    "    coding_count = 0\n",
    "    trna_count = 0\n",
    "    seen = set()  # Store seen barcodes\n",
    "\n",
    "    for feature in record.features:\n",
    "        start = feature.location.start\n",
    "        end = feature.location.end\n",
    "        length = len(FeatureLocation(start, end))\n",
    "        barcode = f\"{start}_{end}_{length}\"\n",
    "\n",
    "        if feature.type != 'source' and barcode not in seen:\n",
    "            coding_count += length\n",
    "            seen.add(barcode)\n",
    "\n",
    "        if feature.type == 'CDS':\n",
    "            if feature.location.strand == 1:\n",
    "                plus += 1\n",
    "            elif feature.location.strand == -1:\n",
    "                minus += 1\n",
    "        elif feature.type == 'tRNA':\n",
    "            trna_count += 1\n",
    "\n",
    "\n",
    "    # Calculate total number of CDS\n",
    "    total_CDS = plus + minus\n",
    "\n",
    "    # Calculate strand usage as a percentage\n",
    "    per_plus = round((plus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "    per_minus = round((minus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "\n",
    "    # Calculate coding capacity as a percentage\n",
    "    coding_capacity = (coding_count / total_length) * 100\n",
    "\n",
    "    # Extract molecule_type and topology\n",
    "    molecule_type = record.annotations.get('molecule_type', 'N/A')\n",
    "    topology = record.annotations.get('topology', 'N/A')\n",
    "\n",
    "    # Append data to lists\n",
    "    ids.append(record.id)\n",
    "    genome_lengths.append(total_length)\n",
    "    gc_contents.append(gc_content)\n",
    "    sequences.append(sequence)\n",
    "    reverse_complements.append(str(sequence[::-1]))\n",
    "    cds_numbers.append(total_CDS)\n",
    "    positive_strands.append(per_plus)\n",
    "    negative_strands.append(per_minus)\n",
    "    coding_capacities.append(coding_capacity)\n",
    "    molecule_types.append(molecule_type)\n",
    "    topologies.append(topology)\n",
    "    trna_counts.append(trna_count)\n",
    "\n",
    "# Convert lists to pandas DataFrame\n",
    "df = pd.DataFrame({\n",
    "    'id': ids,\n",
    "    'genome_length': genome_lengths,\n",
    "    'gc_%': gc_contents,\n",
    "    'sequence': sequences,\n",
    "    'reverse_complement': reverse_complements,\n",
    "    'cds_number': cds_numbers,\n",
    "    'positive_strand_%': positive_strands,\n",
    "    'negative_strand_%': negative_strands,\n",
    "    'coding_capacity': coding_capacities,\n",
    "    'molecule_type': molecule_types,\n",
    "    'topology': topologies,\n",
    "    'trna_count': trna_counts\n",
    "})\n",
    "\n",
    "\n",
    "# Filter df based on previously known sequences with a staining class\n",
    "accession_values = pd.read_csv('../data/processed/model_data.csv')\n",
    "accession_values = accession_values['Accession'].tolist()\n",
    "df['id'] = df['id'].str[:-2]\n",
    "# df = df[df['id'].isin(accession_values)]\n",
    "\n",
    "df['jumbophage'] = df['genome_length'].apply(lambda x: x >= 200000)\n",
    "# df = pd.get_dummies(df, columns=['molecule_type'])\n",
    "# columns_to_convert = ['Jumbophage', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']\n",
    "\n",
    "df['jumbophage'] = df['jumbophage'].astype(int)  # Convert True/False to 1/0\n",
    "# Change order of entries\n",
    "df = df[['id', 'genome_length', 'jumbophage', 'molecule_type', 'topology', 'gc_%', 'trna_count','cds_number', 'coding_capacity','positive_strand_%','negative_strand_%' ,'sequence', 'reverse_complement']]\n",
    "\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assign gram staining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>jumbophage</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>topology</th>\n",
       "      <th>gc_%</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>cds_number</th>\n",
       "      <th>coding_capacity</th>\n",
       "      <th>positive_strand_%</th>\n",
       "      <th>negative_strand_%</th>\n",
       "      <th>sequence</th>\n",
       "      <th>reverse_complement</th>\n",
       "      <th>staining</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GU339467</td>\n",
       "      <td>53332</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>64.530</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>92.857946</td>\n",
       "      <td>38.95</td>\n",
       "      <td>61.05</td>\n",
       "      <td>TGCGGCTGCCCCATCCTGTACGGGTTTCCAAGTCGATCTCGGGGGC...</td>\n",
       "      <td>AATGGCTGGCCAGTCCTTCTGTGGGGGACCCCCATAGGGGGGACCC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KT462701</td>\n",
       "      <td>3580</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.296</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>99.748603</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGAGACAGATATGACTAATCTGACATTCGCTTTCACTAATGAATGG...</td>\n",
       "      <td>ATCTCGGAAGACACGCCTACTAGGCCGTGGACTGAATAGTACCAGT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KT462700</td>\n",
       "      <td>3609</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>41.701</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>94.846218</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>ACGTCTGTCGTCGCACGATTATTAAGGAGGTGCTATATGGAAATTT...</td>\n",
       "      <td>AGTACAGTCCCTTCTTGCTGTTCTATGTAAGACAGTTCTTATTAGA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KT462698</td>\n",
       "      <td>3776</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>44.147</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>98.013771</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CAGTAATCGCAATCCAAGGAAAGGGAAATTTACTTATCCCATTCTT...</td>\n",
       "      <td>GAACGGGGGCCGAAACCGCATTTCGGACGGTCGTTGATAAGCTAGT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MF417929</td>\n",
       "      <td>32618</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>39.218</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>88.635109</td>\n",
       "      <td>16.67</td>\n",
       "      <td>83.33</td>\n",
       "      <td>ATGTTGTCTAGTCCATCATAGGTGCAACGGATATACGAGCATTTTT...</td>\n",
       "      <td>AATTTTTATTAATTATTATTAGTTTTTCAATTATATTTTTTATATT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MH616963</td>\n",
       "      <td>94878</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>28.507</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>91.829507</td>\n",
       "      <td>55.56</td>\n",
       "      <td>44.44</td>\n",
       "      <td>ATGTCGAGTATGAGGAATAAGAAACGCTTAATGACGGGGGAGTTCA...</td>\n",
       "      <td>TTAAAAAAGACTCTACTTAAAATTTGGCTACCACGCCGGCGACGTG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MH552500</td>\n",
       "      <td>90023</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>29.180</td>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>92.209769</td>\n",
       "      <td>50.00</td>\n",
       "      <td>50.00</td>\n",
       "      <td>ATGAGTTGTAATCTTTCAATAGACAAGAACATTATTCTTGATAAGA...</td>\n",
       "      <td>TTATATTTTCAAAAATATTAAATAAAAACAAACTAAAGTCTGATAA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BK010471</td>\n",
       "      <td>97065</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>29.274</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>194.522227</td>\n",
       "      <td>50.00</td>\n",
       "      <td>50.00</td>\n",
       "      <td>TAGTAAGAGTGAAAATAATGTATATTGGTGTTTTGGTGAGGCTAGT...</td>\n",
       "      <td>AATGGTAATAGTAATGGTAATAATAGAAATAAAAGTAGAAGTAATA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NR_074910</td>\n",
       "      <td>1544</td>\n",
       "      <td>0</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>54.663</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...</td>\n",
       "      <td>CGAAGAAATTCCATTCCTCCACTAGGTTGGCGTCCAAGGGGATGCC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AC171169.</td>\n",
       "      <td>104373</td>\n",
       "      <td>0</td>\n",
       "      <td>DNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>38.783</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGTGCTATACAAAAGCGGGGGGCACGAGTCCACCCCGTTTTTCAC...</td>\n",
       "      <td>CCCTCACGTAAGGCGGTTATGTAAAAAAAATTCTCACCCTTTTTAT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BK014221</td>\n",
       "      <td>3524</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>57.577</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>89.131669</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GAGGGCTCTGCCCTCCTCTTCTAGATCATCCACTACCACAAGTGGG...</td>\n",
       "      <td>CCCCTCCTACGTTGCGTAGAAGGTGGAACCCGTTTAACCCGGGTAG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BK014220</td>\n",
       "      <td>3326</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>53.428</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>95.971137</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTTTCCCAGTCACGATAAGCCACAAACCAACTTTGCGAGGTAGCTA...</td>\n",
       "      <td>CAAAGGGTCAGTGCTATGAGGAGGCCGATCAATCCATTGGTGGCTG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>BK014219</td>\n",
       "      <td>3642</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>44.783</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>97.391543</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTTTCCCAGTCACGATAACCCCATGGTGTAGGTGTCTTTGGGAGGC...</td>\n",
       "      <td>CAAAGGGTCAGTGCTATGTCAATGCTAAATTTAGGAAGTGCGCGGA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>BK014218</td>\n",
       "      <td>3441</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>47.632</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>92.850915</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTCCCGAAAGGGACCTCTAGAACCTATGGTTCTCTCCGTCACGCAT...</td>\n",
       "      <td>GAGTCAAAATTTTCCCCCATCTTGGGGAGAGGTCTACTGATTTTTC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>BK014217</td>\n",
       "      <td>3499</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.558</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>91.311803</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCGGCGCCCAGTGCCCTTCAGACGGAGTCAACTACACATGCACAAG...</td>\n",
       "      <td>GGGAGGAGAGGCTTGGTCGATGTTCAATCGACCGCCCGAGGAGTGA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>BK014216</td>\n",
       "      <td>3533</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.259</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96.688367</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCCTTCCTGTCCTATCCATTCCGGTCCCTATTCTTCAGTTGGAAG...</td>\n",
       "      <td>GGCGGAGGATGCATAATTGCATCTGGGAGCAGGCGGAAAAGGAATT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>BK014215</td>\n",
       "      <td>3727</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.697</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>90.796888</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TCTCAGGGGCCTTTCTTGTAACCGTGTACTTTTCAGGCACGGCTGT...</td>\n",
       "      <td>CCCCCAACTCGTGTCGTTAGTTGGGGTACAATACTGAAAGCAGGAG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>BK014214</td>\n",
       "      <td>3753</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.346</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>94.084732</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AAAGCTCCGGAACATCCTTTCCCGGAACTATCGCTATGGCTAACCT...</td>\n",
       "      <td>AGACGTGACCCTTCGCCCCGTATTCTCGGGTCCCCTTGTAGAAACC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BK014213</td>\n",
       "      <td>3849</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.922</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>93.141076</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>AACGGCTTCCAATCAATGGCCCTCGTTTAGGACGTGAGTATAATCA...</td>\n",
       "      <td>GAGACGAAGACGAAGCGATTCCTGGCAGGGTACATTTTACCAAGCT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>BK014212</td>\n",
       "      <td>3592</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.782</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>92.455457</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCGAAGGCGTGGATAACCGTTCCATGGAGTCCCTGCATGAAATCCT...</td>\n",
       "      <td>CAAAAATATGACTTCGGGGTCGCAGCTCTCAAGGAATGAATCCACG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>BK014211</td>\n",
       "      <td>4061</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.018</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>95.666092</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CGCCACTTTACAAACCCAACGGTTATGAAAGGAAGGTTAATGACTT...</td>\n",
       "      <td>GAAACGCAGTCCCCCATCGGGTGCTCCCCGCTAACGATGTAGTTGC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>BK014210</td>\n",
       "      <td>3831</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>49.256</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96.162882</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GAGAGCACTTTGCTCTCCCTTATACGTATACTAATACGTATTGGCT...</td>\n",
       "      <td>CGGACGAGGTTCTTTGACACCTTCCAGGAGGAGATTTCCTTCCTCG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>BK014209</td>\n",
       "      <td>3933</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.259</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96.643783</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGGATTGTCCCCCTCCTGCTCATCCTTGAGAACTGCTATGCCGAC...</td>\n",
       "      <td>TAGGCTACACTACGAATAGATAGAGCCATTACTAGGCTCTAAAGAT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>BK014208</td>\n",
       "      <td>4032</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>53.596</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>90.401786</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGGCCCAGGCGTCATCCACTGGGACCGCTGTTGCGTCGCTCAGGG...</td>\n",
       "      <td>GAGACGAAATGTCCCATTAATGGGGTGGTCGGAAAGCCGCGAAGAT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>BK014207</td>\n",
       "      <td>3512</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>48.804</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>94.988610</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CATGGGGGTTCCTTCTGCATGGATGGACAGGTATTTCAGTCAGTGA...</td>\n",
       "      <td>ATTGCGACCAATTGTCGCAAGGGGGCTCCGAGTTTTGCGGAGCAGT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>BK014206</td>\n",
       "      <td>3526</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>39.989</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>96.596710</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>TACTATGTCAGCAAAAATTCTTATTATCATTATATACGTTATCATG...</td>\n",
       "      <td>GCCGATAGGGTTATCCAAAGGTGTGGATGCCCTGGCTTAGAGTAAA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>BK014205</td>\n",
       "      <td>3646</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.962</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>89.687329</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCCAGGCATGCCTGGAGAGCCTCGGTGCATCGATTCCTCGATCAC...</td>\n",
       "      <td>TAGGTGAACCCCACCACGGGGATGCTGGGTTTGAACCCGGAGGTGA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>BK014204</td>\n",
       "      <td>3681</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.757</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>90.627547</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCAGTCTATAGCCTGCGGTTCCACGACGTGTGGATTACGCAGGTA...</td>\n",
       "      <td>AAAGTAAGATAGGGGGCGCTCCCCCGAGGTGGCCACAAAGTGGCCC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>BK014203</td>\n",
       "      <td>3756</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>49.760</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>93.929712</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCTGGCTCCGCTTATCCATTAGCAGAGCTGCGGCGGGCTCATTGGC...</td>\n",
       "      <td>GACTCGTAGTACGAAGTAGTGTGGTCCGCCGAGGAATAGCTCCTCG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>BK014202</td>\n",
       "      <td>3763</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.564</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>88.493223</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGATTGGATCTACTGCGAACGCTGTGAAGCGTAAGAGTAGCTCGA...</td>\n",
       "      <td>GCGCGGGGGTGGGTGAGGGCGATCCCTCCCGGCGGCCGACAGTGCG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>BK014201</td>\n",
       "      <td>3764</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>47.556</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>95.164718</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCCGCTATCAATTTGAGGAGTAACAAACCATGATTTATATTCACC...</td>\n",
       "      <td>GAAGGTTGTCTAATGCTAGTCACTAGTTGGGAGGCCCGTGAATTGG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>BK014200</td>\n",
       "      <td>3766</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.460</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>93.520977</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCCCATAATGGGAGGCTCTCGAGCGATAGCTTGCCGAATCCCTCG...</td>\n",
       "      <td>GAATGCGGCGGGAAGGAGACTTTCTTCTCGAACTTCTTAATAGAAG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>BK014199</td>\n",
       "      <td>3809</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>50.223</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>94.197952</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTCACCTTGTTGGTGACACCGTATATTGAGATTCTCTCTGTATCGG...</td>\n",
       "      <td>GCGGTACGACAATGTAGTTGGTTGCACAAAGGCTGTGGAGGGTCAA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>BK014198</td>\n",
       "      <td>3888</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.415</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96.373457</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCTCGCGGGGCCCTCGGCGTCGTCGACGCTGTTGTATTTTTCTACT...</td>\n",
       "      <td>GCCGTAAGATGACAAATACCAGTTACTGGGGGGTAGCTTGTGATCG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>BK014197</td>\n",
       "      <td>3901</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>46.116</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>95.206357</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GTTCCCCCTCTCTTTAGTGTCATACTTATTGTATGACGCGACTCAG...</td>\n",
       "      <td>TGGCGCACATATGGGTGGAAACTGACGAGGTTCATCATCAGATGGT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>BK014196</td>\n",
       "      <td>3911</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>54.641</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>92.354896</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGGACGACAAGTGTCGTCCCTCCAGGCGAAAGCCTGATCCTTTTTA...</td>\n",
       "      <td>TTTGAGAGGCGCGGAGTCCTCTATATCTAACCCTTGTCGATTCGAC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>BK014195</td>\n",
       "      <td>3966</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>50.429</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>98.310640</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GGTCCGACTATGTCGGTGGGCAATCGCCCGGGATCCGGGTAAACCG...</td>\n",
       "      <td>GCCCGACCGAAATAGTCACGTTTATCTGTGGAGACCTTTAGGCCCC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>BK014194</td>\n",
       "      <td>4025</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>50.882</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>90.857143</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCCCTCTCGGGGAGCACCGTTTCAACAGCGGTGACTCTGTCTGCAG...</td>\n",
       "      <td>CCCTCAGAGAGCGTAGAGGACGGCAAGAGAATAGATACCAACCGAC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>BK014193</td>\n",
       "      <td>4031</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>50.360</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>86.703051</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GAGACCCCGAGTGGGGCCTCCCGTAGAGTAATCTCTATCGATTTCA...</td>\n",
       "      <td>GCGGTGGGTTCATCAGAAACGAGCACTCTCGAGTCTTGATCACCTT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>BK014192</td>\n",
       "      <td>4073</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.394</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>92.290695</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>GCGGGGACCCCGAAACTAGACTACCTGCTAGTGTAGGTGTAAGTTC...</td>\n",
       "      <td>GATGAATGAATTCCATCATTGGTGCCACGATCTGGTCGTGGCGGTG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>BK014191</td>\n",
       "      <td>3477</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>47.800</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>97.382801</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTCTAGCTACCTTGGTAGCTGTTTGCTCAAAAGAGCAGACCTGGGC...</td>\n",
       "      <td>TGGCTTAACCACGAATGGTTGTGGGGAATTTCTTTCCTCGGGGAAT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>BK014190</td>\n",
       "      <td>3553</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>51.112</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>92.288207</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CGTAAGGGCTGCCTTCCTTCATGCTAAGTTGGCACAAAGGAGTGAG...</td>\n",
       "      <td>CGAATACGAGGTAGGGTTCCACCTACTACTTAGGTGCGCGGGTCGC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>BK014189</td>\n",
       "      <td>3863</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>47.010</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>94.123738</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CTGGCCTCCTTCACTACTTCTCAGAAGACTCTTGTCTACTGGCAGA...</td>\n",
       "      <td>CAGCGAATGGTATAACCCCCCCCCAGCCGTGACGTTCACGTGCTGG...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>BK014188</td>\n",
       "      <td>3626</td>\n",
       "      <td>0</td>\n",
       "      <td>RNA</td>\n",
       "      <td>linear</td>\n",
       "      <td>52.592</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>93.739658</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>CCCCATTGGGGGGCCCTGCATGCACTTGAGGAGGAGACACCATTAT...</td>\n",
       "      <td>ACGCTTACCGACGAGCCCCTTCGAACGACGTAACACGTCGAATCGA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           id  genome_length  jumbophage molecule_type  topology    gc_%  \\\n",
       "0    GU339467          53332           0           DNA    linear  64.530   \n",
       "1    KT462701           3580           0           RNA    linear  48.296   \n",
       "2    KT462700           3609           0           RNA    linear  41.701   \n",
       "3    KT462698           3776           0           RNA    linear  44.147   \n",
       "4    MF417929          32618           0           DNA    linear  39.218   \n",
       "5    MH616963          94878           0           DNA  circular  28.507   \n",
       "6    MH552500          90023           0           DNA  circular  29.180   \n",
       "7    BK010471          97065           0           DNA  circular  29.274   \n",
       "8   NR_074910           1544           0          rRNA    linear  54.663   \n",
       "9   AC171169.         104373           0           DNA    linear  38.783   \n",
       "10   BK014221           3524           0           RNA    linear  57.577   \n",
       "11   BK014220           3326           0           RNA    linear  53.428   \n",
       "12   BK014219           3642           0           RNA    linear  44.783   \n",
       "13   BK014218           3441           0           RNA    linear  47.632   \n",
       "14   BK014217           3499           0           RNA    linear  52.558   \n",
       "15   BK014216           3533           0           RNA    linear  48.259   \n",
       "16   BK014215           3727           0           RNA    linear  52.697   \n",
       "17   BK014214           3753           0           RNA    linear  51.346   \n",
       "18   BK014213           3849           0           RNA    linear  48.922   \n",
       "19   BK014212           3592           0           RNA    linear  51.782   \n",
       "20   BK014211           4061           0           RNA    linear  48.018   \n",
       "21   BK014210           3831           0           RNA    linear  49.256   \n",
       "22   BK014209           3933           0           RNA    linear  51.259   \n",
       "23   BK014208           4032           0           RNA    linear  53.596   \n",
       "24   BK014207           3512           0           RNA    linear  48.804   \n",
       "25   BK014206           3526           0           RNA    linear  39.989   \n",
       "26   BK014205           3646           0           RNA    linear  52.962   \n",
       "27   BK014204           3681           0           RNA    linear  52.757   \n",
       "28   BK014203           3756           0           RNA    linear  49.760   \n",
       "29   BK014202           3763           0           RNA    linear  52.564   \n",
       "30   BK014201           3764           0           RNA    linear  47.556   \n",
       "31   BK014200           3766           0           RNA    linear  51.460   \n",
       "32   BK014199           3809           0           RNA    linear  50.223   \n",
       "33   BK014198           3888           0           RNA    linear  51.415   \n",
       "34   BK014197           3901           0           RNA    linear  46.116   \n",
       "35   BK014196           3911           0           RNA    linear  54.641   \n",
       "36   BK014195           3966           0           RNA    linear  50.429   \n",
       "37   BK014194           4025           0           RNA    linear  50.882   \n",
       "38   BK014193           4031           0           RNA    linear  50.360   \n",
       "39   BK014192           4073           0           RNA    linear  52.394   \n",
       "40   BK014191           3477           0           RNA    linear  47.800   \n",
       "41   BK014190           3553           0           RNA    linear  51.112   \n",
       "42   BK014189           3863           0           RNA    linear  47.010   \n",
       "43   BK014188           3626           0           RNA    linear  52.592   \n",
       "\n",
       "    trna_count  cds_number  coding_capacity  positive_strand_%  \\\n",
       "0            1          95        92.857946              38.95   \n",
       "1            0           1        99.748603             100.00   \n",
       "2            0           3        94.846218             100.00   \n",
       "3            0           3        98.013771             100.00   \n",
       "4            0          42        88.635109              16.67   \n",
       "5            0          81        91.829507              55.56   \n",
       "6            0          80        92.209769              50.00   \n",
       "7            0          88       194.522227              50.00   \n",
       "8            0           0       100.000000               0.00   \n",
       "9            0           0         0.000000               0.00   \n",
       "10           0           3        89.131669             100.00   \n",
       "11           0           3        95.971137             100.00   \n",
       "12           0           3        97.391543             100.00   \n",
       "13           0           3        92.850915             100.00   \n",
       "14           0           3        91.311803             100.00   \n",
       "15           0           3        96.688367             100.00   \n",
       "16           0           3        90.796888             100.00   \n",
       "17           0           3        94.084732             100.00   \n",
       "18           0           3        93.141076             100.00   \n",
       "19           0           3        92.455457             100.00   \n",
       "20           0           4        95.666092             100.00   \n",
       "21           0           3        96.162882             100.00   \n",
       "22           0           3        96.643783             100.00   \n",
       "23           0           3        90.401786             100.00   \n",
       "24           0           4        94.988610             100.00   \n",
       "25           0           4        96.596710             100.00   \n",
       "26           0           3        89.687329             100.00   \n",
       "27           0           3        90.627547             100.00   \n",
       "28           0           3        93.929712             100.00   \n",
       "29           0           3        88.493223             100.00   \n",
       "30           0           4        95.164718             100.00   \n",
       "31           0           3        93.520977             100.00   \n",
       "32           0           4        94.197952             100.00   \n",
       "33           0           3        96.373457             100.00   \n",
       "34           0           4        95.206357             100.00   \n",
       "35           0           3        92.354896             100.00   \n",
       "36           0           3        98.310640             100.00   \n",
       "37           0           3        90.857143             100.00   \n",
       "38           0           3        86.703051             100.00   \n",
       "39           0           4        92.290695             100.00   \n",
       "40           0           3        97.382801             100.00   \n",
       "41           0           4        92.288207             100.00   \n",
       "42           0           3        94.123738             100.00   \n",
       "43           0           3        93.739658             100.00   \n",
       "\n",
       "    negative_strand_%                                           sequence  \\\n",
       "0               61.05  TGCGGCTGCCCCATCCTGTACGGGTTTCCAAGTCGATCTCGGGGGC...   \n",
       "1                0.00  GGAGACAGATATGACTAATCTGACATTCGCTTTCACTAATGAATGG...   \n",
       "2                0.00  ACGTCTGTCGTCGCACGATTATTAAGGAGGTGCTATATGGAAATTT...   \n",
       "3                0.00  CAGTAATCGCAATCCAAGGAAAGGGAAATTTACTTATCCCATTCTT...   \n",
       "4               83.33  ATGTTGTCTAGTCCATCATAGGTGCAACGGATATACGAGCATTTTT...   \n",
       "5               44.44  ATGTCGAGTATGAGGAATAAGAAACGCTTAATGACGGGGGAGTTCA...   \n",
       "6               50.00  ATGAGTTGTAATCTTTCAATAGACAAGAACATTATTCTTGATAAGA...   \n",
       "7               50.00  TAGTAAGAGTGAAAATAATGTATATTGGTGTTTTGGTGAGGCTAGT...   \n",
       "8                0.00  GTTTGATCATGGCTCAGATTGAACGCTGGCGGCAGGCCTAACACAT...   \n",
       "9                0.00  GGGTGCTATACAAAAGCGGGGGGCACGAGTCCACCCCGTTTTTCAC...   \n",
       "10               0.00  GAGGGCTCTGCCCTCCTCTTCTAGATCATCCACTACCACAAGTGGG...   \n",
       "11               0.00  GTTTCCCAGTCACGATAAGCCACAAACCAACTTTGCGAGGTAGCTA...   \n",
       "12               0.00  GTTTCCCAGTCACGATAACCCCATGGTGTAGGTGTCTTTGGGAGGC...   \n",
       "13               0.00  GTCCCGAAAGGGACCTCTAGAACCTATGGTTCTCTCCGTCACGCAT...   \n",
       "14               0.00  CCGGCGCCCAGTGCCCTTCAGACGGAGTCAACTACACATGCACAAG...   \n",
       "15               0.00  CTCCTTCCTGTCCTATCCATTCCGGTCCCTATTCTTCAGTTGGAAG...   \n",
       "16               0.00  TCTCAGGGGCCTTTCTTGTAACCGTGTACTTTTCAGGCACGGCTGT...   \n",
       "17               0.00  AAAGCTCCGGAACATCCTTTCCCGGAACTATCGCTATGGCTAACCT...   \n",
       "18               0.00  AACGGCTTCCAATCAATGGCCCTCGTTTAGGACGTGAGTATAATCA...   \n",
       "19               0.00  CCGAAGGCGTGGATAACCGTTCCATGGAGTCCCTGCATGAAATCCT...   \n",
       "20               0.00  CGCCACTTTACAAACCCAACGGTTATGAAAGGAAGGTTAATGACTT...   \n",
       "21               0.00  GAGAGCACTTTGCTCTCCCTTATACGTATACTAATACGTATTGGCT...   \n",
       "22               0.00  GGGGATTGTCCCCCTCCTGCTCATCCTTGAGAACTGCTATGCCGAC...   \n",
       "23               0.00  GGGGCCCAGGCGTCATCCACTGGGACCGCTGTTGCGTCGCTCAGGG...   \n",
       "24               0.00  CATGGGGGTTCCTTCTGCATGGATGGACAGGTATTTCAGTCAGTGA...   \n",
       "25               0.00  TACTATGTCAGCAAAAATTCTTATTATCATTATATACGTTATCATG...   \n",
       "26               0.00  CTCCAGGCATGCCTGGAGAGCCTCGGTGCATCGATTCCTCGATCAC...   \n",
       "27               0.00  CTCAGTCTATAGCCTGCGGTTCCACGACGTGTGGATTACGCAGGTA...   \n",
       "28               0.00  GCTGGCTCCGCTTATCCATTAGCAGAGCTGCGGCGGGCTCATTGGC...   \n",
       "29               0.00  GGGATTGGATCTACTGCGAACGCTGTGAAGCGTAAGAGTAGCTCGA...   \n",
       "30               0.00  CTCCGCTATCAATTTGAGGAGTAACAAACCATGATTTATATTCACC...   \n",
       "31               0.00  CTCCCATAATGGGAGGCTCTCGAGCGATAGCTTGCCGAATCCCTCG...   \n",
       "32               0.00  GTCACCTTGTTGGTGACACCGTATATTGAGATTCTCTCTGTATCGG...   \n",
       "33               0.00  CCTCGCGGGGCCCTCGGCGTCGTCGACGCTGTTGTATTTTTCTACT...   \n",
       "34               0.00  GTTCCCCCTCTCTTTAGTGTCATACTTATTGTATGACGCGACTCAG...   \n",
       "35               0.00  GGGACGACAAGTGTCGTCCCTCCAGGCGAAAGCCTGATCCTTTTTA...   \n",
       "36               0.00  GGTCCGACTATGTCGGTGGGCAATCGCCCGGGATCCGGGTAAACCG...   \n",
       "37               0.00  CCCCTCTCGGGGAGCACCGTTTCAACAGCGGTGACTCTGTCTGCAG...   \n",
       "38               0.00  GAGACCCCGAGTGGGGCCTCCCGTAGAGTAATCTCTATCGATTTCA...   \n",
       "39               0.00  GCGGGGACCCCGAAACTAGACTACCTGCTAGTGTAGGTGTAAGTTC...   \n",
       "40               0.00  CTCTAGCTACCTTGGTAGCTGTTTGCTCAAAAGAGCAGACCTGGGC...   \n",
       "41               0.00  CGTAAGGGCTGCCTTCCTTCATGCTAAGTTGGCACAAAGGAGTGAG...   \n",
       "42               0.00  CTGGCCTCCTTCACTACTTCTCAGAAGACTCTTGTCTACTGGCAGA...   \n",
       "43               0.00  CCCCATTGGGGGGCCCTGCATGCACTTGAGGAGGAGACACCATTAT...   \n",
       "\n",
       "                                   reverse_complement staining  \n",
       "0   AATGGCTGGCCAGTCCTTCTGTGGGGGACCCCCATAGGGGGGACCC...      NaN  \n",
       "1   ATCTCGGAAGACACGCCTACTAGGCCGTGGACTGAATAGTACCAGT...      NaN  \n",
       "2   AGTACAGTCCCTTCTTGCTGTTCTATGTAAGACAGTTCTTATTAGA...      NaN  \n",
       "3   GAACGGGGGCCGAAACCGCATTTCGGACGGTCGTTGATAAGCTAGT...      NaN  \n",
       "4   AATTTTTATTAATTATTATTAGTTTTTCAATTATATTTTTTATATT...      NaN  \n",
       "5   TTAAAAAAGACTCTACTTAAAATTTGGCTACCACGCCGGCGACGTG...      NaN  \n",
       "6   TTATATTTTCAAAAATATTAAATAAAAACAAACTAAAGTCTGATAA...      NaN  \n",
       "7   AATGGTAATAGTAATGGTAATAATAGAAATAAAAGTAGAAGTAATA...      NaN  \n",
       "8   CGAAGAAATTCCATTCCTCCACTAGGTTGGCGTCCAAGGGGATGCC...      NaN  \n",
       "9   CCCTCACGTAAGGCGGTTATGTAAAAAAAATTCTCACCCTTTTTAT...      NaN  \n",
       "10  CCCCTCCTACGTTGCGTAGAAGGTGGAACCCGTTTAACCCGGGTAG...      NaN  \n",
       "11  CAAAGGGTCAGTGCTATGAGGAGGCCGATCAATCCATTGGTGGCTG...      NaN  \n",
       "12  CAAAGGGTCAGTGCTATGTCAATGCTAAATTTAGGAAGTGCGCGGA...      NaN  \n",
       "13  GAGTCAAAATTTTCCCCCATCTTGGGGAGAGGTCTACTGATTTTTC...      NaN  \n",
       "14  GGGAGGAGAGGCTTGGTCGATGTTCAATCGACCGCCCGAGGAGTGA...      NaN  \n",
       "15  GGCGGAGGATGCATAATTGCATCTGGGAGCAGGCGGAAAAGGAATT...      NaN  \n",
       "16  CCCCCAACTCGTGTCGTTAGTTGGGGTACAATACTGAAAGCAGGAG...      NaN  \n",
       "17  AGACGTGACCCTTCGCCCCGTATTCTCGGGTCCCCTTGTAGAAACC...      NaN  \n",
       "18  GAGACGAAGACGAAGCGATTCCTGGCAGGGTACATTTTACCAAGCT...      NaN  \n",
       "19  CAAAAATATGACTTCGGGGTCGCAGCTCTCAAGGAATGAATCCACG...      NaN  \n",
       "20  GAAACGCAGTCCCCCATCGGGTGCTCCCCGCTAACGATGTAGTTGC...      NaN  \n",
       "21  CGGACGAGGTTCTTTGACACCTTCCAGGAGGAGATTTCCTTCCTCG...      NaN  \n",
       "22  TAGGCTACACTACGAATAGATAGAGCCATTACTAGGCTCTAAAGAT...      NaN  \n",
       "23  GAGACGAAATGTCCCATTAATGGGGTGGTCGGAAAGCCGCGAAGAT...      NaN  \n",
       "24  ATTGCGACCAATTGTCGCAAGGGGGCTCCGAGTTTTGCGGAGCAGT...      NaN  \n",
       "25  GCCGATAGGGTTATCCAAAGGTGTGGATGCCCTGGCTTAGAGTAAA...      NaN  \n",
       "26  TAGGTGAACCCCACCACGGGGATGCTGGGTTTGAACCCGGAGGTGA...      NaN  \n",
       "27  AAAGTAAGATAGGGGGCGCTCCCCCGAGGTGGCCACAAAGTGGCCC...      NaN  \n",
       "28  GACTCGTAGTACGAAGTAGTGTGGTCCGCCGAGGAATAGCTCCTCG...      NaN  \n",
       "29  GCGCGGGGGTGGGTGAGGGCGATCCCTCCCGGCGGCCGACAGTGCG...      NaN  \n",
       "30  GAAGGTTGTCTAATGCTAGTCACTAGTTGGGAGGCCCGTGAATTGG...      NaN  \n",
       "31  GAATGCGGCGGGAAGGAGACTTTCTTCTCGAACTTCTTAATAGAAG...      NaN  \n",
       "32  GCGGTACGACAATGTAGTTGGTTGCACAAAGGCTGTGGAGGGTCAA...      NaN  \n",
       "33  GCCGTAAGATGACAAATACCAGTTACTGGGGGGTAGCTTGTGATCG...      NaN  \n",
       "34  TGGCGCACATATGGGTGGAAACTGACGAGGTTCATCATCAGATGGT...      NaN  \n",
       "35  TTTGAGAGGCGCGGAGTCCTCTATATCTAACCCTTGTCGATTCGAC...      NaN  \n",
       "36  GCCCGACCGAAATAGTCACGTTTATCTGTGGAGACCTTTAGGCCCC...      NaN  \n",
       "37  CCCTCAGAGAGCGTAGAGGACGGCAAGAGAATAGATACCAACCGAC...      NaN  \n",
       "38  GCGGTGGGTTCATCAGAAACGAGCACTCTCGAGTCTTGATCACCTT...      NaN  \n",
       "39  GATGAATGAATTCCATCATTGGTGCCACGATCTGGTCGTGGCGGTG...      NaN  \n",
       "40  TGGCTTAACCACGAATGGTTGTGGGGAATTTCTTTCCTCGGGGAAT...      NaN  \n",
       "41  CGAATACGAGGTAGGGTTCCACCTACTACTTAGGTGCGCGGGTCGC...      NaN  \n",
       "42  CAGCGAATGGTATAACCCCCCCCCAGCCGTGACGTTCACGTGCTGG...      NaN  \n",
       "43  ACGCTTACCGACGAGCCCCTTCGAACGACGTAACACGTCGAATCGA...      NaN  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stain = pd.read_csv(\"../data/interim/gram_staining/staining_assignation.csv\", index_col = 0)\n",
    "stain = stain[['Accession', 'staining']]\n",
    "stain = stain.rename(columns={'Accession': 'id'})\n",
    "\n",
    "# Merge the two dataframes on the 'id' column\n",
    "df = pd.merge(df, stain, on='id', how='left')\n",
    "\n",
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "phage",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
