{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import FeatureLocation\n",
    "from Bio.Seq import UndefinedSequenceError\n",
    "import subprocess\n",
    "\n",
    "def unix_call(command):\n",
    "    job = subprocess.run(command.split()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dealing with GenBank record EF380009.1\n"
     ]
    }
   ],
   "source": [
    "gbk_filename =  \"/Users/Alvaro/Desktop/phage/sequence.gb\"\n",
    "faa_filename = \"NC_005213_converted.fna\"\n",
    "input_handle  = open(gbk_filename, \"r\")\n",
    "output_handle = open(faa_filename, \"w\")\n",
    "\n",
    "for seq_record in SeqIO.parse(input_handle, \"genbank\") :\n",
    "    print (\"Dealing with GenBank record %s\" % seq_record.id)\n",
    "    output_handle.write(\">%s %s\\n%s\\n\" % (\n",
    "           seq_record.id,\n",
    "           seq_record.description,\n",
    "           seq_record.seq))\n",
    "\n",
    "output_handle.close()\n",
    "input_handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calling one sequence as a test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_directory = \"../data/interim/pharokka/\"\n",
    "database_dir = \"../data/external/pharokka_v1.4.0_databases\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: EF380009.1\n",
      "Name: EF380009.1\n",
      "Description: EF380009.1 Enterobacteria phage phiX174 isolate AP100, complete genome\n",
      "Number of features: 8\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=30-OCT-2023\n",
      "/accessions=['EF380009']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=\n",
      "/organism=.\n",
      "/taxonomy=[]\n",
      "Seq('GAGTTTTATCGCTTCCATGACGCAGAAGTTAACACTTTCGGATATTTCTGATGA...GCA')\n",
      "ID: MN335248.1\n",
      "Name: MN335248.1\n",
      "Description: MN335248.1 prueba\n",
      "Number of features: 16\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=VRL\n",
      "/date=30-OCT-2023\n",
      "/accessions=['MN335248']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=\n",
      "/organism=.\n",
      "/taxonomy=[]\n",
      "Seq('AGTACCGCCCGAATTTCGCAGCAACCCAACCGACGCAAGCCCAACCAACAGAGA...CAC')\n",
      "Processing record: MN335248.1\n",
      "\n",
      "AGTACCGCCCGAATTTCGCAGCAACCCAACCGACGCAAGCCCAACCAACAGAGAGAACGAAGACCATGAGCAACGCACCGAAGATCACCATCAACAGCGCCGTCGAAACCCGCACTGTCACCACCTCGAAGGGCATGCCGAAGGCCATTTACAGCCAGCGCGCCACGCTCGAAACCGAAGCGATGCGTATCCAGATCGAAGTCGAATGCGATGGCCCGGACAAGGGCTATCCGGTCGGCACGGTCAAGGAATGGGATTTGGTCACCGATCTGGTGCCGGGCCGTTTCGGTGTCGAACTGGCGCGCCGCATGACGCTGGTCGATCCGCAGAGCGGCAAGCCGCAGCAGCGGCAGGCAGCGTAATCGATGGCCGTGCTCATCCCCGCCTGCCGGGAAGCCGATCTAGACGCGGCAGCGGGGACGTGCACGGCGGTCATCTGGATTCCTCAACCGTCACTCTTGCCGGAACTGGCGGTGAAGGATGCCCAGGCCATGGGCAGCGCTATCGCGTTCCTGTGGGCCACGGCGTACGTGTTCCGGCTCATCCGCAAGAAAATTCAACAGTCCTAGGAGGACATGCAATGCGCAAGCTCAAGACCCTGTTCAAGAACAAGACCGCTGCTCTGGCCGCAGTTGGTTCGGCCGCTCTCGTCTCCGCTCCGGCCTTCGCCACTGGCGGTGGTGGTGTGGATGTGGGTGACGTGGTGTCGGCCATCCAGGGCGCAGCAGGCCCGATTGCAGCCATCGGCGGTGCAGTGCTGACTGTCATGGTGGGCATCAAGGTCTACAAGTGGGTGCGCCGCGCCATGTAACGACCACCGGCGGACAGGGCCAACTCCCTCCCGCCGGTCTTTTATGGGGATAGGGCAGGGGAACGGGCGATGGAAGGGTGGATTTGGTTAGTCGCATGGCTGGTGGCCTGCGCGATTGTCTTCGTGGATTTCGAATAATGGGCTGGCTCGCACGCGTGTTTGCATCCGCGATTGCGCGTCGCCTCGCATACGTGCTTGTAGCGGCAACGCTCGCATGGTGTGGCATGGGCAAGGCGCACGCAGCGAATTGTGATGCAACCTCTGACATGTGCAGCGAGGGCGATGCCTGGAATGCGGCACGTCAGCTTGCAGATTCCCGTGGTCCTGACACCTGCAAACTGGTTGGCGGCAACAACGCCTCCTACACAGGCCCAAATATTGAGAAAGACCCGGGCAATTCTTCGAAGGACGTGGGTGCGCTCGCTGTGCGAGTTTCCTGTTCGAATGGGTCGCTGCCGTTCTTAGGCACCAAGTACTACGTGACCGCAAAATCGTGCGAACAAGAGCCTGGCTACACCGGCGGCGGTCCGTGGGGGACATATGTTGGCACCGCGCGCAACGGCAGTATCGGTTGTCGCAATGGCTGCGACGGTGTGTGGTTTGGCAATGGCGATGGCTCGATGACATGGGGCAGTACGGGCGCCGTGTGTCCGACTGATCCGGAGAAGACGTGCGATGCCATGGGCAAGGGGTACGGATGGAACGGCTACCTTGGCGTGTGTGAGCCGCCGCCGACTGAAGAGTGCCCAGAGGGTCAGGTGCCGGACGGGAAGGGTGGCTGTTCAACCAACAAATGCCCCGAAGGCATGCTATTGCAGGCTGACGGCACCTGCACACCGAAGAAGAACGATTGCCCTGCTGGGATGATCAAGTCGCCAACTGGTTCATGCCTTCCAGGCGACGGCCAGTGCGCCGCTGGTGAAGTGCGTGGTCCTGATGGAACGTGCAAGAAAGATGCAGATGGTGATGGTGAACCTGACGGGGACGGCGAGGGCGGCGAGGGCGAGGGCGATAGCAAGAGCGAATTTTCTGGCGGCGACAGCTGTGACTCTCCGCCTAGTTGCAGCGGTGACGCAATTATGTGCGGACAGGCGCGCATTCAATGGCGCATTGATTGCAACACCCGCCGCGATGTAAACATCACAGGTGGTTCGTGCGCGGCCATGCCTGTATGCGTTGGCAAGAACTGCAAGGCATTGGAGTATTCCCAGCTGTTGATGCAATGGCGTGCTGCATGTGCCTTGGAGAAGGCAGCGAACAACAGTGGCGGCGGTACTGGAAACAATGCAGACGTCAAGGCTATTCGCGATGCGATCACCGGCAATGGTTCTGCCGATATCGGTGCCGATGGCAAGCCTGCTGACGCGTTCTCTGATGAATCCGGATACGGCGAAGATGGCTACCCCACTGGCGAACTCGATACACAGGGCTTCGGCTATAGCCGCACGTGCCCAACGATTCCCGATGTAGCCGTGTTCGGCCAGACGCTGCATTTCGACACGTCCAAGTTCTGTCAGTGGATGGTGCTGGGCGGGCAGATTGTCTTGGTCATGGCGTCCTTGGTGTCCTTGCGTTTGATGAGTCAAGGAGGTAGCGCGTAATGCCCTGGTTAATCGCACAACTCGTCACCGCGCTTGCATGGCTGTTCAAGTCGCGTATCGGCTTATGGATCATGACCGCGCTTGTGTGGCTCGGTATCAATTTCGGCACGATCAAGATGGTGGTCGAGCCGGCAATCGATCTGCTCAAGGATTACGCCCAAGGCATGGGCAATGGCAACGGTCAGCTTGGCGCAGACGCCATGGCGTGGTTCGGCGTGTTGCAGTTCGATAAGGCGTTGACCATGGTCATCTCTGCCATCGCTGCCAAGCACGCCATCATGCAAGGTCGCCTGTTCCTGTTTAAGCGTGGATTCGGAGCGAAGCCGTAATGCCAATCGAGCTATACACCGGGCAACCCGGCAACGGTAAAACCGCTCTCATGATGGAGCGCCTGGTCGAAGAGTCGAAGCGCGCCGAGCGACCGATTTTCGCTGCGGGTATCGACGGCCTACAGGACGGCTTGGCGACTGTGCTCGATGACCCGCGCAAGTGGAACGATAAGGATGCAGACGGCAACTACGTTGTGCCCAATGGCTCGTTGATCTTCGTTGACGAGGCGTGGAAGTGGTATGGCCACCTGCACGACGCTACGCGCCAGCAGACGCCGAAACATGTGCTCGATCTTGCTGAGCATCGGCATCGCGGTCTGGATTTTGTGTGGACCACGCAGCAGCCGAACCAGCTGTATCCGTTCGTGCGTGGATTGATCGGTGCACATACGCATGTGGTGCGTCGGTTCAGCACGAAGATGATCGATGTGTTCCGCTGGGGCGAGTTGAACGAAGAAATCAAGTCGTCAGCCAAGCGCGATCTTGCCCAGCGCACCACGCGCCTTCTGCCGTCCTCGATCTTCGGCGCATACAAGTCGGCCGAGGTCCACACGATCAAGCCGCGTATTCCGTGGAAAGTGCTGGCGTTGCCGGGATTGGTCATCCTTGCCATCGCGCTTGGATGGCTCGCCTACACGATGCTCAAGCCTAGCGCGATGGCCGCAAAACTCGGAGATAAGGGGACGCAATCGGCGTTAGCCGATGCGGCCCCTGGCGGGTCGGCGAACTCCGCACGGCGTGATGGTCCGCGTTGGGAATCTCCCACCGAATATGCCAAGCAACACCTGCCCCGGTTCGGCACCATGCCGTGGACTGCACCGGTGTTCGATGTTGGGAATCTCCCACCGAATATGCCAAGCAACACCTGCCCCGGTTCGGCACCATGCCGTGGACTGCACCGGTGTTCGATGACCGCAGCATCACCGCCGATCCGATGCTGATCTGCATGTCGTCGCTCGCGGGCACGGACGCGCAGGGCAAGCACAAGGAAGCGTCTTGCACGTGCATGACAGAGCAGGGCACCGCGTACGACCTCGACCAGCCGCAGTGCCGCACCATTGCCAAGCGCGGTCCGGTGTACAACCCGTATCGCGAACGGCGTGAGAACGAGCAGCAGCCCGCGCAGCAGCAACAGGCGGTGCAGGGTGGGGCGGCTGCGCCTGGTCTGTCCGGCGTGGTCGTGCAGCGTGGCATGCGCACACAGGGCACCTTCCCGGAATCGAAGGGCTACAACACCAAGACCACCACGCCATCAACGTCGCTGGAGATGTGACATGACCAGCAGCGGCCGCGAGGCATTGAAGTGGATTGCGCTGGTGTTGATGACCGGCGATCACGTGGTAAAGGTGTTTTTCGGCGGCTATGTGCCGGTGCTATCCGAATTGGGGCGGATCGCGTTCCCGGTGTTCGCGTTGGTCATGGCGTACAACCTTGCCCAGCCACGGGCCGATTATGCGAAATCGGTGCTGCGCCTTGCGGGCTGGGGGCTGCTGGCGCAGCCATTTCACGCCTGGGCGTTCGGGTACTGGCTGCCGCTCAACGTCCTGCTGACTTTCGCGTTGGCGGCGCTGCTCGTCTGGACGGTCCACGCCCGGCATTGGCTCTATGCTGCTGTGTGCTGTCTCGTCCTGCCGCTCGTGGTCGATTACCAATGGTCCGGCGTCTGGCTCGTTCTGGCGGCGTGGTGCTGGTTCCGCACGGGCCGGTTGGAAGCCTTCGGTGGCGTCCTGGCCACCATGGCTGCGTTGTGCTGGTACAACGGCAATCTGTGGGCGCTGGGATCGCTCCCGGTGCTGGCCCTCGGCTACGTCTGGTGGCCGCTGCCGCGCCTGCGCTGGGCGTTCTACGGCTATTACGTTGGCCACCTCGGGCTGCTCGTGCTCATCGCGTCGTTGCCGGCGTTTCAGCAGCACGTCTCGTAGGTAGATGACGTTGTTCCTTGCAGAAATTCGGCTCCCTGGAACCCTTGTGGCAGTAGGCGTTTCAGGATTTGCCGATCGAGCATTGCGCTCGGTGGCCATCATCAATGACCACTCGCGAGCGATGCAGCAGGTTAGCGACCAGAACCGCATGTCGCACGGCTCGATGGTCTTGCCCTCAGGCGTAAAGAAGTGGTGGCCCTGGAAACCAAAACCGGCCCAAGGGCCGGCAAGGTCTACGCGGTCGTAGGTGTCGATCACGCTGCGATTTCCGTATCAGAAGGACCATGGGACGGTAGCAGCGAGGTAAGCCAGAGCCATAGCCAGCGCGCCCGCGATTGCAGCGCCTGCCGCACACGATTTCGCATAATGTATAGAGAGTTAGCGTTATGTTCGGTCGAAAAGTGACTTTGATCACTGATTTTCGATGCCTCAGCCCGGCCCGGCAGGGACACGCCAAGTAGCAGCAAGACGCCCAGCGCGCTTGCAGCTATGCGTTGAAGGATTCCTTTGTGTTTGATGTTCTCCGGTGAGTGCATGCGCACTTCCGATGCTGTCTTGATGATTGGCAATTTCGCTAGCGCTGCAATTTCAGCTGCTTGTTCGAAGTTCGGTACAGAAGTGCCGTCTAACCACCTGATCACCGCGACGTGCGACACCTTGGCTATGGTGGCGAACTCTCGAATGCTCTTTGCATTCGTTTCTTTCATCGCCAGCTCAACCAACTCTTTTGCGTCCATCGTCATTTCCAGTGGAACGGGAGGTTGACAGCGTATCTGAACTGCTGGTAACGTCCCTACTAGAAGCTACCACTGCCCACCCGCCGGCCCCGCCCGGTGCCGGTTGGGCGGGTTCTACCGGGCACCGGGCAGGGGATACACCTTGAAAGCACTTATCCAGTTTTGCCGTGAATTCGGCCTGCTTGTCTGCATGGCATGCGCTACCGGGTTTGGCATTGGCGGACTCGTTGGCTTTATCGCTGGTCAGGGCGCTGCAAGTAGCGTTGAACACCAGATTCAGCGCGATGAGGCGCTCCAGCAGCTTTGCATCGACGGTAACGAAGGCGCCTGCCGCGTTATGGAGTTCCGCAGATGAGCCGCCCAACCGACACCGAACGCGGCGCACGCATTGCGCTCGACTACGTTGAATCAAAGCTTATTCAACGCGATTTATTTCCAACCCGCCGCGCGCCGTCTTTAAAGTTCTGGCGCGAAATAAAGGCGATTGCGACGGAACACCTTGCCGAATGCAAGGCACTCCGCGAGGCCCGCGCATGAACACGGGCGACCTGAAATTCTGGCAGCTGATCGAGCTGAAAAAGAAAGCCGAAACCGACGACGAACGCGCCGAGCTAGACCGACTGATAGAGGAACGCATCAGGTGGGTAGGCTCGGGGCAGGGCGATGAGCCATGAAGACGCAGCCGTCCAGGCTAGCCCATTACCCGAACAGCCCCTGCTATCAGTGCGGGGGAGCGCAGTTTCAGACGTTGAACGCGTGGGATGCGAAGTTGACGGTCTGCACCGATTGCGGCGTGCTGATCTCGAAGCGCAGGGATATGCAGAGTTCTTACAGCGGATGCAGTGGCAACAGTTCTGGACACTCACATTTCGCATCGAAGAAGCCGGCCGAACCGGTGGTGTCCACCCAGAAAAGGCTGATAAAGCGTTCCGATTCTTCGCCAGCTGTATCAACCGGCAAATATACGGCCGTGAGTGGTCAAAGCGTCCGCATGGCGGCATCCAGTGGGCACGGGGGCAAGAGTTCCACAAAGACGGTCGATTGCATTTCCACGCCGTCGCAGCTGCACCTACCGATGATTTAAACCGGCTTATGAGCCGCTATGAGTGGCATGAGTTTTGGTTCAAGGAATTCGGACGTAATCGCATAGAGGCACCACGCAGCCAGCTCGATATAACCGGCTACGTGTCGAAGTACGTAACGAAGGGGGGAGTGGTGGACGTGTCGAAGAACTTCGGCGCCTGGACGCCACCACCGATCGACTACACGCGCCGCCCAGTGCAGGCCGAGTTCGATCAAACAACGCGTAAGGGGAGCATCGAATCACTGGACCGGGGTGTAGGGGCAGCGCCCCTACGGACTGGTCAACGAAACCGCCGGTGATCGCTCACCGGGACATGCAACACCGCCCCCGGTCTGGGAGGCACCGAAGCCGCAGCCTGTTCGCCTACGCGCACCAGCACGGCAACGCCAGGAATCCACCCCTGAAGACCCGCCTTCGCTCGCAGGCGACCTAAAGCGACCGGTAGGCAGACACCGTGACAGTCAGCCCCGAGCGGCAGGTGGCAAGCCCCGCGCACGGTCATCGCTAAGCCTCACCGCACCCCCCGCATGGGGGGTAAGGGGGGCCTTAGCTTGACCCCAC\n"
     ]
    }
   ],
   "source": [
    "genbank_file_path2 =  \"../data/raw/inphared_8Sep2023/8Sep2023_phages_downloaded_from_genbank.gb\"\n",
    "genbank_file_path2 =  \"../data/interim/genbank_engineering/sequence.gb\"\n",
    "genbank_file_path2 = \"/Users/Alvaro/Desktop/prueba_pharokka2/pharokka.gbk\"\n",
    "# Read the GenBank file\n",
    "for record in SeqIO.parse(genbank_file_path2, \"genbank\"):\n",
    "\n",
    "    \n",
    "    if record.id == \"MN335248.1\":\n",
    "        print(f\"Processing record: {record.id}\\n\")\n",
    "        \n",
    "        sequence = str(record.seq)\n",
    "        print(sequence)\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-10-30 10:08:57.331 | INFO     | input_commands:instantiate_dirs:160 - Removing output directory /Users/Alvaro/Desktop/prueba_pharokka as -f or --force was specified.\n",
      "2023-10-30 10:08:57.339 | INFO     | __main__:main:84 - Starting Pharokka v1.4.1\n",
      "2023-10-30 10:08:57.340 | INFO     | __main__:main:85 - Command executed: Namespace(citation=False, coding_table='11', custom_hmm='', database='/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/', dnaapler=False, evalue='1E-05', fast=False, force=True, genbank=False, gene_predictor='phanotate', infile='../data/interim/genbank_engineering/sequence.fna', locustag='Default', meta=False, meta_hmm=False, mmseqs2_only=False, outdir='/Users/Alvaro/Desktop/prueba_pharokka', prefix='Default', split=False, terminase=False, terminase_start='nothing', terminase_strand='nothing', threads='2')\n",
      "2023-10-30 10:08:57.340 | INFO     | __main__:main:86 - Repository homepage is https://github.com/gbouras13/pharokka\n",
      "2023-10-30 10:08:57.340 | INFO     | __main__:main:87 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
      "2023-10-30 10:08:57.340 | INFO     | __main__:main:89 - Checking database installation in /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/.\n",
      "2023-10-30 10:08:57.341 | INFO     | __main__:main:92 - All databases have been successfully checked.\n",
      "2023-10-30 10:08:57.341 | INFO     | __main__:main:108 - Checking dependencies.\n",
      "2023-10-30 10:08:57.861 | INFO     | input_commands:check_dependencies:322 - Phanotate version found is v1.5.1\n",
      "2023-10-30 10:08:57.863 | INFO     | input_commands:check_dependencies:331 - Phanotate version is ok.\n",
      "2023-10-30 10:08:57.906 | INFO     | input_commands:check_dependencies:353 - MMseqs2 version found is v13.45111\n",
      "2023-10-30 10:08:57.909 | INFO     | input_commands:check_dependencies:362 - MMseqs2 version is ok.\n",
      "2023-10-30 10:08:58.245 | INFO     | input_commands:check_dependencies:386 - tRNAscan-SE version found is v2.0.12\n",
      "2023-10-30 10:08:58.247 | INFO     | input_commands:check_dependencies:397 - tRNAscan-SE version is ok.\n",
      "2023-10-30 10:08:58.547 | INFO     | input_commands:check_dependencies:421 - MinCED version found is v0.4.2\n",
      "2023-10-30 10:08:58.550 | INFO     | input_commands:check_dependencies:432 - MinCED version is ok.\n",
      "2023-10-30 10:08:58.564 | INFO     | input_commands:check_dependencies:458 - ARAGORN version found is v1.2.41\n",
      "2023-10-30 10:08:58.566 | INFO     | input_commands:check_dependencies:469 - ARAGORN version is ok.\n",
      "2023-10-30 10:08:58.589 | INFO     | input_commands:check_dependencies:485 - mash version found is v2.3\n",
      "2023-10-30 10:08:58.591 | INFO     | input_commands:check_dependencies:492 - mash version is ok.\n",
      "2023-10-30 10:09:00.943 | INFO     | input_commands:check_dependencies:511 - Dnaapler version found is v0.3.2\n",
      "2023-10-30 10:09:00.944 | INFO     | input_commands:check_dependencies:518 - Dnaapler version is ok.\n",
      "2023-10-30 10:09:00.945 | INFO     | input_commands:check_dependencies:529 - Pyrodigal version is v3.0.1\n",
      "2023-10-30 10:09:00.945 | INFO     | input_commands:check_dependencies:530 - Pyrodigal version is ok.\n",
      "2023-10-30 10:09:00.971 | INFO     | input_commands:validate_fasta:205 - Checking Input FASTA.\n",
      "2023-10-30 10:09:00.972 | INFO     | input_commands:validate_fasta:207 - FASTA checked.\n",
      "2023-10-30 10:09:00.972 | INFO     | input_commands:validate_gene_predictor:214 - Phanotate will be used for gene prediction.\n",
      "2023-10-30 10:09:00.973 | INFO     | __main__:main:268 - Running Phanotate.\n",
      "2023-10-30 10:09:00.976 | INFO     | external_tools:run:50 - Started running phanotate.py -f fasta -o /Users/Alvaro/Desktop/prueba_pharokka/phanotate_out_tmp.fasta ../data/interim/genbank_engineering/sequence.fna ...\n",
      "2023-10-30 10:09:04.628 | INFO     | external_tools:run:52 - Done running phanotate.py -f fasta -o /Users/Alvaro/Desktop/prueba_pharokka/phanotate_out_tmp.fasta ../data/interim/genbank_engineering/sequence.fna\n",
      "2023-10-30 10:09:04.630 | INFO     | external_tools:run:50 - Started running phanotate.py -f tabular -o /Users/Alvaro/Desktop/prueba_pharokka/phanotate_out.txt ../data/interim/genbank_engineering/sequence.fna ...\n",
      "2023-10-30 10:09:08.803 | INFO     | external_tools:run:52 - Done running phanotate.py -f tabular -o /Users/Alvaro/Desktop/prueba_pharokka/phanotate_out.txt ../data/interim/genbank_engineering/sequence.fna\n",
      "2023-10-30 10:09:08.823 | INFO     | __main__:main:291 - Starting tRNA-scanSE.\n",
      "2023-10-30 10:09:08.824 | INFO     | external_tools:run:50 - Started running tRNAscan-SE --thread 2 -G -Q -j /Users/Alvaro/Desktop/prueba_pharokka/trnascan_out.gff ../data/interim/genbank_engineering/sequence.fna ...\n",
      "2023-10-30 10:09:09.169 | INFO     | external_tools:run:52 - Done running tRNAscan-SE --thread 2 -G -Q -j /Users/Alvaro/Desktop/prueba_pharokka/trnascan_out.gff ../data/interim/genbank_engineering/sequence.fna\n",
      "2023-10-30 10:09:09.180 | INFO     | processes:run_minced:694 - Running MinCED.\n",
      "2023-10-30 10:09:09.186 | INFO     | external_tools:run:50 - Started running minced ../data/interim/genbank_engineering/sequence.fna /Users/Alvaro/Desktop/prueba_pharokka/pharokka_minced_spacers.txt /Users/Alvaro/Desktop/prueba_pharokka/pharokka_minced.gff ...\n",
      "2023-10-30 10:09:09.570 | INFO     | external_tools:run:52 - Done running minced ../data/interim/genbank_engineering/sequence.fna /Users/Alvaro/Desktop/prueba_pharokka/pharokka_minced_spacers.txt /Users/Alvaro/Desktop/prueba_pharokka/pharokka_minced.gff\n",
      "2023-10-30 10:09:09.573 | INFO     | processes:run_aragorn:723 - Running Aragorn.\n",
      "2023-10-30 10:09:09.574 | INFO     | external_tools:run:50 - Started running aragorn -l -gcbact -w -m -o /Users/Alvaro/Desktop/prueba_pharokka/pharokka_aragorn.txt ../data/interim/genbank_engineering/sequence.fna ...\n",
      "2023-10-30 10:09:09.596 | INFO     | external_tools:run:52 - Done running aragorn -l -gcbact -w -m -o /Users/Alvaro/Desktop/prueba_pharokka/pharokka_aragorn.txt ../data/interim/genbank_engineering/sequence.fna\n",
      "2023-10-30 10:09:09.598 | INFO     | __main__:main:299 - Starting MMseqs2.\n",
      "2023-10-30 10:09:09.599 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on PHROG Database.\n",
      "2023-10-30 10:09:09.604 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs ...\n",
      "2023-10-30 10:09:10.083 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs\n",
      "2023-10-30 10:09:10.090 | INFO     | external_tools:run:50 - Started running mmseqs search -e 1E-05 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/tmp_dir/ -s 8.5 --threads 2 ...\n",
      "2023-10-30 10:24:08.907 | INFO     | external_tools:run:52 - Done running mmseqs search -e 1E-05 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/tmp_dir/ -s 8.5 --threads 2\n",
      "2023-10-30 10:24:08.911 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs_results.tsv --full-header --threads 2 ...\n",
      "2023-10-30 10:24:08.979 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/prueba_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/mmseqs_results.tsv --full-header --threads 2\n",
      "2023-10-30 10:24:08.982 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on CARD Database.\n",
      "2023-10-30 10:24:08.984 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs ...\n",
      "2023-10-30 10:24:09.094 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs\n",
      "2023-10-30 10:24:09.099 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/CARD_dir/ -s 8.5 --threads 2 ...\n",
      "2023-10-30 10:26:28.181 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/CARD_dir/ -s 8.5 --threads 2\n",
      "2023-10-30 10:26:28.184 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/CARD_results.tsv --full-header --threads 2 ...\n",
      "2023-10-30 10:26:28.212 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/prueba_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/CARD_results.tsv --full-header --threads 2\n",
      "2023-10-30 10:26:28.216 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on VFDB Database.\n",
      "2023-10-30 10:26:28.218 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs ...\n",
      "2023-10-30 10:26:28.531 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/prueba_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs\n",
      "2023-10-30 10:26:28.536 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/vfdb /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB_dir/ -s 8.5 --threads 2 ...\n",
      "2023-10-30 10:30:22.885 | INFO     | external_tools:run:52 - Done running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/vfdb /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB_dir/ -s 8.5 --threads 2\n",
      "2023-10-30 10:30:22.890 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/vfdb /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/vfdb_results.tsv --full-header --threads 2 ...\n",
      "2023-10-30 10:30:22.923 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/vfdb /Users/Alvaro/Desktop/prueba_pharokka/VFDB_target_dir/target_seqs /Users/Alvaro/Desktop/prueba_pharokka/VFDB/results_mmseqs /Users/Alvaro/Desktop/prueba_pharokka/vfdb_results.tsv --full-header --threads 2\n",
      "2023-10-30 10:30:22.930 | INFO     | __main__:main:330 - Running PyHMMER on PHROGs.\n",
      "2023-10-30 10:30:43.325 | INFO     | __main__:main:346 - Post Processing Output.\n",
      "2023-10-30 10:30:43.366 | INFO     | post_processing:create_mmseqs_tophits:1920 - Processing mmseqs2 outputs.\n",
      "2023-10-30 10:30:43.366 | INFO     | post_processing:create_mmseqs_tophits:1921 - Processing PHROGs output.\n",
      "2023-10-30 10:30:43.578 | INFO     | post_processing:process_vfdb_results:2134 - Processing VFDB output.\n",
      "2023-10-30 10:30:43.595 | INFO     | post_processing:process_vfdb_results:2224 - 0 VFDB virulence factors identified.\n",
      "2023-10-30 10:30:43.597 | INFO     | post_processing:process_card_results:2241 - Processing CARD output.\n",
      "2023-10-30 10:30:43.609 | INFO     | post_processing:process_card_results:2367 - 0 CARD AMR genes identified.\n",
      "2023-10-30 10:30:43.993 | INFO     | __main__:main:404 - Converting gff to genbank.\n",
      "2023-10-30 10:30:44.060 | INFO     | __main__:main:416 - Finding the closest match for each contig in INPHARED using mash.\n",
      "2023-10-30 10:30:44.061 | INFO     | external_tools:run:50 - Started running mash sketch ../data/interim/genbank_engineering/sequence.fna -o /Users/Alvaro/Desktop/prueba_pharokka/input_mash_sketch.msh -i ...\n",
      "2023-10-30 10:30:44.097 | INFO     | external_tools:run:52 - Done running mash sketch ../data/interim/genbank_engineering/sequence.fna -o /Users/Alvaro/Desktop/prueba_pharokka/input_mash_sketch.msh -i\n",
      "2023-10-30 10:30:44.100 | INFO     | external_tools:run_to_stdout:59 - Started running mash dist /Users/Alvaro/Desktop/prueba_pharokka/input_mash_sketch.msh /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/1Aug2023_genomes.fa.msh -d 0.2 -i ...\n",
      "2023-10-30 10:30:45.617 | INFO     | external_tools:run_to_stdout:61 - Done running mash dist /Users/Alvaro/Desktop/prueba_pharokka/input_mash_sketch.msh /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/1Aug2023_genomes.fa.msh -d 0.2 -i\n",
      "2023-10-30 10:30:45.974 | INFO     | __main__:main:431 - Pharokka has finished.\n",
      "2023-10-30 10:30:45.974 | INFO     | __main__:main:432 - Elapsed time: 1308.64 seconds\n",
      "2023-10-30 10:30:45.975 | INFO     | __main__:main:434 - If you use Pharokka in your research, please cite:\n",
      "2023-10-30 10:30:45.975 | INFO     | __main__:main:435 - George Bouras, Roshan Nepal, Ghais Houtak, Alkis James Psaltis, Peter-John Wormald, Sarah Vreugde.\n",
      "2023-10-30 10:30:45.975 | INFO     | __main__:main:438 - Pharokka: a fast scalable bacteriophage annotation tool.\n",
      "2023-10-30 10:30:45.976 | INFO     | __main__:main:439 - Bioinformatics, Volume 39, Issue 1, January 2023, btac776.\n",
      "2023-10-30 10:30:45.976 | INFO     | __main__:main:440 - https://doi.org/10.1093/bioinformatics/btac776.\n",
      "2023-10-30 10:30:45.979 | INFO     | __main__:main:441 - You should also cite the full list of tools Pharokka uses, which can be found at https://github.com/gbouras13/pharokka#citation.\n"
     ]
    }
   ],
   "source": [
    "# database_dir = \"/Users/Alvaro/Downloads/pharokka_v1.4.0_databases\"\n",
    "database_dir = \"/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/\"\n",
    "sequence = \"../data/interim/genbank_engineering/double_sequence.fna\"\n",
    "sequence = \"../data/interim/genbank_engineering/sequence.fna\"\n",
    "output_dir = \"/Users/Alvaro/Desktop/prueba_pharokka\"\n",
    "unix_call(f'pharokka.py -i {sequence} -o {output_dir} -f -t 2 -d {database_dir}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-10-30 11:12:31.387 | INFO     | input_commands:instantiate_dirs:160 - Removing output directory /Users/Alvaro/Desktop/full_pharokka as -f or --force was specified.\n",
      "2023-10-30 11:12:31.396 | INFO     | __main__:main:84 - Starting Pharokka v1.4.1\n",
      "2023-10-30 11:12:31.397 | INFO     | __main__:main:85 - Command executed: Namespace(citation=False, coding_table='11', custom_hmm='', database='/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/', dnaapler=False, evalue='1E-05', fast=False, force=True, genbank=False, gene_predictor='phanotate', infile='../data/interim/genbank_engineering/double_sequence.fna', locustag='Default', meta=False, meta_hmm=False, mmseqs2_only=False, outdir='/Users/Alvaro/Desktop/full_pharokka', prefix='Default', split=False, terminase=False, terminase_start='nothing', terminase_strand='nothing', threads='10')\n",
      "2023-10-30 11:12:31.397 | INFO     | __main__:main:86 - Repository homepage is https://github.com/gbouras13/pharokka\n",
      "2023-10-30 11:12:31.398 | INFO     | __main__:main:87 - Written by George Bouras: george.bouras@adelaide.edu.au\n",
      "2023-10-30 11:12:31.399 | INFO     | __main__:main:89 - Checking database installation in /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/.\n",
      "2023-10-30 11:12:31.401 | INFO     | __main__:main:92 - All databases have been successfully checked.\n",
      "2023-10-30 11:12:31.402 | INFO     | __main__:main:108 - Checking dependencies.\n",
      "2023-10-30 11:12:32.424 | INFO     | input_commands:check_dependencies:322 - Phanotate version found is v1.5.1\n",
      "2023-10-30 11:12:32.427 | INFO     | input_commands:check_dependencies:331 - Phanotate version is ok.\n",
      "2023-10-30 11:12:32.505 | INFO     | input_commands:check_dependencies:353 - MMseqs2 version found is v13.45111\n",
      "2023-10-30 11:12:32.508 | INFO     | input_commands:check_dependencies:362 - MMseqs2 version is ok.\n",
      "2023-10-30 11:12:32.957 | INFO     | input_commands:check_dependencies:386 - tRNAscan-SE version found is v2.0.12\n",
      "2023-10-30 11:12:32.959 | INFO     | input_commands:check_dependencies:397 - tRNAscan-SE version is ok.\n",
      "2023-10-30 11:12:33.385 | INFO     | input_commands:check_dependencies:421 - MinCED version found is v0.4.2\n",
      "2023-10-30 11:12:33.387 | INFO     | input_commands:check_dependencies:432 - MinCED version is ok.\n",
      "2023-10-30 11:12:33.403 | INFO     | input_commands:check_dependencies:458 - ARAGORN version found is v1.2.41\n",
      "2023-10-30 11:12:33.405 | INFO     | input_commands:check_dependencies:469 - ARAGORN version is ok.\n",
      "2023-10-30 11:12:33.437 | INFO     | input_commands:check_dependencies:485 - mash version found is v2.3\n",
      "2023-10-30 11:12:33.439 | INFO     | input_commands:check_dependencies:492 - mash version is ok.\n",
      "2023-10-30 11:12:39.011 | INFO     | input_commands:check_dependencies:511 - Dnaapler version found is v0.3.2\n",
      "2023-10-30 11:12:39.013 | INFO     | input_commands:check_dependencies:518 - Dnaapler version is ok.\n",
      "2023-10-30 11:12:39.013 | INFO     | input_commands:check_dependencies:529 - Pyrodigal version is v3.0.1\n",
      "2023-10-30 11:12:39.013 | INFO     | input_commands:check_dependencies:530 - Pyrodigal version is ok.\n",
      "2023-10-30 11:12:39.110 | INFO     | input_commands:validate_fasta:205 - Checking Input FASTA.\n",
      "2023-10-30 11:12:39.111 | INFO     | input_commands:validate_fasta:207 - FASTA checked.\n",
      "2023-10-30 11:12:39.112 | INFO     | input_commands:validate_gene_predictor:214 - Phanotate will be used for gene prediction.\n",
      "2023-10-30 11:12:39.115 | INFO     | input_commands:validate_meta:245 - More than one contig detected in the input file. Re-running pharokka with -m meta mode is recommended unless this is a fragmented isolate genome. Continuing.\n",
      "2023-10-30 11:12:39.116 | INFO     | __main__:main:268 - Running Phanotate.\n",
      "2023-10-30 11:12:39.119 | INFO     | external_tools:run:50 - Started running phanotate.py -f fasta -o /Users/Alvaro/Desktop/full_pharokka/phanotate_out_tmp.fasta ../data/interim/genbank_engineering/double_sequence.fna ...\n",
      "2023-10-30 11:13:00.901 | INFO     | external_tools:run:52 - Done running phanotate.py -f fasta -o /Users/Alvaro/Desktop/full_pharokka/phanotate_out_tmp.fasta ../data/interim/genbank_engineering/double_sequence.fna\n",
      "2023-10-30 11:13:00.904 | INFO     | external_tools:run:50 - Started running phanotate.py -f tabular -o /Users/Alvaro/Desktop/full_pharokka/phanotate_out.txt ../data/interim/genbank_engineering/double_sequence.fna ...\n",
      "2023-10-30 11:13:20.402 | INFO     | external_tools:run:52 - Done running phanotate.py -f tabular -o /Users/Alvaro/Desktop/full_pharokka/phanotate_out.txt ../data/interim/genbank_engineering/double_sequence.fna\n",
      "2023-10-30 11:13:20.440 | INFO     | __main__:main:291 - Starting tRNA-scanSE.\n",
      "2023-10-30 11:13:20.442 | INFO     | external_tools:run:50 - Started running tRNAscan-SE --thread 10 -G -Q -j /Users/Alvaro/Desktop/full_pharokka/trnascan_out.gff ../data/interim/genbank_engineering/double_sequence.fna ...\n",
      "2023-10-30 11:13:21.287 | INFO     | external_tools:run:52 - Done running tRNAscan-SE --thread 10 -G -Q -j /Users/Alvaro/Desktop/full_pharokka/trnascan_out.gff ../data/interim/genbank_engineering/double_sequence.fna\n",
      "2023-10-30 11:13:21.290 | INFO     | processes:run_minced:694 - Running MinCED.\n",
      "2023-10-30 11:13:21.292 | INFO     | external_tools:run:50 - Started running minced ../data/interim/genbank_engineering/double_sequence.fna /Users/Alvaro/Desktop/full_pharokka/pharokka_minced_spacers.txt /Users/Alvaro/Desktop/full_pharokka/pharokka_minced.gff ...\n",
      "2023-10-30 11:13:21.608 | INFO     | external_tools:run:52 - Done running minced ../data/interim/genbank_engineering/double_sequence.fna /Users/Alvaro/Desktop/full_pharokka/pharokka_minced_spacers.txt /Users/Alvaro/Desktop/full_pharokka/pharokka_minced.gff\n",
      "2023-10-30 11:13:21.611 | INFO     | processes:run_aragorn:723 - Running Aragorn.\n",
      "2023-10-30 11:13:21.614 | INFO     | external_tools:run:50 - Started running aragorn -l -gcbact -w -m -o /Users/Alvaro/Desktop/full_pharokka/pharokka_aragorn.txt ../data/interim/genbank_engineering/double_sequence.fna ...\n",
      "2023-10-30 11:13:21.638 | INFO     | external_tools:run:52 - Done running aragorn -l -gcbact -w -m -o /Users/Alvaro/Desktop/full_pharokka/pharokka_aragorn.txt ../data/interim/genbank_engineering/double_sequence.fna\n",
      "2023-10-30 11:13:21.641 | INFO     | __main__:main:299 - Starting MMseqs2.\n",
      "2023-10-30 11:13:21.642 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on PHROG Database.\n",
      "2023-10-30 11:13:21.644 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/full_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs ...\n",
      "2023-10-30 11:13:21.923 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/full_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs\n",
      "2023-10-30 11:13:21.929 | INFO     | external_tools:run:50 - Started running mmseqs search -e 1E-05 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/tmp_dir/ -s 8.5 --threads 10 ...\n",
      "2023-10-30 11:32:59.444 | INFO     | external_tools:run:52 - Done running mmseqs search -e 1E-05 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/tmp_dir/ -s 8.5 --threads 10\n",
      "2023-10-30 11:32:59.450 | INFO     | external_tools:run:50 - Started running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/mmseqs_results.tsv --full-header --threads 10 ...\n",
      "2023-10-30 11:33:00.118 | INFO     | external_tools:run:52 - Done running mmseqs createtsv /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/phrogs_profile_db /Users/Alvaro/Desktop/full_pharokka/target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/mmseqs/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/mmseqs_results.tsv --full-header --threads 10\n",
      "2023-10-30 11:33:00.128 | INFO     | processes:run_mmseqs:549 - Running MMseqs2 on CARD Database.\n",
      "2023-10-30 11:33:00.130 | INFO     | external_tools:run:50 - Started running mmseqs createdb /Users/Alvaro/Desktop/full_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/full_pharokka/CARD_target_dir/target_seqs ...\n",
      "2023-10-30 11:33:00.365 | INFO     | external_tools:run:52 - Done running mmseqs createdb /Users/Alvaro/Desktop/full_pharokka/phanotate_aas_tmp.fasta /Users/Alvaro/Desktop/full_pharokka/CARD_target_dir/target_seqs\n",
      "2023-10-30 11:33:00.371 | INFO     | external_tools:run:50 - Started running mmseqs search --min-seq-id 0.8 -c 0.4 /Users/Alvaro/miniconda3/envs/pharokkaENV/databases/CARD /Users/Alvaro/Desktop/full_pharokka/CARD_target_dir/target_seqs /Users/Alvaro/Desktop/full_pharokka/CARD/results_mmseqs /Users/Alvaro/Desktop/full_pharokka/CARD_dir/ -s 8.5 --threads 10 ...\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/Alvaro/Library/Mobile Documents/com~apple~CloudDocs/DTU/Autumn 2023/Phage project/phage/notebooks/05_feature_engineering.ipynb Cell 8\u001b[0m line \u001b[0;36m5\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m sequence \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m../data/interim/genbank_engineering/double_sequence.fna\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m output_dir \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m/Users/Alvaro/Desktop/full_pharokka\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m unix_call(\u001b[39mf\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mpharokka.py -i \u001b[39;49m\u001b[39m{\u001b[39;49;00msequence\u001b[39m}\u001b[39;49;00m\u001b[39m -o \u001b[39;49m\u001b[39m{\u001b[39;49;00moutput_dir\u001b[39m}\u001b[39;49;00m\u001b[39m -f -t 10 -d \u001b[39;49m\u001b[39m{\u001b[39;49;00mdatabase_dir\u001b[39m}\u001b[39;49;00m\u001b[39m'\u001b[39;49m)\n",
      "\u001b[1;32m/Users/Alvaro/Library/Mobile Documents/com~apple~CloudDocs/DTU/Autumn 2023/Phage project/phage/notebooks/05_feature_engineering.ipynb Cell 8\u001b[0m line \u001b[0;36m9\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39munix_call\u001b[39m(command):\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X12sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m     job \u001b[39m=\u001b[39m subprocess\u001b[39m.\u001b[39;49mrun(command\u001b[39m.\u001b[39;49msplit())\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:495\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    493\u001b[0m \u001b[39mwith\u001b[39;00m Popen(\u001b[39m*\u001b[39mpopenargs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs) \u001b[39mas\u001b[39;00m process:\n\u001b[1;32m    494\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 495\u001b[0m         stdout, stderr \u001b[39m=\u001b[39m process\u001b[39m.\u001b[39;49mcommunicate(\u001b[39minput\u001b[39;49m, timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[1;32m    496\u001b[0m     \u001b[39mexcept\u001b[39;00m TimeoutExpired \u001b[39mas\u001b[39;00m exc:\n\u001b[1;32m    497\u001b[0m         process\u001b[39m.\u001b[39mkill()\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:1020\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1018\u001b[0m         stderr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mread()\n\u001b[1;32m   1019\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mclose()\n\u001b[0;32m-> 1020\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mwait()\n\u001b[1;32m   1021\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   1022\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:1083\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1081\u001b[0m     endtime \u001b[39m=\u001b[39m _time() \u001b[39m+\u001b[39m timeout\n\u001b[1;32m   1082\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1083\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_wait(timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[1;32m   1084\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1085\u001b[0m     \u001b[39m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1086\u001b[0m     \u001b[39m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1087\u001b[0m     \u001b[39m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1088\u001b[0m     \u001b[39m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1089\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:1822\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1820\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mreturncode \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m   1821\u001b[0m     \u001b[39mbreak\u001b[39;00m  \u001b[39m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 1822\u001b[0m (pid, sts) \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_try_wait(\u001b[39m0\u001b[39;49m)\n\u001b[1;32m   1823\u001b[0m \u001b[39m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   1824\u001b[0m \u001b[39m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   1825\u001b[0m \u001b[39m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   1826\u001b[0m \u001b[39mif\u001b[39;00m pid \u001b[39m==\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpid:\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/subprocess.py:1780\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1778\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   1779\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1780\u001b[0m     (pid, sts) \u001b[39m=\u001b[39m os\u001b[39m.\u001b[39;49mwaitpid(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mpid, wait_flags)\n\u001b[1;32m   1781\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mChildProcessError\u001b[39;00m:\n\u001b[1;32m   1782\u001b[0m     \u001b[39m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   1783\u001b[0m     \u001b[39m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   1784\u001b[0m     \u001b[39m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   1785\u001b[0m     pid \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpid\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# database_dir = \"/Users/Alvaro/Downloads/pharokka_v1.4.0_databases\"\n",
    "database_dir = \"/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/\"\n",
    "sequence = \"../data/interim/genbank_engineering/double_sequence.fna\"\n",
    "output_dir = \"/Users/Alvaro/Desktop/full_pharokka\"\n",
    "unix_call(f'pharokka.py -i {sequence} -o {output_dir} -f -t 10 -d {database_dir}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# After pharokka: processing genbank files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Info: Entry with id 'MT366568' has molecule type 'cRNA'. Changing it to 'RNA'.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>jumbophage</th>\n",
       "      <th>topology</th>\n",
       "      <th>gc_%</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>cds_number</th>\n",
       "      <th>coding_capacity</th>\n",
       "      <th>positive_strand_%</th>\n",
       "      <th>negative_strand_%</th>\n",
       "      <th>molecule_type_ss-DNA</th>\n",
       "      <th>molecule_type_DNA</th>\n",
       "      <th>molecule_type_RNA</th>\n",
       "      <th>molecule_type_ss-RNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MN335248</td>\n",
       "      <td>7045</td>\n",
       "      <td>0</td>\n",
       "      <td>linear</td>\n",
       "      <td>60.298</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>104.229950</td>\n",
       "      <td>85.71</td>\n",
       "      <td>14.29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MK250029</td>\n",
       "      <td>540217</td>\n",
       "      <td>1</td>\n",
       "      <td>circular</td>\n",
       "      <td>25.796</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MK250028</td>\n",
       "      <td>550053</td>\n",
       "      <td>1</td>\n",
       "      <td>circular</td>\n",
       "      <td>26.012</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MK250027</td>\n",
       "      <td>551627</td>\n",
       "      <td>1</td>\n",
       "      <td>circular</td>\n",
       "      <td>26.022</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MK250026</td>\n",
       "      <td>550702</td>\n",
       "      <td>1</td>\n",
       "      <td>circular</td>\n",
       "      <td>26.020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11149</th>\n",
       "      <td>EF380009</td>\n",
       "      <td>5386</td>\n",
       "      <td>0</td>\n",
       "      <td>circular</td>\n",
       "      <td>44.653</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>175.362050</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11150</th>\n",
       "      <td>DQ490056</td>\n",
       "      <td>26537</td>\n",
       "      <td>0</td>\n",
       "      <td>linear</td>\n",
       "      <td>37.099</td>\n",
       "      <td>0</td>\n",
       "      <td>47</td>\n",
       "      <td>91.547650</td>\n",
       "      <td>72.34</td>\n",
       "      <td>27.66</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11151</th>\n",
       "      <td>AJ414696</td>\n",
       "      <td>32308</td>\n",
       "      <td>0</td>\n",
       "      <td>linear</td>\n",
       "      <td>25.269</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>93.475300</td>\n",
       "      <td>51.11</td>\n",
       "      <td>48.89</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11152</th>\n",
       "      <td>AJ344259</td>\n",
       "      <td>35450</td>\n",
       "      <td>0</td>\n",
       "      <td>linear</td>\n",
       "      <td>25.199</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>79.904090</td>\n",
       "      <td>53.70</td>\n",
       "      <td>46.30</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11153</th>\n",
       "      <td>AJ550635</td>\n",
       "      <td>4554</td>\n",
       "      <td>0</td>\n",
       "      <td>circular</td>\n",
       "      <td>41.041</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>94.466403</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11154 rows  14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             id  genome_length  jumbophage  topology    gc_%  trna_count  \\\n",
       "0      MN335248           7045           0    linear  60.298           0   \n",
       "1      MK250029         540217           1  circular  25.796           0   \n",
       "2      MK250028         550053           1  circular  26.012           0   \n",
       "3      MK250027         551627           1  circular  26.022           0   \n",
       "4      MK250026         550702           1  circular  26.020           0   \n",
       "...         ...            ...         ...       ...     ...         ...   \n",
       "11149  EF380009           5386           0  circular  44.653           0   \n",
       "11150  DQ490056          26537           0    linear  37.099           0   \n",
       "11151  AJ414696          32308           0    linear  25.269           0   \n",
       "11152  AJ344259          35450           0    linear  25.199           0   \n",
       "11153  AJ550635           4554           0  circular  41.041           0   \n",
       "\n",
       "       cds_number  coding_capacity  positive_strand_%  negative_strand_%  \\\n",
       "0              14       104.229950              85.71              14.29   \n",
       "1               0         0.000000               0.00               0.00   \n",
       "2               0         0.000000               0.00               0.00   \n",
       "3               0         0.000000               0.00               0.00   \n",
       "4               0         0.000000               0.00               0.00   \n",
       "...           ...              ...                ...                ...   \n",
       "11149          11       175.362050             100.00               0.00   \n",
       "11150          47        91.547650              72.34              27.66   \n",
       "11151          45        93.475300              51.11              48.89   \n",
       "11152          54        79.904090              53.70              46.30   \n",
       "11153           8        94.466403             100.00               0.00   \n",
       "\n",
       "       molecule_type_ss-DNA  molecule_type_DNA  molecule_type_RNA  \\\n",
       "0                      True              False              False   \n",
       "1                     False               True              False   \n",
       "2                     False               True              False   \n",
       "3                     False               True              False   \n",
       "4                     False               True              False   \n",
       "...                     ...                ...                ...   \n",
       "11149                  True              False              False   \n",
       "11150                 False               True              False   \n",
       "11151                 False               True              False   \n",
       "11152                 False               True              False   \n",
       "11153                 False               True              False   \n",
       "\n",
       "       molecule_type_ss-RNA  \n",
       "0                     False  \n",
       "1                     False  \n",
       "2                     False  \n",
       "3                     False  \n",
       "4                     False  \n",
       "...                     ...  \n",
       "11149                 False  \n",
       "11150                 False  \n",
       "11151                 False  \n",
       "11152                 False  \n",
       "11153                 False  \n",
       "\n",
       "[11154 rows x 14 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_file_path = \"../data/interim/genbank_engineering/sequences.gb\"\n",
    "#genbank_file_path = \"../data/interim/pharokka/pharokka.gbk\"\n",
    "genbank_file_path = \"../data/interim/genbank_engineering/genbank_filtered_model_entries.gb\"\n",
    "# Lists to hold data\n",
    "ids = []\n",
    "genome_lengths = []\n",
    "gc_contents = []\n",
    "sequences = []\n",
    "reverse_complements = []\n",
    "cds_numbers = []\n",
    "positive_strands = []\n",
    "negative_strands = []\n",
    "coding_capacities = []\n",
    "molecule_types = []\n",
    "topologies = []\n",
    "trna_counts = []\n",
    "\n",
    "# Read the GenBank file\n",
    "for record in SeqIO.parse(genbank_file_path, \"genbank\"):\n",
    "    try:\n",
    "        # Attempt to access the sequence, which may raise UndefinedSequenceError\n",
    "        sequence = str(record.seq)\n",
    "        # print(record.id)\n",
    "    except UndefinedSequenceError:\n",
    "        # print(f\"Skipping record {record.id} as sequence is undefined.\")\n",
    "        continue  # Skip this record\n",
    "\n",
    "    # Calculate genome length and GC content\n",
    "    total_length = len(sequence)\n",
    "    gc_content = round((sequence.count('C') + sequence.count('G')) / total_length * 100, 3)\n",
    "\n",
    "    # Initialize counters\n",
    "    plus = 0\n",
    "    minus = 0\n",
    "    coding_count = 0\n",
    "    trna_count = 0\n",
    "    seen = set()  # Store seen barcodes\n",
    "\n",
    "    for feature in record.features:\n",
    "        start = feature.location.start\n",
    "        end = feature.location.end\n",
    "        length = len(FeatureLocation(start, end))\n",
    "        barcode = f\"{start}_{end}_{length}\"\n",
    "\n",
    "        if feature.type != 'source' and barcode not in seen:\n",
    "            coding_count += length\n",
    "            seen.add(barcode)\n",
    "\n",
    "        if feature.type == 'CDS':\n",
    "            if feature.location.strand == 1:\n",
    "                plus += 1\n",
    "            elif feature.location.strand == -1:\n",
    "                minus += 1\n",
    "        elif feature.type == 'tRNA':\n",
    "            trna_count += 1\n",
    "\n",
    "\n",
    "    # Calculate total number of CDS\n",
    "    total_CDS = plus + minus\n",
    "\n",
    "    # Calculate strand usage as a percentage\n",
    "    per_plus = round((plus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "    per_minus = round((minus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "\n",
    "    # Calculate coding capacity as a percentage\n",
    "    coding_capacity = (coding_count / total_length) * 100\n",
    "\n",
    "    # Extract molecule_type and topology\n",
    "    molecule_type = record.annotations.get('molecule_type', 'N/A')\n",
    "    topology = record.annotations.get('topology', 'N/A')\n",
    "\n",
    "    # Append data to lists\n",
    "    ids.append(record.id)\n",
    "    genome_lengths.append(total_length)\n",
    "    gc_contents.append(gc_content)\n",
    "    sequences.append(sequence)\n",
    "    reverse_complements.append(str(sequence[::-1]))\n",
    "    cds_numbers.append(total_CDS)\n",
    "    positive_strands.append(per_plus)\n",
    "    negative_strands.append(per_minus)\n",
    "    coding_capacities.append(coding_capacity)\n",
    "    molecule_types.append(molecule_type)\n",
    "    topologies.append(topology)\n",
    "    trna_counts.append(trna_count)\n",
    "\n",
    "# Convert lists to pandas DataFrame\n",
    "df = pd.DataFrame({\n",
    "    'id': ids,\n",
    "    'genome_length': genome_lengths,\n",
    "    'gc_%': gc_contents,\n",
    "    'sequence': sequences,\n",
    "    'reverse_complement': reverse_complements,\n",
    "    'cds_number': cds_numbers,\n",
    "    'positive_strand_%': positive_strands,\n",
    "    'negative_strand_%': negative_strands,\n",
    "    'coding_capacity': coding_capacities,\n",
    "    'molecule_type': molecule_types,\n",
    "    'topology': topologies,\n",
    "    'trna_count': trna_counts\n",
    "})\n",
    "\n",
    "\n",
    "# Filter df based on previously known sequences with a staining class\n",
    "accession_values = pd.read_csv('../data/processed/model_data.csv')\n",
    "accession_values = accession_values['Accession'].tolist()\n",
    "df['id'] = df['id'].str[:-2]\n",
    "# df = df[df['id'].isin(accession_values)]\n",
    "\n",
    "\n",
    "# Check for unexpected molecule types\n",
    "expected_molecule_types = ['ss-DNA', 'DNA', 'RNA', 'ss-RNA']\n",
    "\n",
    "# Check and correct 'cRNA' entries\n",
    "cRNA_entries = df[df['molecule_type'] == 'cRNA']\n",
    "if not cRNA_entries.empty:\n",
    "    for entry_id in cRNA_entries['id']:\n",
    "        print(f\"Info: Entry with id '{entry_id}' has molecule type 'cRNA'. Changing it to 'RNA'.\")\n",
    "    df.loc[df['molecule_type'] == 'cRNA', 'molecule_type'] = 'RNA'\n",
    "\n",
    "# Check and correct 'cDNA' entries\n",
    "cDNA_entries = df[df['molecule_type'] == 'cDNA']\n",
    "if not cDNA_entries.empty:\n",
    "    for entry_id in cDNA_entries['id']:\n",
    "        print(f\"Info: Entry with id '{entry_id}' has molecule type 'cDNA'. Changing it to 'DNA'.\")\n",
    "    df.loc[df['molecule_type'] == 'cDNA', 'molecule_type'] = 'DNA'\n",
    "\n",
    "unexpected_types = set(df['molecule_type']) - set(expected_molecule_types)\n",
    "\n",
    "if unexpected_types:\n",
    "    for utype in unexpected_types:\n",
    "        # Get ids of entries with the unexpected molecule type\n",
    "        ids_to_exclude = df[df['molecule_type'] == utype]['id'].tolist()\n",
    "        for entry_id in ids_to_exclude:\n",
    "            print(f\"Warning: Entry with id '{entry_id}' has unrecognized molecule type '{utype}'. It will not be considered.\")\n",
    "        df = df[df['molecule_type'] != utype]\n",
    "\n",
    "        \n",
    "df = pd.get_dummies(df, columns=['molecule_type'])\n",
    "\n",
    "expected_columns = ['jumbophage', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']\n",
    "for col in expected_columns:\n",
    "    if col not in df.columns:\n",
    "        df[col] = 0  # Filling with zeros\n",
    "    df[col] = df[col].astype(bool)  # Convert to boolean\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "df['jumbophage'] = df['genome_length'].apply(lambda x: x >= 200000)\n",
    "df['jumbophage'] = df['jumbophage'].astype(int)  # Convert True/False to 1/0\n",
    "df = pd.get_dummies(df, columns=['topology'])\n",
    "# Change order of entries\n",
    "# df = df[['id', 'genome_length', 'jumbophage', 'topology', 'gc_%', 'trna_count','cds_number', 'coding_capacity','positive_strand_%','negative_strand_%', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']]\n",
    "# df = df[['id', 'genome_length', 'jumbophage', 'molecule_type', 'topology', 'gc_%', 'trna_count','cds_number', 'coding_capacity','positive_strand_%','negative_strand_%' ,'sequence', 'reverse_complement']]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id', 'genome_length', 'jumbophage', 'gc_%', 'trna_count', 'cds_number',\n",
       "       'coding_capacity', 'positive_strand_%', 'negative_strand_%',\n",
       "       'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA',\n",
       "       'molecule_type_ss-RNA', 'topology_circular', 'topology_linear'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.get_dummies(df, columns=['topology'])\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assign gram staining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>staining</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>jumbophage</th>\n",
       "      <th>gc_%</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>cds_number</th>\n",
       "      <th>coding_capacity</th>\n",
       "      <th>positive_strand_%</th>\n",
       "      <th>negative_strand_%</th>\n",
       "      <th>molecule_type_ss-DNA</th>\n",
       "      <th>molecule_type_DNA</th>\n",
       "      <th>molecule_type_RNA</th>\n",
       "      <th>molecule_type_ss-RNA</th>\n",
       "      <th>topology_circular</th>\n",
       "      <th>topology_linear</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MN335248</td>\n",
       "      <td>negative</td>\n",
       "      <td>7045</td>\n",
       "      <td>0</td>\n",
       "      <td>60.298</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>104.22995</td>\n",
       "      <td>85.71</td>\n",
       "      <td>14.29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MK250029</td>\n",
       "      <td>negative</td>\n",
       "      <td>540217</td>\n",
       "      <td>1</td>\n",
       "      <td>25.796</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MK250028</td>\n",
       "      <td>negative</td>\n",
       "      <td>550053</td>\n",
       "      <td>1</td>\n",
       "      <td>26.012</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MK250027</td>\n",
       "      <td>negative</td>\n",
       "      <td>551627</td>\n",
       "      <td>1</td>\n",
       "      <td>26.022</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MK250026</td>\n",
       "      <td>negative</td>\n",
       "      <td>550702</td>\n",
       "      <td>1</td>\n",
       "      <td>26.020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         id  staining  genome_length  jumbophage    gc_%  trna_count  \\\n",
       "0  MN335248  negative           7045           0  60.298           0   \n",
       "1  MK250029  negative         540217           1  25.796           0   \n",
       "2  MK250028  negative         550053           1  26.012           0   \n",
       "3  MK250027  negative         551627           1  26.022           0   \n",
       "4  MK250026  negative         550702           1  26.020           0   \n",
       "\n",
       "   cds_number  coding_capacity  positive_strand_%  negative_strand_%  \\\n",
       "0          14        104.22995              85.71              14.29   \n",
       "1           0          0.00000               0.00               0.00   \n",
       "2           0          0.00000               0.00               0.00   \n",
       "3           0          0.00000               0.00               0.00   \n",
       "4           0          0.00000               0.00               0.00   \n",
       "\n",
       "   molecule_type_ss-DNA  molecule_type_DNA  molecule_type_RNA  \\\n",
       "0                  True              False              False   \n",
       "1                 False               True              False   \n",
       "2                 False               True              False   \n",
       "3                 False               True              False   \n",
       "4                 False               True              False   \n",
       "\n",
       "   molecule_type_ss-RNA  topology_circular  topology_linear  \n",
       "0                 False              False             True  \n",
       "1                 False               True            False  \n",
       "2                 False               True            False  \n",
       "3                 False               True            False  \n",
       "4                 False               True            False  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def staining_feature(staining_df, features_df):\n",
    "    stain = pd.read_csv(staining_df, index_col = 0)\n",
    "    stain = stain[['Accession', 'staining']]\n",
    "    stain = stain.rename(columns={'Accession': 'id'})\n",
    "\n",
    "    features_df = pd.merge(features_df, stain, on='id', how='left')\n",
    "    features_df = features_df[['id', 'staining', 'genome_length', 'jumbophage', 'gc_%', 'trna_count','cds_number', 'coding_capacity',\n",
    "                               'positive_strand_%','negative_strand_%', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA',\n",
    "                               'topology_circular', 'topology_linear']]\n",
    "\n",
    "    return features_df\n",
    "\n",
    "staining_df = \"../data/interim/gram_staining/staining_assignation.csv\"\n",
    "df = staining_feature(staining_df, df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fixing predict_model.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def engineer_features(genbank_file):\n",
    "    # Lists to hold data\n",
    "    ids = []\n",
    "    genome_lengths = []\n",
    "    gc_contents = []\n",
    "    sequences = []\n",
    "    reverse_complements = []\n",
    "    cds_numbers = []\n",
    "    positive_strands = []\n",
    "    negative_strands = []\n",
    "    coding_capacities = []\n",
    "    molecule_types = []\n",
    "    topologies = []\n",
    "    trna_counts = []\n",
    "\n",
    "    \n",
    "    # Read the GenBank file\n",
    "    for record in SeqIO.parse(genbank_file, \"genbank\"):\n",
    "        try:\n",
    "            # Attempt to access the sequence, which may raise UndefinedSequenceError\n",
    "            sequence = str(record.seq)\n",
    "            # print(record.id)\n",
    "        except UndefinedSequenceError:\n",
    "            # print(f\"Skipping record {record.id} as sequence is undefined.\")\n",
    "            continue  # Skip this record\n",
    "\n",
    "        # Calculate genome length and GC content\n",
    "        total_length = len(sequence)\n",
    "        gc_content = round((sequence.count('C') + sequence.count('G')) / total_length * 100, 3)\n",
    "\n",
    "        # Initialize counters\n",
    "        plus = 0\n",
    "        minus = 0\n",
    "        coding_count = 0\n",
    "        trna_count = 0\n",
    "        seen = set()  # Store seen barcodes\n",
    "\n",
    "        for feature in record.features:\n",
    "            start = feature.location.start\n",
    "            end = feature.location.end\n",
    "            length = len(FeatureLocation(start, end))\n",
    "            barcode = f\"{start}_{end}_{length}\"\n",
    "\n",
    "            if feature.type != 'source' and barcode not in seen:\n",
    "                coding_count += length\n",
    "                seen.add(barcode)\n",
    "\n",
    "            if feature.type == 'CDS':\n",
    "                if feature.location.strand == 1:\n",
    "                    plus += 1\n",
    "                elif feature.location.strand == -1:\n",
    "                    minus += 1\n",
    "            elif feature.type == 'tRNA':\n",
    "                trna_count += 1\n",
    "\n",
    "        \n",
    "        # Calculate total number of CDS\n",
    "        total_CDS = plus + minus\n",
    "\n",
    "        # Calculate strand usage as a percentage\n",
    "        per_plus = round((plus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "        per_minus = round((minus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "\n",
    "        # Calculate coding capacity as a percentage\n",
    "        coding_capacity = (coding_count / total_length) * 100\n",
    "\n",
    "        # Extract molecule_type and topology\n",
    "        molecule_type = record.annotations.get('molecule_type', 'N/A')\n",
    "        topology = record.annotations.get('topology', 'N/A')\n",
    "\n",
    "        # Append data to lists\n",
    "        ids.append(record.id)\n",
    "        genome_lengths.append(total_length)\n",
    "        gc_contents.append(gc_content)\n",
    "        sequences.append(sequence)\n",
    "        reverse_complements.append(str(sequence[::-1]))\n",
    "        cds_numbers.append(total_CDS)\n",
    "        positive_strands.append(per_plus)\n",
    "        negative_strands.append(per_minus)\n",
    "        coding_capacities.append(coding_capacity)\n",
    "        molecule_types.append(molecule_type)\n",
    "        topologies.append(topology)\n",
    "        trna_counts.append(trna_count)\n",
    "    \n",
    "    print(\"Processing the entries...\")\n",
    "    # Convert lists to pandas DataFrame\n",
    "    df = pd.DataFrame({\n",
    "        'id': ids,\n",
    "        'genome_length': genome_lengths,\n",
    "        'gc_%': gc_contents,\n",
    "        'sequence': sequences,\n",
    "        'reverse_complement': reverse_complements,\n",
    "        'cds_number': cds_numbers,\n",
    "        'positive_strand_%': positive_strands,\n",
    "        'negative_strand_%': negative_strands,\n",
    "        'coding_capacity': coding_capacities,\n",
    "        'molecule_type': molecule_types,\n",
    "        'topology': topologies,\n",
    "        'trna_count': trna_counts\n",
    "    })\n",
    "\n",
    "\n",
    "    df['id'] = df['id'].str[:-2]\n",
    "\n",
    "    # Check for unexpected molecule types\n",
    "    expected_molecule_types = ['ss-DNA', 'DNA', 'RNA', 'ss-RNA']\n",
    "\n",
    "    # Check and correct 'cRNA' entries\n",
    "    cRNA_entries = df[df['molecule_type'] == 'cRNA']\n",
    "    if not cRNA_entries.empty:\n",
    "        for entry_id in cRNA_entries['id']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cRNA'. Changing it to 'RNA'.\")\n",
    "        df.loc[df['molecule_type'] == 'cRNA', 'molecule_type'] = 'RNA'\n",
    "\n",
    "    # Check and correct 'cDNA' entries\n",
    "    cDNA_entries = df[df['molecule_type'] == 'cDNA']\n",
    "    if not cDNA_entries.empty:\n",
    "        for entry_id in cDNA_entries['id']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cDNA'. Changing it to 'DNA'.\")\n",
    "        df.loc[df['molecule_type'] == 'cDNA', 'molecule_type'] = 'DNA'\n",
    "\n",
    "    unexpected_types = set(df['molecule_type']) - set(expected_molecule_types)\n",
    "\n",
    "    if unexpected_types:\n",
    "        for utype in unexpected_types:\n",
    "            # Get ids of entries with the unexpected molecule type\n",
    "            ids_to_exclude = df[df['molecule_type'] == utype]['id'].tolist()\n",
    "            for entry_id in ids_to_exclude:\n",
    "                print(f\"Warning: Entry with id '{entry_id}' has unrecognized molecule type '{utype}'. It will not be considered.\")\n",
    "            df = df[df['molecule_type'] != utype]\n",
    "            \n",
    "    df = pd.get_dummies(df, columns=['molecule_type'])\n",
    "\n",
    "    expected_columns = ['jumbophage', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']\n",
    "    for col in expected_columns:\n",
    "        if col not in df.columns:\n",
    "            df[col] = 0  # Filling with zeros\n",
    "        df[col] = df[col].astype(bool)  # Convert to boolean\n",
    "\n",
    "    df['jumbophage'] = df['genome_length'].apply(lambda x: x >= 200000)\n",
    "    df['jumbophage'] = df['jumbophage'].astype(int)  # Convert True/False to 1/0\n",
    "    df = pd.get_dummies(df, columns=['topology'])\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing the entries...\n",
      "Warning: Entry with id 'NR_074910' has unrecognized molecule type 'rRNA'. It will not be considered.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_length</th>\n",
       "      <th>jumbophage</th>\n",
       "      <th>gc_%</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>cds_number</th>\n",
       "      <th>coding_capacity</th>\n",
       "      <th>positive_strand_%</th>\n",
       "      <th>negative_strand_%</th>\n",
       "      <th>molecule_type_ss-DNA</th>\n",
       "      <th>molecule_type_DNA</th>\n",
       "      <th>molecule_type_RNA</th>\n",
       "      <th>molecule_type_ss-RNA</th>\n",
       "      <th>topology_circular</th>\n",
       "      <th>topology_linear</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GU339467</th>\n",
       "      <td>53332</td>\n",
       "      <td>0</td>\n",
       "      <td>64.530</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>92.857946</td>\n",
       "      <td>38.95</td>\n",
       "      <td>61.05</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KT462701</th>\n",
       "      <td>3580</td>\n",
       "      <td>0</td>\n",
       "      <td>48.296</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>99.748603</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KT462700</th>\n",
       "      <td>3609</td>\n",
       "      <td>0</td>\n",
       "      <td>41.701</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>94.846218</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KT462698</th>\n",
       "      <td>3776</td>\n",
       "      <td>0</td>\n",
       "      <td>44.147</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>98.013771</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MF417929</th>\n",
       "      <td>32618</td>\n",
       "      <td>0</td>\n",
       "      <td>39.218</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>88.635109</td>\n",
       "      <td>16.67</td>\n",
       "      <td>83.33</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          genome_length  jumbophage    gc_%  trna_count  cds_number  \\\n",
       "id                                                                    \n",
       "GU339467          53332           0  64.530           1          95   \n",
       "KT462701           3580           0  48.296           0           1   \n",
       "KT462700           3609           0  41.701           0           3   \n",
       "KT462698           3776           0  44.147           0           3   \n",
       "MF417929          32618           0  39.218           0          42   \n",
       "\n",
       "          coding_capacity  positive_strand_%  negative_strand_%  \\\n",
       "id                                                                \n",
       "GU339467        92.857946              38.95              61.05   \n",
       "KT462701        99.748603             100.00               0.00   \n",
       "KT462700        94.846218             100.00               0.00   \n",
       "KT462698        98.013771             100.00               0.00   \n",
       "MF417929        88.635109              16.67              83.33   \n",
       "\n",
       "          molecule_type_ss-DNA  molecule_type_DNA  molecule_type_RNA  \\\n",
       "id                                                                     \n",
       "GU339467                 False               True              False   \n",
       "KT462701                 False              False               True   \n",
       "KT462700                 False              False               True   \n",
       "KT462698                 False              False               True   \n",
       "MF417929                 False               True              False   \n",
       "\n",
       "          molecule_type_ss-RNA  topology_circular  topology_linear  \n",
       "id                                                                  \n",
       "GU339467                 False              False             True  \n",
       "KT462701                 False              False             True  \n",
       "KT462700                 False              False             True  \n",
       "KT462698                 False              False             True  \n",
       "MF417929                 False              False             True  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"../data/interim/genbank_engineering/50_sequences.gb\"\n",
    "df = engineer_features(path)\n",
    "features = ['genome_length', 'jumbophage', 'gc_%',\n",
    "       'trna_count', 'cds_number', 'coding_capacity', 'positive_strand_%',\n",
    "       'negative_strand_%', 'molecule_type_ss-DNA', 'molecule_type_DNA',\n",
    "       'molecule_type_RNA', 'molecule_type_ss-RNA', 'topology_circular','topology_linear']\n",
    "df = df.set_index(df.columns[0])\n",
    "df = df[features]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         id prediction  prediction_probability\n",
      "0  GU339467   positive                    0.76\n",
      "1  KT462701   negative                    0.97\n",
      "2  KT462700   negative                    0.92\n",
      "3  KT462698   negative                    0.97\n",
      "4  MF417929   negative                    0.63\n"
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Load the model using the provided path\n",
    "rf = pickle.load(open(\"../models/random_forest.pkl\", \"rb\"))\n",
    "\n",
    "# Predictions\n",
    "new_data_pred = rf.predict(df)\n",
    "\n",
    "# Get the probabilities for the predicted class for each instance\n",
    "probas = rf.predict_proba(df)\n",
    "predicted_indices = np.argmax(probas, axis=1)  # Get index of max proba for each sample\n",
    "new_data_pred_proba = [probas[i][predicted_indices[i]] for i in range(len(predicted_indices))]\n",
    "\n",
    "# Prepare the output DataFrame\n",
    "output_df = pd.DataFrame({\n",
    "    'id': df.index,\n",
    "    'prediction': new_data_pred,\n",
    "    'prediction_probability': new_data_pred_proba\n",
    "})\n",
    "\n",
    "# Uncomment the following line if you want to save the output to a CSV\n",
    "# output_df.to_csv('../data/output_predictions.csv', index=False)\n",
    "\n",
    "print(output_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.24, 0.76],\n",
       "       [0.97, 0.03],\n",
       "       [0.92, 0.08],\n",
       "       [0.97, 0.03],\n",
       "       [0.63, 0.37],\n",
       "       [0.9 , 0.1 ],\n",
       "       [0.88, 0.12],\n",
       "       [0.82, 0.18],\n",
       "       [0.4 , 0.6 ],\n",
       "       [0.94, 0.06],\n",
       "       [0.97, 0.03],\n",
       "       [0.96, 0.04],\n",
       "       [0.91, 0.09],\n",
       "       [0.95, 0.05],\n",
       "       [0.97, 0.03],\n",
       "       [0.95, 0.05],\n",
       "       [0.87, 0.13],\n",
       "       [0.89, 0.11],\n",
       "       [0.88, 0.12],\n",
       "       [0.97, 0.03],\n",
       "       [0.95, 0.05],\n",
       "       [0.94, 0.06],\n",
       "       [0.94, 0.06],\n",
       "       [0.95, 0.05],\n",
       "       [0.95, 0.05],\n",
       "       [0.95, 0.05],\n",
       "       [0.93, 0.07],\n",
       "       [0.9 , 0.1 ],\n",
       "       [0.95, 0.05],\n",
       "       [0.96, 0.04],\n",
       "       [0.89, 0.11],\n",
       "       [0.97, 0.03],\n",
       "       [0.96, 0.04],\n",
       "       [0.94, 0.06],\n",
       "       [0.92, 0.08],\n",
       "       [0.97, 0.03],\n",
       "       [0.82, 0.18],\n",
       "       [0.91, 0.09],\n",
       "       [0.92, 0.08],\n",
       "       [0.97, 0.03],\n",
       "       [0.83, 0.17],\n",
       "       [0.94, 0.06],\n",
       "       [0.92, 0.08]])"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rf.predict_proba(df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "phage",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
