{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import FeatureLocation\n",
    "from Bio.Seq import UndefinedSequenceError\n",
    "import subprocess\n",
    "\n",
    "def unix_call(command):\n",
    "    job = subprocess.run(command.split()) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using pyrodigal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">EF380009.1_1\n",
      "MSRKIILIKQELLLLVYELNRSGLLAENEKIRPILAQLEKLLLCDLSPSTNDSVKN*\n",
      ">EF380009.1_2\n",
      "MSQVTEQSVRFQTALASIKLIQASAVLDLTEDDFDFLTSNKVWIATDRSRARRCVEACVYGTLDFVGYPRFPAPVEFIAAVIAYYVHPVNIQTACLIMEGAEFTENIINGVERPVKAAELFAFTLRVRAGNTDVLTDAEENVRQKLRAEGVM*\n",
      ">EF380009.1_3\n",
      "MSKGKKRSGARPGRPQPLRGTKGKRKGARLWYVGGQQF*\n",
      ">EF380009.1_4\n",
      "MSNIQTGAERMPHDLSHLGFLAGQIGRLITISTTPVIAGDSFEMDAVGALRLSPLRRGLAIDSTVDIFTFYVPHRHVYGEQWIKFMKDGVNATPLPTVNTTGYIDHAAFLGTINPDTNKIPKHLFQGYLNIYNNYFKAPWMPDRTEANPNELNEDDARYGFRCCHLKNIWTAPLPPETELSRQMTTSTTSIDIMGLQAAYANLHTDQERDYFMQRYRDVISSFGGKTSYDADNRPLLVMRSNFWASGYDVDGTDQTSLGQFSGRVQQTYKHSVPRFFVPEHGTMFTLALVRFPPTATKEIQYLNAKGALTYTDIAGDPVLYGNLPPREISMKDVFRSGDSSKKFKIAEGQWYRYAPSYVSPVYHLLEGFPFIQEPPSGDLQERVLIRHHDYDQCFQSVQLLQWNSQVKFNVTVYRNLPTTRDSIMTS*\n",
      ">EF380009.1_5\n",
      "MFQTFISRHNSNFFSDKLVLTSVTPASSAPVLQTPKATSSTLYFDSLTVNAGNGGFLHCIQMDTSVNAANQVVSVGADIAFDADPKFFACLVRFESSSVPTTLPTAYDVYPLDGRHDGGYYTVKDCVTIDVLPRTPGNNVYVGFMVWSNFTATKCRGLVSLNQVIKEIICLQPLK*\n",
      ">EF380009.1_6\n",
      "MFGAIAGGIASALAGGAMSKLFGGGQKAASGGIQGDVLATDNNTVGMGDAGIKSAIQGSNVPNHDEAAPNFVSGAMAKAGKGLLEGTLQAGTSAVSDKLLDLVGLGGKSAADKGKDTRDYLAAAFPELNAWERAGADASSAGMVDAGFENQKELTKMQLDNQKEIAEMQNETQKEIAGIQSATSRQNTKDQVYAQNEMLAYQQKESTARVASIMENTNLSKQQQVSEIMRQMLTQAQTAGQYFTNDQIKEMTRKVSAEVDLVHQQTQNQRYGSSHIGATAKDISNVVTDAASGVVDIFHGIDKAVADTWNNFWKDGKADGIGSNLSRK*\n",
      ">EF380009.1_7\n",
      "MVRSYYPSECHADYFDFERIEALKPAIEACGISTLSQSPMLGFHKQMDNRIKLLEEILSFRMQGVEFDNGDMYVDGHKAASDVRDEFVSVTEKLMDELAQCYNVLPQLDINNTIDHRPEGDEKWFLENEKTVTQFCRKLAAERPLKDIRDEYNYPKKKGIKDECSRLLEASTMKSRRGFTIQRLMNAMRQAHADGWFIVFDTLTLADDRLEAFYDNPNALRDYFRDIGRMVLAAEGRKANDSHADCYQYFCVPEYGTANGRLHFHAVHFMRTLPTGSVDPNFGRRVRNRRQLNSLQNTWPYGYSMPIAVRYTQDAFSRSGWLWPVDAKGEPLKATSYMAVGFYVAKYVNKKSDMDLAAKGLGAKEWNNSLKTKLSLLPKKLFRIRMSRNFGMKMLTMTNLSTECLIQLTKLGYDATPFNQILKQNAKREMRLRLGKVTVADVLAAQPVTTNLLKFMRASIKMIGVSNL\n"
     ]
    }
   ],
   "source": [
    "import Bio.SeqIO\n",
    "import pyrodigal\n",
    "\n",
    "record = Bio.SeqIO.read(\"../data/interim/genbank_engineering/sequence.gb\", \"genbank\")\n",
    "\n",
    "orf_finder = pyrodigal.GeneFinder(meta=True)\n",
    "for i, pred in enumerate(orf_finder.find_genes(bytes(record.seq))):\n",
    "    print(f\">{record.id}_{i+1}\")\n",
    "    print(pred.translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'import Bio.SeqIO\\nimport pyrodigal\\n\\ninput_handle = \"../data/interim/genbank_engineering/50_sequences_filtered.gb\"\\nfor record in SeqIO.parse(input_handle, \"genbank\") :\\n\\n    orf_finder = pyrodigal.GeneFinder(meta=True)\\n    for i, pred in enumerate(orf_finder.find_genes(bytes(record.seq))):\\n        print(f\">{record.id}_{i+1}\")\\n        print(pred.translate())'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''import Bio.SeqIO\n",
    "import pyrodigal\n",
    "\n",
    "input_handle = \"../data/interim/genbank_engineering/50_sequences_filtered.gb\"\n",
    "for record in SeqIO.parse(input_handle, \"genbank\") :\n",
    "\n",
    "    orf_finder = pyrodigal.GeneFinder(meta=True)\n",
    "    for i, pred in enumerate(orf_finder.find_genes(bytes(record.seq))):\n",
    "        print(f\">{record.id}_{i+1}\")\n",
    "        print(pred.translate())'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dealing with GenBank record EF380009.1\n"
     ]
    }
   ],
   "source": [
    "gbk_filename =  \"../data/interim/genbank_engineering/sequence.gb\"\n",
    "faa_filename = \"NC_005213_converted.fna\"\n",
    "input_handle  = open(gbk_filename, \"r\")\n",
    "output_handle = open(faa_filename, \"w\")\n",
    "\n",
    "for seq_record in SeqIO.parse(input_handle, \"genbank\") :\n",
    "    print (\"Dealing with GenBank record %s\" % seq_record.id)\n",
    "    output_handle.write(\">%s %s\\n%s\\n\" % (\n",
    "           seq_record.id,\n",
    "           seq_record.description,\n",
    "           seq_record.seq))\n",
    "\n",
    "output_handle.close()\n",
    "input_handle.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calling one sequence as a test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_directory = \"../data/interim/pharokka/\"\n",
    "database_dir = \"../data/external/pharokka_v1.4.0_databases\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'genbank_file_path2 =  \"../data/raw/inphared_8Sep2023/8Sep2023_phages_downloaded_from_genbank.gb\"\\ngenbank_file_path2 =  \"../data/interim/genbank_engineering/sequence.gb\"\\ngenbank_file_path2 = \"/Users/Alvaro/Desktop/prueba_pharokka2/pharokka.gbk\"\\n# Read the GenBank file\\nfor record in SeqIO.parse(genbank_file_path2, \"genbank\"):\\n\\n    \\n    if record.id == \"MN335248.1\":\\n        print(f\"Processing record: {record.id}\\n\")\\n        \\n        sequence = str(record.seq)\\n        print(sequence)\\n        break'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''genbank_file_path2 =  \"../data/raw/inphared_8Sep2023/8Sep2023_phages_downloaded_from_genbank.gb\"\n",
    "genbank_file_path2 =  \"../data/interim/genbank_engineering/sequence.gb\"\n",
    "genbank_file_path2 = \"/Users/Alvaro/Desktop/prueba_pharokka2/pharokka.gbk\"\n",
    "# Read the GenBank file\n",
    "for record in SeqIO.parse(genbank_file_path2, \"genbank\"):\n",
    "\n",
    "    \n",
    "    if record.id == \"MN335248.1\":\n",
    "        print(f\"Processing record: {record.id}\\n\")\n",
    "        \n",
    "        sequence = str(record.seq)\n",
    "        print(sequence)\n",
    "        break'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# database_dir = \"/Users/Alvaro/Downloads/pharokka_v1.4.0_databases\"\\ndatabase_dir = \"/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/\"\\nsequence = \"../data/interim/genbank_engineering/double_sequence.fna\"\\nsequence = \"../data/interim/genbank_engineering/sequence.fna\"\\noutput_dir = \"/Users/Alvaro/Desktop/prueba_pharokka\"\\nunix_call(f\\'pharokka.py -i {sequence} -o {output_dir} -f -t 2 -d {database_dir}\\')'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''# database_dir = \"/Users/Alvaro/Downloads/pharokka_v1.4.0_databases\"\n",
    "database_dir = \"/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/\"\n",
    "sequence = \"../data/interim/genbank_engineering/double_sequence.fna\"\n",
    "sequence = \"../data/interim/genbank_engineering/sequence.fna\"\n",
    "output_dir = \"/Users/Alvaro/Desktop/prueba_pharokka\"\n",
    "unix_call(f'pharokka.py -i {sequence} -o {output_dir} -f -t 2 -d {database_dir}')'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'# database_dir = \"/Users/Alvaro/Downloads/pharokka_v1.4.0_databases\"\\ndatabase_dir = \"/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/\"\\nsequence = \"../data/interim/genbank_engineering/double_sequence.fna\"\\noutput_dir = \"/Users/Alvaro/Desktop/full_pharokka\"\\nunix_call(f\\'pharokka.py -i {sequence} -o {output_dir} -f -t 10 -d {database_dir}\\')'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''# database_dir = \"/Users/Alvaro/Downloads/pharokka_v1.4.0_databases\"\n",
    "database_dir = \"/Users/Alvaro/miniconda3/envs/pharokkaENV/databases/\"\n",
    "sequence = \"../data/interim/genbank_engineering/double_sequence.fna\"\n",
    "output_dir = \"/Users/Alvaro/Desktop/full_pharokka\"\n",
    "unix_call(f'pharokka.py -i {sequence} -o {output_dir} -f -t 10 -d {database_dir}')'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# After pharokka: processing genbank files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/Alvaro/Library/Mobile Documents/com~apple~CloudDocs/DTU/Autumn 2023/Phage project/phage/notebooks/05_feature_engineering.ipynb Cell 12\u001b[0m line \u001b[0;36m1\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X14sZmlsZQ%3D%3D?line=15'>16</a>\u001b[0m trna_counts \u001b[39m=\u001b[39m []\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X14sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m \u001b[39m# Read the GenBank file\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X14sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m \u001b[39mfor\u001b[39;00m record \u001b[39min\u001b[39;00m SeqIO\u001b[39m.\u001b[39mparse(genbank_file_path, \u001b[39m\"\u001b[39m\u001b[39mgenbank\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X14sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X14sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m         \u001b[39m# Attempt to access the sequence, which may raise UndefinedSequenceError\u001b[39;00m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X14sZmlsZQ%3D%3D?line=21'>22</a>\u001b[0m         sequence \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(record\u001b[39m.\u001b[39mseq)\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/Bio/SeqIO/Interfaces.py:72\u001b[0m, in \u001b[0;36mSequenceIterator.__next__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     70\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Return the next entry.\"\"\"\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> 72\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mnext\u001b[39;49m(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mrecords)\n\u001b[1;32m     73\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m:\n\u001b[1;32m     74\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mshould_close_stream:\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/Bio/GenBank/Scanner.py:516\u001b[0m, in \u001b[0;36mInsdcScanner.parse_records\u001b[0;34m(self, handle, do_features)\u001b[0m\n\u001b[1;32m    514\u001b[0m \u001b[39mwith\u001b[39;00m as_handle(handle) \u001b[39mas\u001b[39;00m handle:\n\u001b[1;32m    515\u001b[0m     \u001b[39mwhile\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[0;32m--> 516\u001b[0m         record \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mparse(handle, do_features)\n\u001b[1;32m    517\u001b[0m         \u001b[39mif\u001b[39;00m record \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    518\u001b[0m             \u001b[39mbreak\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/Bio/GenBank/Scanner.py:499\u001b[0m, in \u001b[0;36mInsdcScanner.parse\u001b[0;34m(self, handle, do_features)\u001b[0m\n\u001b[1;32m    493\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mBio\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mGenBank\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m \u001b[39mimport\u001b[39;00m FeatureValueCleaner\n\u001b[1;32m    495\u001b[0m consumer \u001b[39m=\u001b[39m _FeatureConsumer(\n\u001b[1;32m    496\u001b[0m     use_fuzziness\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m, feature_cleaner\u001b[39m=\u001b[39mFeatureValueCleaner()\n\u001b[1;32m    497\u001b[0m )\n\u001b[0;32m--> 499\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mfeed(handle, consumer, do_features):\n\u001b[1;32m    500\u001b[0m     \u001b[39mreturn\u001b[39;00m consumer\u001b[39m.\u001b[39mdata\n\u001b[1;32m    501\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/Bio/GenBank/Scanner.py:470\u001b[0m, in \u001b[0;36mInsdcScanner.feed\u001b[0;34m(self, handle, consumer, do_features)\u001b[0m\n\u001b[1;32m    468\u001b[0m \u001b[39m# Features (common to both EMBL and GenBank):\u001b[39;00m\n\u001b[1;32m    469\u001b[0m \u001b[39mif\u001b[39;00m do_features:\n\u001b[0;32m--> 470\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_feed_feature_table(consumer, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mparse_features(skip\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m))\n\u001b[1;32m    471\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    472\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mparse_features(skip\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)  \u001b[39m# ignore the data\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/Bio/GenBank/Scanner.py:425\u001b[0m, in \u001b[0;36mInsdcScanner._feed_feature_table\u001b[0;34m(consumer, feature_tuples)\u001b[0m\n\u001b[1;32m    423\u001b[0m     consumer\u001b[39m.\u001b[39mfeature_qualifier(q_key, q_value)\n\u001b[1;32m    424\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 425\u001b[0m     consumer\u001b[39m.\u001b[39;49mfeature_qualifier(q_key, q_value\u001b[39m.\u001b[39;49mreplace(\u001b[39m\"\u001b[39;49m\u001b[39m\\n\u001b[39;49;00m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39m \u001b[39;49m\u001b[39m\"\u001b[39;49m))\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/Bio/GenBank/__init__.py:772\u001b[0m, in \u001b[0;36m_FeatureConsumer.feature_qualifier\u001b[0;34m(self, key, value)\u001b[0m\n\u001b[1;32m    769\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_cur_feature\u001b[39m.\u001b[39mqualifiers[key]\u001b[39m.\u001b[39mappend(value)\n\u001b[1;32m    770\u001b[0m \u001b[39m# otherwise start a new list of the key with its values\u001b[39;00m\n\u001b[1;32m    771\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 772\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_cur_feature\u001b[39m.\u001b[39;49mqualifiers[key] \u001b[39m=\u001b[39m [value]\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "genbank_file_path = \"../data/interim/genbank_engineering/sequences.gb\"\n",
    "#Â genbank_file_path = \"../data/interim/pharokka/pharokka.gbk\"\n",
    "genbank_file_path = \"../data/interim/genbank_engineering/genbank_filtered_model_entries.gb\"\n",
    "# Lists to hold data\n",
    "ids = []\n",
    "genome_lengths = []\n",
    "gc_contents = []\n",
    "sequences = []\n",
    "reverse_complements = []\n",
    "cds_numbers = []\n",
    "positive_strands = []\n",
    "negative_strands = []\n",
    "coding_capacities = []\n",
    "molecule_types = []\n",
    "topologies = []\n",
    "trna_counts = []\n",
    "\n",
    "# Read the GenBank file\n",
    "for record in SeqIO.parse(genbank_file_path, \"genbank\"):\n",
    "    try:\n",
    "        # Attempt to access the sequence, which may raise UndefinedSequenceError\n",
    "        sequence = str(record.seq)\n",
    "        # print(record.id)\n",
    "    except UndefinedSequenceError:\n",
    "        # print(f\"Skipping record {record.id} as sequence is undefined.\")\n",
    "        continue  # Skip this record\n",
    "\n",
    "    # Calculate genome length and GC content\n",
    "    total_length = len(sequence)\n",
    "    gc_content = round((sequence.count('C') + sequence.count('G')) / total_length * 100, 3)\n",
    "\n",
    "    # Initialize counters\n",
    "    plus = 0\n",
    "    minus = 0\n",
    "    coding_count = 0\n",
    "    trna_count = 0\n",
    "    seen = set()  # Store seen barcodes\n",
    "\n",
    "    for feature in record.features:\n",
    "        start = feature.location.start\n",
    "        end = feature.location.end\n",
    "        length = len(FeatureLocation(start, end))\n",
    "        barcode = f\"{start}_{end}_{length}\"\n",
    "\n",
    "        if feature.type != 'source' and barcode not in seen:\n",
    "            coding_count += length\n",
    "            seen.add(barcode)\n",
    "\n",
    "        if feature.type == 'CDS':\n",
    "            if feature.location.strand == 1:\n",
    "                plus += 1\n",
    "            elif feature.location.strand == -1:\n",
    "                minus += 1\n",
    "        elif feature.type == 'tRNA':\n",
    "            trna_count += 1\n",
    "\n",
    "\n",
    "    # Calculate total number of CDS\n",
    "    total_CDS = plus + minus\n",
    "\n",
    "    # Calculate strand usage as a percentage\n",
    "    per_plus = round((plus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "    per_minus = round((minus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "\n",
    "    # Calculate coding capacity as a percentage\n",
    "    coding_capacity = (coding_count / total_length) * 100\n",
    "\n",
    "    # Extract molecule_type and topology\n",
    "    molecule_type = record.annotations.get('molecule_type', 'N/A')\n",
    "    topology = record.annotations.get('topology', 'N/A')\n",
    "\n",
    "    # Append data to lists\n",
    "    ids.append(record.id)\n",
    "    genome_lengths.append(total_length)\n",
    "    gc_contents.append(gc_content)\n",
    "    sequences.append(sequence)\n",
    "    reverse_complements.append(str(sequence[::-1]))\n",
    "    cds_numbers.append(total_CDS)\n",
    "    positive_strands.append(per_plus)\n",
    "    negative_strands.append(per_minus)\n",
    "    coding_capacities.append(coding_capacity)\n",
    "    molecule_types.append(molecule_type)\n",
    "    topologies.append(topology)\n",
    "    trna_counts.append(trna_count)\n",
    "\n",
    "# Convert lists to pandas DataFrame\n",
    "df = pd.DataFrame({\n",
    "    'id': ids,\n",
    "    'genome_length': genome_lengths,\n",
    "    'gc_%': gc_contents,\n",
    "    'sequence': sequences,\n",
    "    'reverse_complement': reverse_complements,\n",
    "    'cds_number': cds_numbers,\n",
    "    'positive_strand_%': positive_strands,\n",
    "    'negative_strand_%': negative_strands,\n",
    "    'coding_capacity': coding_capacities,\n",
    "    'molecule_type': molecule_types,\n",
    "    'topology': topologies,\n",
    "    'trna_count': trna_counts\n",
    "})\n",
    "\n",
    "\n",
    "# Filter df based on previously known sequences with a staining class\n",
    "accession_values = pd.read_csv('../data/processed/model_data.csv')\n",
    "accession_values = accession_values['Accession'].tolist()\n",
    "df['id'] = df['id'].str[:-2]\n",
    "# df = df[df['id'].isin(accession_values)]\n",
    "\n",
    "\n",
    "# Check for unexpected molecule types\n",
    "expected_molecule_types = ['ss-DNA', 'DNA', 'RNA', 'ss-RNA']\n",
    "\n",
    "# Check and correct 'cRNA' entries\n",
    "cRNA_entries = df[df['molecule_type'] == 'cRNA']\n",
    "if not cRNA_entries.empty:\n",
    "    for entry_id in cRNA_entries['id']:\n",
    "        print(f\"Info: Entry with id '{entry_id}' has molecule type 'cRNA'. Changing it to 'RNA'.\")\n",
    "    df.loc[df['molecule_type'] == 'cRNA', 'molecule_type'] = 'RNA'\n",
    "\n",
    "# Check and correct 'cDNA' entries\n",
    "cDNA_entries = df[df['molecule_type'] == 'cDNA']\n",
    "if not cDNA_entries.empty:\n",
    "    for entry_id in cDNA_entries['id']:\n",
    "        print(f\"Info: Entry with id '{entry_id}' has molecule type 'cDNA'. Changing it to 'DNA'.\")\n",
    "    df.loc[df['molecule_type'] == 'cDNA', 'molecule_type'] = 'DNA'\n",
    "\n",
    "unexpected_types = set(df['molecule_type']) - set(expected_molecule_types)\n",
    "\n",
    "if unexpected_types:\n",
    "    for utype in unexpected_types:\n",
    "        # Get ids of entries with the unexpected molecule type\n",
    "        ids_to_exclude = df[df['molecule_type'] == utype]['id'].tolist()\n",
    "        for entry_id in ids_to_exclude:\n",
    "            print(f\"Warning: Entry with id '{entry_id}' has unrecognized molecule type '{utype}'. It will not be considered.\")\n",
    "        df = df[df['molecule_type'] != utype]\n",
    "\n",
    "        \n",
    "df = pd.get_dummies(df, columns=['molecule_type'])\n",
    "\n",
    "expected_columns = ['jumbophage', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']\n",
    "for col in expected_columns:\n",
    "    if col not in df.columns:\n",
    "        df[col] = 0  # Filling with zeros\n",
    "    df[col] = df[col].astype(bool)  # Convert to boolean\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "df['jumbophage'] = df['genome_length'].apply(lambda x: x >= 200000)\n",
    "df['jumbophage'] = df['jumbophage'].astype(int)  # Convert True/False to 1/0\n",
    "df = pd.get_dummies(df, columns=['topology'])\n",
    "# Change order of entries\n",
    "# df = df[['id', 'genome_length', 'jumbophage', 'topology', 'gc_%', 'trna_count','cds_number', 'coding_capacity','positive_strand_%','negative_strand_%', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']]\n",
    "# df = df[['id', 'genome_length', 'jumbophage', 'molecule_type', 'topology', 'gc_%', 'trna_count','cds_number', 'coding_capacity','positive_strand_%','negative_strand_%' ,'sequence', 'reverse_complement']]\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id', 'genome_length', 'jumbophage', 'gc_%', 'trna_count', 'cds_number',\n",
       "       'coding_capacity', 'positive_strand_%', 'negative_strand_%',\n",
       "       'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA',\n",
       "       'molecule_type_ss-RNA', 'topology_circular', 'topology_linear'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.get_dummies(df, columns=['topology'])\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Assign gram staining"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>staining</th>\n",
       "      <th>genome_length</th>\n",
       "      <th>jumbophage</th>\n",
       "      <th>gc_%</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>cds_number</th>\n",
       "      <th>coding_capacity</th>\n",
       "      <th>positive_strand_%</th>\n",
       "      <th>negative_strand_%</th>\n",
       "      <th>molecule_type_ss-DNA</th>\n",
       "      <th>molecule_type_DNA</th>\n",
       "      <th>molecule_type_RNA</th>\n",
       "      <th>molecule_type_ss-RNA</th>\n",
       "      <th>topology_circular</th>\n",
       "      <th>topology_linear</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MN335248</td>\n",
       "      <td>negative</td>\n",
       "      <td>7045</td>\n",
       "      <td>0</td>\n",
       "      <td>60.298</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>104.22995</td>\n",
       "      <td>85.71</td>\n",
       "      <td>14.29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MK250029</td>\n",
       "      <td>negative</td>\n",
       "      <td>540217</td>\n",
       "      <td>1</td>\n",
       "      <td>25.796</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MK250028</td>\n",
       "      <td>negative</td>\n",
       "      <td>550053</td>\n",
       "      <td>1</td>\n",
       "      <td>26.012</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MK250027</td>\n",
       "      <td>negative</td>\n",
       "      <td>551627</td>\n",
       "      <td>1</td>\n",
       "      <td>26.022</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MK250026</td>\n",
       "      <td>negative</td>\n",
       "      <td>550702</td>\n",
       "      <td>1</td>\n",
       "      <td>26.020</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         id  staining  genome_length  jumbophage    gc_%  trna_count  \\\n",
       "0  MN335248  negative           7045           0  60.298           0   \n",
       "1  MK250029  negative         540217           1  25.796           0   \n",
       "2  MK250028  negative         550053           1  26.012           0   \n",
       "3  MK250027  negative         551627           1  26.022           0   \n",
       "4  MK250026  negative         550702           1  26.020           0   \n",
       "\n",
       "   cds_number  coding_capacity  positive_strand_%  negative_strand_%  \\\n",
       "0          14        104.22995              85.71              14.29   \n",
       "1           0          0.00000               0.00               0.00   \n",
       "2           0          0.00000               0.00               0.00   \n",
       "3           0          0.00000               0.00               0.00   \n",
       "4           0          0.00000               0.00               0.00   \n",
       "\n",
       "   molecule_type_ss-DNA  molecule_type_DNA  molecule_type_RNA  \\\n",
       "0                  True              False              False   \n",
       "1                 False               True              False   \n",
       "2                 False               True              False   \n",
       "3                 False               True              False   \n",
       "4                 False               True              False   \n",
       "\n",
       "   molecule_type_ss-RNA  topology_circular  topology_linear  \n",
       "0                 False              False             True  \n",
       "1                 False               True            False  \n",
       "2                 False               True            False  \n",
       "3                 False               True            False  \n",
       "4                 False               True            False  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def staining_feature(staining_df, features_df):\n",
    "    stain = pd.read_csv(staining_df, index_col = 0)\n",
    "    stain = stain[['Accession', 'staining']]\n",
    "    stain = stain.rename(columns={'Accession': 'id'})\n",
    "\n",
    "    features_df = pd.merge(features_df, stain, on='id', how='left')\n",
    "    features_df = features_df[['id', 'staining', 'genome_length', 'jumbophage', 'gc_%', 'trna_count','cds_number', 'coding_capacity',\n",
    "                               'positive_strand_%','negative_strand_%', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA',\n",
    "                               'topology_circular', 'topology_linear']]\n",
    "\n",
    "    return features_df\n",
    "\n",
    "staining_df = \"../data/interim/gram_staining/staining_assignation.csv\"\n",
    "df = staining_feature(staining_df, df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# After pharokka: merging output from several files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/Alvaro/Library/Mobile Documents/com~apple~CloudDocs/DTU/Autumn 2023/Phage project/phage/notebooks'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import FeatureLocation\n",
    "from Bio.Seq import UndefinedSequenceError\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../data/interim/pharokka/tmp_pharokka/MZ079855__Klebsiella_phage_vB_Kpn_3.fna_pharokka/'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "code = \"MZ079855__Klebsiella_phage_vB_Kpn_3\" \n",
    "entry = f\"../data/interim/pharokka/tmp_pharokka/{code}.fna_pharokka/\"\n",
    "entry"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, get id, topology, molecule type, sequence from genbank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>molecule_type_DNA</th>\n",
       "      <th>topology_linear</th>\n",
       "      <th>topology_circular</th>\n",
       "      <th>molecule_type_ss-DNA</th>\n",
       "      <th>molecule_type_RNA</th>\n",
       "      <th>molecule_type_ss-RNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ079855</td>\n",
       "      <td>TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         id                                           sequence  \\\n",
       "0  MZ079855  TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...   \n",
       "\n",
       "   molecule_type_DNA  topology_linear  topology_circular  \\\n",
       "0               True             True              False   \n",
       "\n",
       "   molecule_type_ss-DNA  molecule_type_RNA  molecule_type_ss-RNA  \n",
       "0                 False              False                 False  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_file_path = f'{entry}pharokka.gbk'\n",
    "def engineer_features(genbank_file):\n",
    "    # Lists to hold data\n",
    "    ids = []\n",
    "    sequences = []\n",
    "    molecule_types = []\n",
    "    topologies = []\n",
    "\n",
    "\n",
    "    # Read the GenBank file\n",
    "    for record in SeqIO.parse(genbank_file_path, \"genbank\"):\n",
    "        try:\n",
    "            # Attempt to access the sequence, which may raise UndefinedSequenceError\n",
    "            sequence = str(record.seq)\n",
    "            # print(record.id)\n",
    "        except UndefinedSequenceError:\n",
    "            # print(f\"Skipping record {record.id} as sequence is undefined.\")\n",
    "            continue  # Skip this record\n",
    "\n",
    "\n",
    "        # Extract molecule_type and topology\n",
    "        molecule_type = record.annotations.get('molecule_type', 'N/A')\n",
    "        topology = record.annotations.get('topology', 'N/A')\n",
    "\n",
    "        # Append data to lists\n",
    "        ids.append(record.id)\n",
    "        sequences.append(sequence)\n",
    "        molecule_types.append(molecule_type)\n",
    "        topologies.append(topology)\n",
    "\n",
    "\n",
    "    # Convert lists to pandas DataFrame\n",
    "    df = pd.DataFrame({\n",
    "        'id': ids,\n",
    "        'sequence': sequences,\n",
    "        'molecule_type': molecule_types,\n",
    "        'topology': topologies\n",
    "    })\n",
    "\n",
    "\n",
    "    # Filter df based on previously known sequences with a staining class\n",
    "    accession_values = pd.read_csv('../data/processed/model_data.csv')\n",
    "    accession_values = accession_values['Accession'].tolist()\n",
    "    df['id'] = df['id'].str[:-2]\n",
    "    # df = df[df['id'].isin(accession_values)]\n",
    "\n",
    "\n",
    "    # Check for unexpected molecule types\n",
    "    expected_molecule_types = ['ss-DNA', 'DNA', 'RNA', 'ss-RNA']\n",
    "\n",
    "    # Check and correct 'cRNA' entries\n",
    "    cRNA_entries = df[df['molecule_type'] == 'cRNA']\n",
    "    if not cRNA_entries.empty:\n",
    "        for entry_id in cRNA_entries['id']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cRNA'. Changing it to 'RNA'.\")\n",
    "        df.loc[df['molecule_type'] == 'cRNA', 'molecule_type'] = 'RNA'\n",
    "\n",
    "    # Check and correct 'cDNA' entries\n",
    "    cDNA_entries = df[df['molecule_type'] == 'cDNA']\n",
    "    if not cDNA_entries.empty:\n",
    "        for entry_id in cDNA_entries['id']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cDNA'. Changing it to 'DNA'.\")\n",
    "        df.loc[df['molecule_type'] == 'cDNA', 'molecule_type'] = 'DNA'\n",
    "\n",
    "    unexpected_types = set(df['molecule_type']) - set(expected_molecule_types)\n",
    "\n",
    "    if unexpected_types:\n",
    "        for utype in unexpected_types:\n",
    "            # Get ids of entries with the unexpected molecule type\n",
    "            ids_to_exclude = df[df['molecule_type'] == utype]['id'].tolist()\n",
    "            for entry_id in ids_to_exclude:\n",
    "                print(f\"Warning: Entry with id '{entry_id}' has unrecognized molecule type '{utype}'. It will not be considered.\")\n",
    "            df = df[df['molecule_type'] != utype]\n",
    "\n",
    "            \n",
    "    df = pd.get_dummies(df, columns=['molecule_type'])\n",
    "\n",
    "    df = pd.get_dummies(df, columns=['topology'])\n",
    "\n",
    "    expected_columns = ['topology_linear', 'topology_circular', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']\n",
    "    for col in expected_columns:\n",
    "        if col not in df.columns:\n",
    "            df[col] = 0  # Filling with zeros\n",
    "        df[col] = df[col].astype(bool)  # Convert to boolean\n",
    "    \n",
    "    df['id'] = df['id'].str.slice(0,8)\n",
    "    return df\n",
    "\n",
    "\n",
    "df = engineer_features(genbank_file_path)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perhaps first, getting everything from inphared"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing the entries...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_inphared</th>\n",
       "      <th>genome_length_inphared</th>\n",
       "      <th>gc_%_inphared</th>\n",
       "      <th>sequence_inphared</th>\n",
       "      <th>reverse_complement_inphared</th>\n",
       "      <th>cds_number_inphared</th>\n",
       "      <th>positive_strand_%_inphared</th>\n",
       "      <th>negative_strand_%_inphared</th>\n",
       "      <th>coding_capacity_inphared</th>\n",
       "      <th>molecule_type_inphared_DNA</th>\n",
       "      <th>topology_inphared_linear</th>\n",
       "      <th>jumbophage_inphared</th>\n",
       "      <th>topology_linear_inphared</th>\n",
       "      <th>topology_circular_inphared</th>\n",
       "      <th>molecule_inphared_type_ss-DNA</th>\n",
       "      <th>molecule_inphared_type_DNA</th>\n",
       "      <th>molecule_inphared_type_RNA</th>\n",
       "      <th>molecule_inphared_type_ss-RNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ079855</td>\n",
       "      <td>112003</td>\n",
       "      <td>41.361</td>\n",
       "      <td>TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...</td>\n",
       "      <td>TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...</td>\n",
       "      <td>186</td>\n",
       "      <td>71.51</td>\n",
       "      <td>28.49</td>\n",
       "      <td>84.663804</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  id_inphared  genome_length_inphared  gc_%_inphared  \\\n",
       "0    MZ079855                  112003         41.361   \n",
       "\n",
       "                                   sequence_inphared  \\\n",
       "0  TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...   \n",
       "\n",
       "                         reverse_complement_inphared  cds_number_inphared  \\\n",
       "0  TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...                  186   \n",
       "\n",
       "   positive_strand_%_inphared  negative_strand_%_inphared  \\\n",
       "0                       71.51                       28.49   \n",
       "\n",
       "   coding_capacity_inphared  molecule_type_inphared_DNA  \\\n",
       "0                 84.663804                        True   \n",
       "\n",
       "   topology_inphared_linear  jumbophage_inphared  topology_linear_inphared  \\\n",
       "0                      True                    0                     False   \n",
       "\n",
       "   topology_circular_inphared  molecule_inphared_type_ss-DNA  \\\n",
       "0                       False                          False   \n",
       "\n",
       "   molecule_inphared_type_DNA  molecule_inphared_type_RNA  \\\n",
       "0                       False                       False   \n",
       "\n",
       "   molecule_inphared_type_ss-RNA  \n",
       "0                          False  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genbank_file_path = f'{entry}pharokka.gbk'\n",
    "\n",
    "def engineer_features(genbank_file):\n",
    "    # Lists to hold data\n",
    "    ids = []\n",
    "    genome_lengths = []\n",
    "    gc_contents = []\n",
    "    sequences = []\n",
    "    reverse_complements = []\n",
    "    cds_numbers = []\n",
    "    positive_strands = []\n",
    "    negative_strands = []\n",
    "    coding_capacities = []\n",
    "    molecule_types = []\n",
    "    topologies = []\n",
    "    trna_counts = []\n",
    "\n",
    "    \n",
    "    # Read the GenBank file\n",
    "    for record in SeqIO.parse(genbank_file, \"genbank\"):\n",
    "        try:\n",
    "            # Attempt to access the sequence, which may raise UndefinedSequenceError\n",
    "            sequence = str(record.seq)\n",
    "            # print(record.id)\n",
    "        except UndefinedSequenceError:\n",
    "            # print(f\"Skipping record {record.id} as sequence is undefined.\")\n",
    "            continue  # Skip this record\n",
    "\n",
    "        # Calculate genome length and GC content\n",
    "        total_length = len(sequence)\n",
    "        gc_content = round((sequence.count('C') + sequence.count('G')) / total_length * 100, 3)\n",
    "\n",
    "        # Initialize counters\n",
    "        plus = 0\n",
    "        minus = 0\n",
    "        coding_count = 0\n",
    "        trna_count = 0\n",
    "        seen = set()  # Store seen barcodes\n",
    "\n",
    "        for feature in record.features:\n",
    "            start = feature.location.start\n",
    "            end = feature.location.end\n",
    "            length = len(FeatureLocation(start, end))\n",
    "            barcode = f\"{start}_{end}_{length}\"\n",
    "\n",
    "            if feature.type != 'source' and barcode not in seen:\n",
    "                coding_count += length\n",
    "                seen.add(barcode)\n",
    "\n",
    "            if feature.type == 'CDS':\n",
    "                if feature.location.strand == 1:\n",
    "                    plus += 1\n",
    "                elif feature.location.strand == -1:\n",
    "                    minus += 1\n",
    "            elif feature.type == 'tRNA':\n",
    "                trna_count += 1\n",
    "\n",
    "        \n",
    "        # Calculate total number of CDS\n",
    "        total_CDS = plus + minus\n",
    "\n",
    "        # Calculate strand usage as a percentage\n",
    "        per_plus = round((plus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "        per_minus = round((minus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "\n",
    "        # Calculate coding capacity as a percentage\n",
    "        coding_capacity = (coding_count / total_length) * 100\n",
    "\n",
    "        # Extract molecule_type and topology\n",
    "        molecule_type = record.annotations.get('molecule_type', 'N/A')\n",
    "        topology = record.annotations.get('topology', 'N/A')\n",
    "\n",
    "        # Append data to lists\n",
    "        ids.append(record.id)\n",
    "        genome_lengths.append(total_length)\n",
    "        gc_contents.append(gc_content)\n",
    "        sequences.append(sequence)\n",
    "        reverse_complements.append(str(sequence[::-1]))\n",
    "        cds_numbers.append(total_CDS)\n",
    "        positive_strands.append(per_plus)\n",
    "        negative_strands.append(per_minus)\n",
    "        coding_capacities.append(coding_capacity)\n",
    "        molecule_types.append(molecule_type)\n",
    "        topologies.append(topology)\n",
    "        trna_counts.append(trna_count)\n",
    "    \n",
    "    print(\"Processing the entries...\")\n",
    "    # Convert lists to pandas DataFrame\n",
    "    df = pd.DataFrame({\n",
    "        'id_inphared': ids,\n",
    "        'genome_length_inphared': genome_lengths,\n",
    "        'gc_%_inphared': gc_contents,\n",
    "        'sequence_inphared': sequences,\n",
    "        'reverse_complement_inphared': reverse_complements,\n",
    "        'cds_number_inphared': cds_numbers,\n",
    "        'positive_strand_%_inphared': positive_strands,\n",
    "        'negative_strand_%_inphared': negative_strands,\n",
    "        'coding_capacity_inphared': coding_capacities,\n",
    "        'molecule_type_inphared': molecule_types,\n",
    "        'topology_inphared': topologies\n",
    "    })\n",
    "\n",
    "\n",
    "    df['id_inphared'] = df['id_inphared'].str[:-2]\n",
    "\n",
    "    # Check for unexpected molecule types\n",
    "    expected_molecule_types = ['ss-DNA', 'DNA', 'RNA', 'ss-RNA']\n",
    "\n",
    "    # Check and correct 'cRNA' entries\n",
    "    cRNA_entries = df[df['molecule_type_inphared'] == 'cRNA']\n",
    "    if not cRNA_entries.empty:\n",
    "        for entry_id in cRNA_entries['id_inphared']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cRNA'. Changing it to 'RNA'.\")\n",
    "        df.loc[df['molecule_type_inphared'] == 'cRNA', 'molecule_type_inphared'] = 'RNA'\n",
    "\n",
    "    # Check and correct 'cDNA' entries\n",
    "    cDNA_entries = df[df['molecule_type_inphared'] == 'cDNA']\n",
    "    if not cDNA_entries.empty:\n",
    "        for entry_id in cDNA_entries['id_inphared']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cDNA'. Changing it to 'DNA'.\")\n",
    "        df.loc[df['molecule_type_inphared'] == 'cDNA', 'molecule_type_inphared'] = 'DNA'\n",
    "\n",
    "    unexpected_types = set(df['molecule_type_inphared']) - set(expected_molecule_types)\n",
    "\n",
    "    if unexpected_types:\n",
    "        for utype in unexpected_types:\n",
    "            # Get ids of entries with the unexpected molecule type\n",
    "            ids_to_exclude = df[df['molecule_type_inphared'] == utype]['id'].tolist()\n",
    "            for entry_id in ids_to_exclude:\n",
    "                print(f\"Warning: Entry with id '{entry_id}' has unrecognized molecule type '{utype}'. It will not be considered.\")\n",
    "            df = df[df['molecule_type'] != utype]\n",
    "            \n",
    "    df = pd.get_dummies(df, columns=['molecule_type_inphared'])\n",
    "    df = pd.get_dummies(df, columns=['topology_inphared'])\n",
    "    expected_columns = ['jumbophage_inphared', 'topology_linear_inphared','topology_circular_inphared',\n",
    "    'molecule_inphared_type_ss-DNA', 'molecule_inphared_type_DNA', 'molecule_inphared_type_RNA', 'molecule_inphared_type_ss-RNA']\n",
    "    for col in expected_columns:\n",
    "        if col not in df.columns:\n",
    "            df[col] = 0  # Filling with zeros\n",
    "        df[col] = df[col].astype(bool)  # Convert to boolean\n",
    "\n",
    "    df['jumbophage_inphared'] = df['genome_length_inphared'].apply(lambda x: x >= 200000)\n",
    "    df['jumbophage_inphared'] = df['jumbophage_inphared'].astype(int)  # Convert True/False to 1/0\n",
    "    \n",
    "    df['id_inphared'] = df['id_inphared'].str.slice(0,8)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "df = engineer_features(genbank_file_path)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parsing pharokka_length_gc_cds_coding_density"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_inphared</th>\n",
       "      <th>genome_length_inphared</th>\n",
       "      <th>gc_%_inphared</th>\n",
       "      <th>sequence_inphared</th>\n",
       "      <th>reverse_complement_inphared</th>\n",
       "      <th>cds_number_inphared</th>\n",
       "      <th>positive_strand_%_inphared</th>\n",
       "      <th>negative_strand_%_inphared</th>\n",
       "      <th>coding_capacity_inphared</th>\n",
       "      <th>molecule_type_inphared_DNA</th>\n",
       "      <th>...</th>\n",
       "      <th>topology_circular_inphared</th>\n",
       "      <th>molecule_inphared_type_ss-DNA</th>\n",
       "      <th>molecule_inphared_type_DNA</th>\n",
       "      <th>molecule_inphared_type_RNA</th>\n",
       "      <th>molecule_inphared_type_ss-RNA</th>\n",
       "      <th>contig</th>\n",
       "      <th>length</th>\n",
       "      <th>gc_perc</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>cds_coding_density</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ079855</td>\n",
       "      <td>112003</td>\n",
       "      <td>41.361</td>\n",
       "      <td>TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...</td>\n",
       "      <td>TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...</td>\n",
       "      <td>186</td>\n",
       "      <td>71.51</td>\n",
       "      <td>28.49</td>\n",
       "      <td>84.663804</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>MZ079855</td>\n",
       "      <td>112003</td>\n",
       "      <td>41.36</td>\n",
       "      <td>11</td>\n",
       "      <td>82.86</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  id_inphared  genome_length_inphared  gc_%_inphared  \\\n",
       "0    MZ079855                  112003         41.361   \n",
       "\n",
       "                                   sequence_inphared  \\\n",
       "0  TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...   \n",
       "\n",
       "                         reverse_complement_inphared  cds_number_inphared  \\\n",
       "0  TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...                  186   \n",
       "\n",
       "   positive_strand_%_inphared  negative_strand_%_inphared  \\\n",
       "0                       71.51                       28.49   \n",
       "\n",
       "   coding_capacity_inphared  molecule_type_inphared_DNA  ...  \\\n",
       "0                 84.663804                        True  ...   \n",
       "\n",
       "   topology_circular_inphared  molecule_inphared_type_ss-DNA  \\\n",
       "0                       False                          False   \n",
       "\n",
       "   molecule_inphared_type_DNA  molecule_inphared_type_RNA  \\\n",
       "0                       False                       False   \n",
       "\n",
       "   molecule_inphared_type_ss-RNA    contig  length  gc_perc transl_table  \\\n",
       "0                          False  MZ079855  112003    41.36           11   \n",
       "\n",
       "   cds_coding_density  \n",
       "0               82.86  \n",
       "\n",
       "[1 rows x 23 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_length_gc_cds_density = pd.read_csv(f\"{entry}pharokka_length_gc_cds_density.tsv\", sep=\"\\t\")\n",
    "df_length_gc_cds_density['contig'] = df_length_gc_cds_density['contig'].str.slice(0,8)\n",
    "df_length_gc_cds_density\n",
    "\n",
    "df = pd.merge(df, df_length_gc_cds_density, left_on='id_inphared', right_on='contig', how='outer')\n",
    "del(df_length_gc_cds_density)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Getting CDS function information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_inphared</th>\n",
       "      <th>genome_length_inphared</th>\n",
       "      <th>gc_%_inphared</th>\n",
       "      <th>sequence_inphared</th>\n",
       "      <th>reverse_complement_inphared</th>\n",
       "      <th>cds_number_inphared</th>\n",
       "      <th>positive_strand_%_inphared</th>\n",
       "      <th>negative_strand_%_inphared</th>\n",
       "      <th>coding_capacity_inphared</th>\n",
       "      <th>molecule_type_inphared_DNA</th>\n",
       "      <th>...</th>\n",
       "      <th>host_takeover</th>\n",
       "      <th>integration and excision</th>\n",
       "      <th>lysis</th>\n",
       "      <th>nucleotide_metabolism</th>\n",
       "      <th>other</th>\n",
       "      <th>tRNAs</th>\n",
       "      <th>tail</th>\n",
       "      <th>tmRNAs</th>\n",
       "      <th>transcription</th>\n",
       "      <th>unkown_function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ079855</td>\n",
       "      <td>112003</td>\n",
       "      <td>41.361</td>\n",
       "      <td>TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...</td>\n",
       "      <td>TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...</td>\n",
       "      <td>186</td>\n",
       "      <td>71.51</td>\n",
       "      <td>28.49</td>\n",
       "      <td>84.663804</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  id_inphared  genome_length_inphared  gc_%_inphared  \\\n",
       "0    MZ079855                  112003         41.361   \n",
       "\n",
       "                                   sequence_inphared  \\\n",
       "0  TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...   \n",
       "\n",
       "                         reverse_complement_inphared  cds_number_inphared  \\\n",
       "0  TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...                  186   \n",
       "\n",
       "   positive_strand_%_inphared  negative_strand_%_inphared  \\\n",
       "0                       71.51                       28.49   \n",
       "\n",
       "   coding_capacity_inphared  molecule_type_inphared_DNA  ...  host_takeover  \\\n",
       "0                 84.663804                        True  ...              4   \n",
       "\n",
       "   integration and excision  lysis  nucleotide_metabolism  other  tRNAs  tail  \\\n",
       "0                         0      5                     38      8     23    21   \n",
       "\n",
       "   tmRNAs transcription  unkown_function  \n",
       "0       0             1               98  \n",
       "\n",
       "[1 rows x 40 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cds = pd.read_csv(f\"{entry}pharokka_cds_functions.tsv\", sep=\"\\t\")\n",
    "# Define a dictionary of replacements\n",
    "replacements = {\n",
    "    \"DNA, RNA and nucleotide metabolism\":\"nucleotide_metabolism\",\n",
    "    \"head and packaging\": \"head_packaging\",\n",
    "    \"moron, auxiliary metabolic gene and host takeover\": \"host_takeover\",\n",
    "    \"transcription regulation\": \"transcription\",\n",
    "    \"unknown function\": \"unkown_function\"\n",
    "}\n",
    "\n",
    "# Replace the values using the dictionary\n",
    "df_cds['Description'] = df_cds['Description'].replace(replacements)\n",
    "\n",
    "df_cds = df_cds.pivot(index='contig', columns='Description', values='Count').reset_index()\n",
    "df_cds['contig'] = df_cds['contig'].str.slice(0,8)\n",
    "\n",
    "df = pd.merge(df, df_cds, left_on='id_inphared', right_on='contig', how='outer')\n",
    "del(df_cds)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CDS final merged output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_frame = pd.read_csv(f\"{entry}pharokka_cds_final_merged_output.tsv\", sep=\"\\t\")\n",
    "frame_counts = df_frame.groupby('contig')['frame'].value_counts().unstack(fill_value=0)\n",
    "frame_counts.columns = ['frame_negative', 'frame_positive']\n",
    "frame_counts = frame_counts.rename_axis(None, axis=1).reset_index()\n",
    "frame_counts['contig'] = frame_counts['contig'].str.slice(0,8)\n",
    "df = pd.merge(df, frame_counts, left_on='id_inphared', right_on='contig', how='outer')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_inphared</th>\n",
       "      <th>genome_length_inphared</th>\n",
       "      <th>gc_%_inphared</th>\n",
       "      <th>sequence_inphared</th>\n",
       "      <th>reverse_complement_inphared</th>\n",
       "      <th>cds_number_inphared</th>\n",
       "      <th>positive_strand_%_inphared</th>\n",
       "      <th>negative_strand_%_inphared</th>\n",
       "      <th>coding_capacity_inphared</th>\n",
       "      <th>molecule_type_inphared_DNA</th>\n",
       "      <th>...</th>\n",
       "      <th>nucleotide_metabolism</th>\n",
       "      <th>other</th>\n",
       "      <th>tRNAs</th>\n",
       "      <th>tail</th>\n",
       "      <th>tmRNAs</th>\n",
       "      <th>transcription</th>\n",
       "      <th>unkown_function</th>\n",
       "      <th>contig</th>\n",
       "      <th>frame_negative</th>\n",
       "      <th>frame_positive</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MZ079855</td>\n",
       "      <td>112003</td>\n",
       "      <td>41.361</td>\n",
       "      <td>TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...</td>\n",
       "      <td>TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...</td>\n",
       "      <td>186</td>\n",
       "      <td>71.51</td>\n",
       "      <td>28.49</td>\n",
       "      <td>84.663804</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>38</td>\n",
       "      <td>8</td>\n",
       "      <td>23</td>\n",
       "      <td>21</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>98</td>\n",
       "      <td>MZ079855</td>\n",
       "      <td>133</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  id_inphared  genome_length_inphared  gc_%_inphared  \\\n",
       "0    MZ079855                  112003         41.361   \n",
       "\n",
       "                                   sequence_inphared  \\\n",
       "0  TGTCACCGTGAGTTCGCTTGTTTTTTAGAGCCACAGGAATTTTTTC...   \n",
       "\n",
       "                         reverse_complement_inphared  cds_number_inphared  \\\n",
       "0  TTTTAAAATAATCGTTTTACCCCGTTTTATCGAACTGTTCACTTTT...                  186   \n",
       "\n",
       "   positive_strand_%_inphared  negative_strand_%_inphared  \\\n",
       "0                       71.51                       28.49   \n",
       "\n",
       "   coding_capacity_inphared  molecule_type_inphared_DNA  ...  \\\n",
       "0                 84.663804                        True  ...   \n",
       "\n",
       "   nucleotide_metabolism  other  tRNAs  tail  tmRNAs  transcription  \\\n",
       "0                     38      8     23    21       0              1   \n",
       "\n",
       "   unkown_function    contig frame_negative  frame_positive  \n",
       "0               98  MZ079855            133              53  \n",
       "\n",
       "[1 rows x 43 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fixing predict_model.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def engineer_features(genbank_file):\n",
    "    # Lists to hold data\n",
    "    ids = []\n",
    "    genome_lengths = []\n",
    "    gc_contents = []\n",
    "    sequences = []\n",
    "    reverse_complements = []\n",
    "    cds_numbers = []\n",
    "    positive_strands = []\n",
    "    negative_strands = []\n",
    "    coding_capacities = []\n",
    "    molecule_types = []\n",
    "    topologies = []\n",
    "    trna_counts = []\n",
    "\n",
    "    \n",
    "    # Read the GenBank file\n",
    "    for record in SeqIO.parse(genbank_file, \"genbank\"):\n",
    "        try:\n",
    "            # Attempt to access the sequence, which may raise UndefinedSequenceError\n",
    "            sequence = str(record.seq)\n",
    "            # print(record.id)\n",
    "        except UndefinedSequenceError:\n",
    "            # print(f\"Skipping record {record.id} as sequence is undefined.\")\n",
    "            continue  # Skip this record\n",
    "\n",
    "        # Calculate genome length and GC content\n",
    "        total_length = len(sequence)\n",
    "        gc_content = round((sequence.count('C') + sequence.count('G')) / total_length * 100, 3)\n",
    "\n",
    "        # Initialize counters\n",
    "        plus = 0\n",
    "        minus = 0\n",
    "        coding_count = 0\n",
    "        trna_count = 0\n",
    "        seen = set()  # Store seen barcodes\n",
    "\n",
    "        for feature in record.features:\n",
    "            start = feature.location.start\n",
    "            end = feature.location.end\n",
    "            length = len(FeatureLocation(start, end))\n",
    "            barcode = f\"{start}_{end}_{length}\"\n",
    "\n",
    "            if feature.type != 'source' and barcode not in seen:\n",
    "                coding_count += length\n",
    "                seen.add(barcode)\n",
    "\n",
    "            if feature.type == 'CDS':\n",
    "                if feature.location.strand == 1:\n",
    "                    plus += 1\n",
    "                elif feature.location.strand == -1:\n",
    "                    minus += 1\n",
    "            elif feature.type == 'tRNA':\n",
    "                trna_count += 1\n",
    "\n",
    "        \n",
    "        # Calculate total number of CDS\n",
    "        total_CDS = plus + minus\n",
    "\n",
    "        # Calculate strand usage as a percentage\n",
    "        per_plus = round((plus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "        per_minus = round((minus / total_CDS) * 100, 2) if total_CDS != 0 else 0\n",
    "\n",
    "        # Calculate coding capacity as a percentage\n",
    "        coding_capacity = (coding_count / total_length) * 100\n",
    "\n",
    "        # Extract molecule_type and topology\n",
    "        molecule_type = record.annotations.get('molecule_type', 'N/A')\n",
    "        topology = record.annotations.get('topology', 'N/A')\n",
    "\n",
    "        # Append data to lists\n",
    "        ids.append(record.id)\n",
    "        genome_lengths.append(total_length)\n",
    "        gc_contents.append(gc_content)\n",
    "        sequences.append(sequence)\n",
    "        reverse_complements.append(str(sequence[::-1]))\n",
    "        cds_numbers.append(total_CDS)\n",
    "        positive_strands.append(per_plus)\n",
    "        negative_strands.append(per_minus)\n",
    "        coding_capacities.append(coding_capacity)\n",
    "        molecule_types.append(molecule_type)\n",
    "        topologies.append(topology)\n",
    "        trna_counts.append(trna_count)\n",
    "    \n",
    "    print(\"Processing the entries...\")\n",
    "    # Convert lists to pandas DataFrame\n",
    "    df = pd.DataFrame({\n",
    "        'id': ids,\n",
    "        'genome_length': genome_lengths,\n",
    "        'gc_%': gc_contents,\n",
    "        'sequence': sequences,\n",
    "        'reverse_complement': reverse_complements,\n",
    "        'cds_number': cds_numbers,\n",
    "        'positive_strand_%': positive_strands,\n",
    "        'negative_strand_%': negative_strands,\n",
    "        'coding_capacity': coding_capacities,\n",
    "        'molecule_type': molecule_types,\n",
    "        'topology': topologies,\n",
    "        'trna_count': trna_counts\n",
    "    })\n",
    "\n",
    "\n",
    "    df['id'] = df['id'].str[:-2]\n",
    "\n",
    "    # Check for unexpected molecule types\n",
    "    expected_molecule_types = ['ss-DNA', 'DNA', 'RNA', 'ss-RNA']\n",
    "\n",
    "    # Check and correct 'cRNA' entries\n",
    "    cRNA_entries = df[df['molecule_type'] == 'cRNA']\n",
    "    if not cRNA_entries.empty:\n",
    "        for entry_id in cRNA_entries['id']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cRNA'. Changing it to 'RNA'.\")\n",
    "        df.loc[df['molecule_type'] == 'cRNA', 'molecule_type'] = 'RNA'\n",
    "\n",
    "    # Check and correct 'cDNA' entries\n",
    "    cDNA_entries = df[df['molecule_type'] == 'cDNA']\n",
    "    if not cDNA_entries.empty:\n",
    "        for entry_id in cDNA_entries['id']:\n",
    "            print(f\"Info: Entry with id '{entry_id}' has molecule type 'cDNA'. Changing it to 'DNA'.\")\n",
    "        df.loc[df['molecule_type'] == 'cDNA', 'molecule_type'] = 'DNA'\n",
    "\n",
    "    unexpected_types = set(df['molecule_type']) - set(expected_molecule_types)\n",
    "\n",
    "    if unexpected_types:\n",
    "        for utype in unexpected_types:\n",
    "            # Get ids of entries with the unexpected molecule type\n",
    "            ids_to_exclude = df[df['molecule_type'] == utype]['id'].tolist()\n",
    "            for entry_id in ids_to_exclude:\n",
    "                print(f\"Warning: Entry with id '{entry_id}' has unrecognized molecule type '{utype}'. It will not be considered.\")\n",
    "            df = df[df['molecule_type'] != utype]\n",
    "            \n",
    "    df = pd.get_dummies(df, columns=['molecule_type'])\n",
    "\n",
    "    expected_columns = ['jumbophage', 'molecule_type_ss-DNA', 'molecule_type_DNA', 'molecule_type_RNA', 'molecule_type_ss-RNA']\n",
    "    for col in expected_columns:\n",
    "        if col not in df.columns:\n",
    "            df[col] = 0  # Filling with zeros\n",
    "        df[col] = df[col].astype(bool)  # Convert to boolean\n",
    "\n",
    "    df['jumbophage'] = df['genome_length'].apply(lambda x: x >= 200000)\n",
    "    df['jumbophage'] = df['jumbophage'].astype(int)  # Convert True/False to 1/0\n",
    "    df = pd.get_dummies(df, columns=['topology'])\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing the entries...\n",
      "Warning: Entry with id 'NR_074910' has unrecognized molecule type 'rRNA'. It will not be considered.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_length</th>\n",
       "      <th>jumbophage</th>\n",
       "      <th>gc_%</th>\n",
       "      <th>trna_count</th>\n",
       "      <th>cds_number</th>\n",
       "      <th>coding_capacity</th>\n",
       "      <th>positive_strand_%</th>\n",
       "      <th>negative_strand_%</th>\n",
       "      <th>molecule_type_ss-DNA</th>\n",
       "      <th>molecule_type_DNA</th>\n",
       "      <th>molecule_type_RNA</th>\n",
       "      <th>molecule_type_ss-RNA</th>\n",
       "      <th>topology_circular</th>\n",
       "      <th>topology_linear</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GU339467</th>\n",
       "      <td>53332</td>\n",
       "      <td>0</td>\n",
       "      <td>64.530</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>92.857946</td>\n",
       "      <td>38.95</td>\n",
       "      <td>61.05</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KT462701</th>\n",
       "      <td>3580</td>\n",
       "      <td>0</td>\n",
       "      <td>48.296</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>99.748603</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KT462700</th>\n",
       "      <td>3609</td>\n",
       "      <td>0</td>\n",
       "      <td>41.701</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>94.846218</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KT462698</th>\n",
       "      <td>3776</td>\n",
       "      <td>0</td>\n",
       "      <td>44.147</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>98.013771</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MF417929</th>\n",
       "      <td>32618</td>\n",
       "      <td>0</td>\n",
       "      <td>39.218</td>\n",
       "      <td>0</td>\n",
       "      <td>42</td>\n",
       "      <td>88.635109</td>\n",
       "      <td>16.67</td>\n",
       "      <td>83.33</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          genome_length  jumbophage    gc_%  trna_count  cds_number  \\\n",
       "id                                                                    \n",
       "GU339467          53332           0  64.530           1          95   \n",
       "KT462701           3580           0  48.296           0           1   \n",
       "KT462700           3609           0  41.701           0           3   \n",
       "KT462698           3776           0  44.147           0           3   \n",
       "MF417929          32618           0  39.218           0          42   \n",
       "\n",
       "          coding_capacity  positive_strand_%  negative_strand_%  \\\n",
       "id                                                                \n",
       "GU339467        92.857946              38.95              61.05   \n",
       "KT462701        99.748603             100.00               0.00   \n",
       "KT462700        94.846218             100.00               0.00   \n",
       "KT462698        98.013771             100.00               0.00   \n",
       "MF417929        88.635109              16.67              83.33   \n",
       "\n",
       "          molecule_type_ss-DNA  molecule_type_DNA  molecule_type_RNA  \\\n",
       "id                                                                     \n",
       "GU339467                 False               True              False   \n",
       "KT462701                 False              False               True   \n",
       "KT462700                 False              False               True   \n",
       "KT462698                 False              False               True   \n",
       "MF417929                 False               True              False   \n",
       "\n",
       "          molecule_type_ss-RNA  topology_circular  topology_linear  \n",
       "id                                                                  \n",
       "GU339467                 False              False             True  \n",
       "KT462701                 False              False             True  \n",
       "KT462700                 False              False             True  \n",
       "KT462698                 False              False             True  \n",
       "MF417929                 False              False             True  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = \"../data/interim/genbank_engineering/50_sequences.gb\"\n",
    "df = engineer_features(path)\n",
    "features = ['genome_length', 'jumbophage', 'gc_%',\n",
    "       'trna_count', 'cds_number', 'coding_capacity', 'positive_strand_%',\n",
    "       'negative_strand_%', 'molecule_type_ss-DNA', 'molecule_type_DNA',\n",
    "       'molecule_type_RNA', 'molecule_type_ss-RNA', 'topology_circular','topology_linear']\n",
    "df = df.set_index(df.columns[0])\n",
    "df = df[features]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/Alvaro/Library/Mobile Documents/com~apple~CloudDocs/DTU/Autumn 2023/Phage project/phage/notebooks/05_feature_engineering.ipynb Cell 32\u001b[0m line \u001b[0;36m6\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X42sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mnumpy\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mnp\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X42sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m \u001b[39m# Load the model using the provided path\u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X42sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m rf \u001b[39m=\u001b[39m pickle\u001b[39m.\u001b[39;49mload(\u001b[39mopen\u001b[39;49m(\u001b[39m\"\u001b[39;49m\u001b[39m../models/random_forest.pkl\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mrb\u001b[39;49m\u001b[39m\"\u001b[39;49m))\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X42sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m \u001b[39m# Predictions\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/05_feature_engineering.ipynb#X42sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m new_data_pred \u001b[39m=\u001b[39m rf\u001b[39m.\u001b[39mpredict(df)\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/sklearn/__init__.py:83\u001b[0m\n\u001b[1;32m     69\u001b[0m     \u001b[39m# We are not importing the rest of scikit-learn during the build\u001b[39;00m\n\u001b[1;32m     70\u001b[0m     \u001b[39m# process, as it may not be compiled yet\u001b[39;00m\n\u001b[1;32m     71\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     77\u001b[0m     \u001b[39m# later is linked to the OpenMP runtime to make it possible to introspect\u001b[39;00m\n\u001b[1;32m     78\u001b[0m     \u001b[39m# it and importing it first would fail if the OpenMP dll cannot be found.\u001b[39;00m\n\u001b[1;32m     79\u001b[0m     \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m \u001b[39mimport\u001b[39;00m (\n\u001b[1;32m     80\u001b[0m         __check_build,  \u001b[39m# noqa: F401\u001b[39;00m\n\u001b[1;32m     81\u001b[0m         _distributor_init,  \u001b[39m# noqa: F401\u001b[39;00m\n\u001b[1;32m     82\u001b[0m     )\n\u001b[0;32m---> 83\u001b[0m     \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mbase\u001b[39;00m \u001b[39mimport\u001b[39;00m clone\n\u001b[1;32m     84\u001b[0m     \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_show_versions\u001b[39;00m \u001b[39mimport\u001b[39;00m show_versions\n\u001b[1;32m     86\u001b[0m     __all__ \u001b[39m=\u001b[39m [\n\u001b[1;32m     87\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mcalibration\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     88\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mcluster\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    129\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mshow_versions\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    130\u001b[0m     ]\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/sklearn/base.py:19\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_config\u001b[39;00m \u001b[39mimport\u001b[39;00m config_context, get_config\n\u001b[1;32m     18\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mexceptions\u001b[39;00m \u001b[39mimport\u001b[39;00m InconsistentVersionWarning\n\u001b[0;32m---> 19\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m \u001b[39mimport\u001b[39;00m _IS_32BIT\n\u001b[1;32m     20\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_estimator_html_repr\u001b[39;00m \u001b[39mimport\u001b[39;00m estimator_html_repr\n\u001b[1;32m     21\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_metadata_requests\u001b[39;00m \u001b[39mimport\u001b[39;00m _MetadataRequester\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/sklearn/utils/__init__.py:22\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_bunch\u001b[39;00m \u001b[39mimport\u001b[39;00m Bunch\n\u001b[1;32m     21\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_estimator_html_repr\u001b[39;00m \u001b[39mimport\u001b[39;00m estimator_html_repr\n\u001b[0;32m---> 22\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_param_validation\u001b[39;00m \u001b[39mimport\u001b[39;00m Interval, validate_params\n\u001b[1;32m     23\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mclass_weight\u001b[39;00m \u001b[39mimport\u001b[39;00m compute_class_weight, compute_sample_weight\n\u001b[1;32m     24\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mdeprecation\u001b[39;00m \u001b[39mimport\u001b[39;00m deprecated\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/sklearn/utils/_param_validation.py:15\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mscipy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39msparse\u001b[39;00m \u001b[39mimport\u001b[39;00m csr_matrix, issparse\n\u001b[1;32m     14\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_config\u001b[39;00m \u001b[39mimport\u001b[39;00m config_context, get_config\n\u001b[0;32m---> 15\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mvalidation\u001b[39;00m \u001b[39mimport\u001b[39;00m _is_arraylike_not_scalar\n\u001b[1;32m     18\u001b[0m \u001b[39mclass\u001b[39;00m \u001b[39mInvalidParameterError\u001b[39;00m(\u001b[39mValueError\u001b[39;00m, \u001b[39mTypeError\u001b[39;00m):\n\u001b[1;32m     19\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Custom exception to be raised when the parameter of a class/method/function\u001b[39;00m\n\u001b[1;32m     20\u001b[0m \u001b[39m    does not have a valid type or value.\u001b[39;00m\n\u001b[1;32m     21\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/sklearn/utils/validation.py:28\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m \u001b[39mimport\u001b[39;00m get_config \u001b[39mas\u001b[39;00m _get_config\n\u001b[1;32m     27\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mexceptions\u001b[39;00m \u001b[39mimport\u001b[39;00m DataConversionWarning, NotFittedError, PositiveSpectrumWarning\n\u001b[0;32m---> 28\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_array_api\u001b[39;00m \u001b[39mimport\u001b[39;00m _asarray_with_order, _is_numpy_namespace, get_namespace\n\u001b[1;32m     29\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_isfinite\u001b[39;00m \u001b[39mimport\u001b[39;00m FiniteStatus, cy_isfinite\n\u001b[1;32m     30\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mfixes\u001b[39;00m \u001b[39mimport\u001b[39;00m _object_dtype_isnan\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/sklearn/utils/_array_api.py:9\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mscipy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mspecial\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mspecial\u001b[39;00m\n\u001b[1;32m      8\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_config\u001b[39;00m \u001b[39mimport\u001b[39;00m get_config\n\u001b[0;32m----> 9\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mfixes\u001b[39;00m \u001b[39mimport\u001b[39;00m parse_version\n\u001b[1;32m     12\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_check_array_api_dispatch\u001b[39m(array_api_dispatch):\n\u001b[1;32m     13\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"Check that array_api_compat is installed and NumPy version is compatible.\u001b[39;00m\n\u001b[1;32m     14\u001b[0m \n\u001b[1;32m     15\u001b[0m \u001b[39m    array_api_compat follows NEP29, which has a higher minimum NumPy version than\u001b[39;00m\n\u001b[1;32m     16\u001b[0m \u001b[39m    scikit-learn.\u001b[39;00m\n\u001b[1;32m     17\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/sklearn/utils/fixes.py:18\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mnumpy\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mnp\u001b[39;00m\n\u001b[1;32m     17\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mscipy\u001b[39;00m\n\u001b[0;32m---> 18\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mscipy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mstats\u001b[39;00m\n\u001b[1;32m     19\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mthreadpoolctl\u001b[39;00m\n\u001b[1;32m     21\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39msklearn\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/scipy/stats/__init__.py:485\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[39m.. _statsrefmanual:\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    480\u001b[0m \n\u001b[1;32m    481\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    483\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_warnings_errors\u001b[39;00m \u001b[39mimport\u001b[39;00m (ConstantInputWarning, NearConstantInputWarning,\n\u001b[1;32m    484\u001b[0m                                DegenerateDataWarning, FitError)\n\u001b[0;32m--> 485\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_stats_py\u001b[39;00m \u001b[39mimport\u001b[39;00m \u001b[39m*\u001b[39m\n\u001b[1;32m    486\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_variation\u001b[39;00m \u001b[39mimport\u001b[39;00m variation\n\u001b[1;32m    487\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mdistributions\u001b[39;00m \u001b[39mimport\u001b[39;00m \u001b[39m*\u001b[39m\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/scipy/stats/_stats_py.py:39\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mnumpy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mlib\u001b[39;00m \u001b[39mimport\u001b[39;00m NumpyVersion\n\u001b[1;32m     37\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mnumpy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mtesting\u001b[39;00m \u001b[39mimport\u001b[39;00m suppress_warnings\n\u001b[0;32m---> 39\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mscipy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mspatial\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mdistance\u001b[39;00m \u001b[39mimport\u001b[39;00m cdist\n\u001b[1;32m     40\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mscipy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mndimage\u001b[39;00m \u001b[39mimport\u001b[39;00m _measurements\n\u001b[1;32m     41\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mscipy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_lib\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_util\u001b[39;00m \u001b[39mimport\u001b[39;00m (check_random_state, MapWrapper,\n\u001b[1;32m     42\u001b[0m                               rng_integers, _rename_parameter, _contains_nan)\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/scipy/spatial/__init__.py:111\u001b[0m\n\u001b[1;32m    109\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_plotutils\u001b[39;00m \u001b[39mimport\u001b[39;00m \u001b[39m*\u001b[39m\n\u001b[1;32m    110\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_procrustes\u001b[39;00m \u001b[39mimport\u001b[39;00m procrustes\n\u001b[0;32m--> 111\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m_geometric_slerp\u001b[39;00m \u001b[39mimport\u001b[39;00m geometric_slerp\n\u001b[1;32m    113\u001b[0m \u001b[39m# Deprecated namespaces, to be removed in v2.0.0\u001b[39;00m\n\u001b[1;32m    114\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m \u001b[39mimport\u001b[39;00m ckdtree, kdtree, qhull\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/scipy/spatial/_geometric_slerp.py:9\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mtyping\u001b[39;00m \u001b[39mimport\u001b[39;00m TYPE_CHECKING\n\u001b[1;32m      8\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mnumpy\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mnp\u001b[39;00m\n\u001b[0;32m----> 9\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mscipy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mspatial\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mdistance\u001b[39;00m \u001b[39mimport\u001b[39;00m euclidean\n\u001b[1;32m     11\u001b[0m \u001b[39mif\u001b[39;00m TYPE_CHECKING:\n\u001b[1;32m     12\u001b[0m     \u001b[39mimport\u001b[39;00m \u001b[39mnumpy\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mtyping\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mnpt\u001b[39;00m\n",
      "File \u001b[0;32m<frozen importlib._bootstrap>:991\u001b[0m, in \u001b[0;36m_find_and_load\u001b[0;34m(name, import_)\u001b[0m\n",
      "File \u001b[0;32m<frozen importlib._bootstrap>:975\u001b[0m, in \u001b[0;36m_find_and_load_unlocked\u001b[0;34m(name, import_)\u001b[0m\n",
      "File \u001b[0;32m<frozen importlib._bootstrap>:671\u001b[0m, in \u001b[0;36m_load_unlocked\u001b[0;34m(spec)\u001b[0m\n",
      "File \u001b[0;32m<frozen importlib._bootstrap_external>:839\u001b[0m, in \u001b[0;36mexec_module\u001b[0;34m(self, module)\u001b[0m\n",
      "File \u001b[0;32m<frozen importlib._bootstrap_external>:934\u001b[0m, in \u001b[0;36mget_code\u001b[0;34m(self, fullname)\u001b[0m\n",
      "File \u001b[0;32m<frozen importlib._bootstrap_external>:1032\u001b[0m, in \u001b[0;36mget_data\u001b[0;34m(self, path)\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Load the model using the provided path\n",
    "rf = pickle.load(open(\"../models/random_forest.pkl\", \"rb\"))\n",
    "\n",
    "# Predictions\n",
    "new_data_pred = rf.predict(df)\n",
    "\n",
    "# Get the probabilities for the predicted class for each instance\n",
    "probas = rf.predict_proba(df)\n",
    "predicted_indices = np.argmax(probas, axis=1)  # Get index of max proba for each sample\n",
    "new_data_pred_proba = [probas[i][predicted_indices[i]] for i in range(len(predicted_indices))]\n",
    "\n",
    "# Prepare the output DataFrame\n",
    "output_df = pd.DataFrame({\n",
    "    'id': df.index,\n",
    "    'prediction': new_data_pred,\n",
    "    'prediction_probability': new_data_pred_proba\n",
    "})\n",
    "\n",
    "# Uncomment the following line if you want to save the output to a CSV\n",
    "# output_df.to_csv('../data/output_predictions.csv', index=False)\n",
    "\n",
    "print(output_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.24, 0.76],\n",
       "       [0.97, 0.03],\n",
       "       [0.92, 0.08],\n",
       "       [0.97, 0.03],\n",
       "       [0.63, 0.37],\n",
       "       [0.9 , 0.1 ],\n",
       "       [0.88, 0.12],\n",
       "       [0.82, 0.18],\n",
       "       [0.4 , 0.6 ],\n",
       "       [0.94, 0.06],\n",
       "       [0.97, 0.03],\n",
       "       [0.96, 0.04],\n",
       "       [0.91, 0.09],\n",
       "       [0.95, 0.05],\n",
       "       [0.97, 0.03],\n",
       "       [0.95, 0.05],\n",
       "       [0.87, 0.13],\n",
       "       [0.89, 0.11],\n",
       "       [0.88, 0.12],\n",
       "       [0.97, 0.03],\n",
       "       [0.95, 0.05],\n",
       "       [0.94, 0.06],\n",
       "       [0.94, 0.06],\n",
       "       [0.95, 0.05],\n",
       "       [0.95, 0.05],\n",
       "       [0.95, 0.05],\n",
       "       [0.93, 0.07],\n",
       "       [0.9 , 0.1 ],\n",
       "       [0.95, 0.05],\n",
       "       [0.96, 0.04],\n",
       "       [0.89, 0.11],\n",
       "       [0.97, 0.03],\n",
       "       [0.96, 0.04],\n",
       "       [0.94, 0.06],\n",
       "       [0.92, 0.08],\n",
       "       [0.97, 0.03],\n",
       "       [0.82, 0.18],\n",
       "       [0.91, 0.09],\n",
       "       [0.92, 0.08],\n",
       "       [0.97, 0.03],\n",
       "       [0.83, 0.17],\n",
       "       [0.94, 0.06],\n",
       "       [0.92, 0.08]])"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rf.predict_proba(df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "phage",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
