{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# from graph_part import train_test_validation_split, stratified_k_fold\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.metrics import accuracy_score, classification_report, f1_score, matthews_corrcoef, log_loss, mean_squared_error, balanced_accuracy_score\n",
    "from math import sqrt\n",
    "import numpy as np\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "from sklearn.utils.class_weight import compute_class_weight\n",
    "\n",
    "def matthews(y_true, y_pred):\n",
    "    \"\"\"\n",
    "    Calculate the Matthews Correlation Coefficient and other metrics.\n",
    "    \"\"\"\n",
    "    if type(y_true) == pd.Series:\n",
    "        y_true = y_true.values\n",
    "\n",
    "    P = len([x for x in y_true if x == 1])\n",
    "    N = len([x for x in y_true if x == 0])\n",
    "\n",
    "    Tp, Fp = 0, 0\n",
    "    for i in range(len(y_true)):\n",
    "        if y_true[i] == 1 and y_pred[i] == 1: Tp += 1\n",
    "        elif y_true[i] == 0 and y_pred[i] == 1: Fp += 1\n",
    "\n",
    "    Tn = N - Fp\n",
    "    Fn = P - Tp\n",
    "\n",
    "    try:\n",
    "        mcc = (Tp * Tn - Fp * Fn) / sqrt(\n",
    "            (Tn + Fn) * (Tn + Fp) * (Tp + Fn) * (Tp + Fp))\n",
    "    except ZeroDivisionError:\n",
    "        mcc = 0\n",
    "\n",
    "    return (mcc, f\" \\n \\\n",
    "    P: {P:_} \\n \\\n",
    "    Tp: {Tp:_} \\n \\\n",
    "    Fp: {Fp:_} \\n \\\n",
    "    N: {N:_} \\n \\\n",
    "    Tn: {Tn:_} \\n \\\n",
    "    Fn: {Fn:_}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11156, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>staining</th>\n",
       "      <th>Genome Length (bp)</th>\n",
       "      <th>Jumbophage</th>\n",
       "      <th>molGC (%)</th>\n",
       "      <th>Number CDS</th>\n",
       "      <th>Positive Strand (%)</th>\n",
       "      <th>Negative Strand (%)</th>\n",
       "      <th>Coding Capacity (%)</th>\n",
       "      <th>tRNAs</th>\n",
       "      <th>Molecule_DNA</th>\n",
       "      <th>Molecule_RNA</th>\n",
       "      <th>Molecule_ss-DNA</th>\n",
       "      <th>Molecule_ss-RNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MN335248</td>\n",
       "      <td>negative</td>\n",
       "      <td>7045</td>\n",
       "      <td>False</td>\n",
       "      <td>60.298</td>\n",
       "      <td>13</td>\n",
       "      <td>84.615385</td>\n",
       "      <td>15.384615</td>\n",
       "      <td>88.828957</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MK250029</td>\n",
       "      <td>negative</td>\n",
       "      <td>540217</td>\n",
       "      <td>True</td>\n",
       "      <td>25.796</td>\n",
       "      <td>830</td>\n",
       "      <td>47.108434</td>\n",
       "      <td>52.891566</td>\n",
       "      <td>68.324951</td>\n",
       "      <td>30</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MK250028</td>\n",
       "      <td>negative</td>\n",
       "      <td>550053</td>\n",
       "      <td>True</td>\n",
       "      <td>26.012</td>\n",
       "      <td>859</td>\n",
       "      <td>52.270081</td>\n",
       "      <td>47.729919</td>\n",
       "      <td>69.188424</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MK250027</td>\n",
       "      <td>negative</td>\n",
       "      <td>551627</td>\n",
       "      <td>True</td>\n",
       "      <td>26.022</td>\n",
       "      <td>860</td>\n",
       "      <td>53.023256</td>\n",
       "      <td>46.976744</td>\n",
       "      <td>69.318761</td>\n",
       "      <td>33</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MK250026</td>\n",
       "      <td>negative</td>\n",
       "      <td>550702</td>\n",
       "      <td>True</td>\n",
       "      <td>26.020</td>\n",
       "      <td>859</td>\n",
       "      <td>53.201397</td>\n",
       "      <td>46.798603</td>\n",
       "      <td>69.363285</td>\n",
       "      <td>33</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Accession  staining  Genome Length (bp)  Jumbophage  molGC (%)  Number CDS  \\\n",
       "0  MN335248  negative                7045       False     60.298          13   \n",
       "1  MK250029  negative              540217        True     25.796         830   \n",
       "2  MK250028  negative              550053        True     26.012         859   \n",
       "3  MK250027  negative              551627        True     26.022         860   \n",
       "4  MK250026  negative              550702        True     26.020         859   \n",
       "\n",
       "   Positive Strand (%)  Negative Strand (%)  Coding Capacity (%)  tRNAs  \\\n",
       "0            84.615385            15.384615            88.828957      0   \n",
       "1            47.108434            52.891566            68.324951     30   \n",
       "2            52.270081            47.729919            69.188424     29   \n",
       "3            53.023256            46.976744            69.318761     33   \n",
       "4            53.201397            46.798603            69.363285     33   \n",
       "\n",
       "   Molecule_DNA  Molecule_RNA  Molecule_ss-DNA  Molecule_ss-RNA  \n",
       "0         False         False             True            False  \n",
       "1          True         False            False            False  \n",
       "2          True         False            False            False  \n",
       "3          True         False            False            False  \n",
       "4          True         False            False            False  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"../data/processed/model_data.csv\")\n",
    "\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "staining\n",
       "negative    7932\n",
       "positive    3224\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['staining'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking why df and df2 are different"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11154, 15)\n",
      "Index(['staining', 'genome_length', 'jumbophage', 'gc_%', 'trna_count',\n",
      "       'cds_number', 'coding_capacity', 'positive_strand_%',\n",
      "       'negative_strand_%', 'molecule_type_ss-DNA', 'molecule_type_DNA',\n",
      "       'molecule_type_RNA', 'molecule_type_ss-RNA', 'topology_circular',\n",
      "       'topology_linear'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "df2 = pd.read_csv(\"../data/processed/model_data2.csv\", index_col=0)\n",
    "\n",
    "print(df2.shape)\n",
    "df2.head()\n",
    "print(df2.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>staining</th>\n",
       "      <th>Genome Length (bp)</th>\n",
       "      <th>Jumbophage</th>\n",
       "      <th>molGC (%)</th>\n",
       "      <th>Number CDS</th>\n",
       "      <th>Positive Strand (%)</th>\n",
       "      <th>Negative Strand (%)</th>\n",
       "      <th>Coding Capacity (%)</th>\n",
       "      <th>tRNAs</th>\n",
       "      <th>Molecule_DNA</th>\n",
       "      <th>Molecule_RNA</th>\n",
       "      <th>Molecule_ss-DNA</th>\n",
       "      <th>Molecule_ss-RNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11080</th>\n",
       "      <td>MPU46938</td>\n",
       "      <td>positive</td>\n",
       "      <td>15664</td>\n",
       "      <td>False</td>\n",
       "      <td>69.344</td>\n",
       "      <td>25</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>91.183606</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11122</th>\n",
       "      <td>L67CG</td>\n",
       "      <td>positive</td>\n",
       "      <td>22195</td>\n",
       "      <td>False</td>\n",
       "      <td>35.990</td>\n",
       "      <td>39</td>\n",
       "      <td>53.846154</td>\n",
       "      <td>46.153846</td>\n",
       "      <td>92.507321</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11126</th>\n",
       "      <td>BL5GENM</td>\n",
       "      <td>positive</td>\n",
       "      <td>2435</td>\n",
       "      <td>False</td>\n",
       "      <td>33.265</td>\n",
       "      <td>4</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>51.252567</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Accession  staining  Genome Length (bp)  Jumbophage  molGC (%)  \\\n",
       "11080  MPU46938  positive               15664       False     69.344   \n",
       "11122     L67CG  positive               22195       False     35.990   \n",
       "11126   BL5GENM  positive                2435       False     33.265   \n",
       "\n",
       "       Number CDS  Positive Strand (%)  Negative Strand (%)  \\\n",
       "11080          25           100.000000             0.000000   \n",
       "11122          39            53.846154            46.153846   \n",
       "11126           4           100.000000             0.000000   \n",
       "\n",
       "       Coding Capacity (%)  tRNAs  Molecule_DNA  Molecule_RNA  \\\n",
       "11080            91.183606      0          True         False   \n",
       "11122            92.507321      0          True         False   \n",
       "11126            51.252567      0          True         False   \n",
       "\n",
       "       Molecule_ss-DNA  Molecule_ss-RNA  \n",
       "11080            False            False  \n",
       "11122            False            False  \n",
       "11126            False            False  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "not_in_df = df[~df['Accession'].isin(df2['id'])]\n",
    "not_in_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homology partitioning (Graph-Part)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>staining</th>\n",
       "      <th>Genome Length (bp)</th>\n",
       "      <th>Jumbophage</th>\n",
       "      <th>molGC (%)</th>\n",
       "      <th>Number CDS</th>\n",
       "      <th>Positive Strand (%)</th>\n",
       "      <th>Negative Strand (%)</th>\n",
       "      <th>Coding Capacity (%)</th>\n",
       "      <th>tRNAs</th>\n",
       "      <th>Molecule_DNA</th>\n",
       "      <th>Molecule_RNA</th>\n",
       "      <th>Molecule_ss-DNA</th>\n",
       "      <th>Molecule_ss-RNA</th>\n",
       "      <th>id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>negative</td>\n",
       "      <td>7045</td>\n",
       "      <td>False</td>\n",
       "      <td>60.298</td>\n",
       "      <td>13</td>\n",
       "      <td>84.615385</td>\n",
       "      <td>15.384615</td>\n",
       "      <td>88.828957</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>MN335248</td>\n",
       "      <td>AGTACCGCCCGAATTTCGCAGCAACCCAACCGACGCAAGCCCAACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>negative</td>\n",
       "      <td>540217</td>\n",
       "      <td>True</td>\n",
       "      <td>25.796</td>\n",
       "      <td>830</td>\n",
       "      <td>47.108434</td>\n",
       "      <td>52.891566</td>\n",
       "      <td>68.324951</td>\n",
       "      <td>30</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>MK250029</td>\n",
       "      <td>GGACAAAGTTTAAAATCAAGAATTGATAGAAAAACATTTAATAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>negative</td>\n",
       "      <td>550053</td>\n",
       "      <td>True</td>\n",
       "      <td>26.012</td>\n",
       "      <td>859</td>\n",
       "      <td>52.270081</td>\n",
       "      <td>47.729919</td>\n",
       "      <td>69.188424</td>\n",
       "      <td>29</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>MK250028</td>\n",
       "      <td>TTACTCATAACTTATCTTACCCATTCATTAATATAACCATATTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>negative</td>\n",
       "      <td>551627</td>\n",
       "      <td>True</td>\n",
       "      <td>26.022</td>\n",
       "      <td>860</td>\n",
       "      <td>53.023256</td>\n",
       "      <td>46.976744</td>\n",
       "      <td>69.318761</td>\n",
       "      <td>33</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>MK250027</td>\n",
       "      <td>TTACTCATAACTTATCTTACCCATTCATTAATATAACCATATTCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>negative</td>\n",
       "      <td>550702</td>\n",
       "      <td>True</td>\n",
       "      <td>26.020</td>\n",
       "      <td>859</td>\n",
       "      <td>53.201397</td>\n",
       "      <td>46.798603</td>\n",
       "      <td>69.363285</td>\n",
       "      <td>33</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>MK250026</td>\n",
       "      <td>TTACTCATAACTTATCTTACCCATTCATTAATATAACCATATTCAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   staining  Genome Length (bp)  Jumbophage  molGC (%)  Number CDS  \\\n",
       "0  negative                7045       False     60.298          13   \n",
       "1  negative              540217        True     25.796         830   \n",
       "2  negative              550053        True     26.012         859   \n",
       "3  negative              551627        True     26.022         860   \n",
       "4  negative              550702        True     26.020         859   \n",
       "\n",
       "   Positive Strand (%)  Negative Strand (%)  Coding Capacity (%)  tRNAs  \\\n",
       "0            84.615385            15.384615            88.828957      0   \n",
       "1            47.108434            52.891566            68.324951     30   \n",
       "2            52.270081            47.729919            69.188424     29   \n",
       "3            53.023256            46.976744            69.318761     33   \n",
       "4            53.201397            46.798603            69.363285     33   \n",
       "\n",
       "   Molecule_DNA  Molecule_RNA  Molecule_ss-DNA  Molecule_ss-RNA        id  \\\n",
       "0         False         False             True            False  MN335248   \n",
       "1          True         False            False            False  MK250029   \n",
       "2          True         False            False            False  MK250028   \n",
       "3          True         False            False            False  MK250027   \n",
       "4          True         False            False            False  MK250026   \n",
       "\n",
       "                                            sequence  \n",
       "0  AGTACCGCCCGAATTTCGCAGCAACCCAACCGACGCAAGCCCAACC...  \n",
       "1  GGACAAAGTTTAAAATCAAGAATTGATAGAAAAACATTTAATAGCA...  \n",
       "2  TTACTCATAACTTATCTTACCCATTCATTAATATAACCATATTCAT...  \n",
       "3  TTACTCATAACTTATCTTACCCATTCATTAATATAACCATATTCAT...  \n",
       "4  TTACTCATAACTTATCTTACCCATTCATTAATATAACCATATTCAT...  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences = pd.read_csv('../data/interim/feature_engineering/feature_engineering_genomes.csv')\n",
    "sequences = sequences[['id', 'sequence']]\n",
    "df.reset_index()\n",
    "df = df.merge(sequences, left_on='Accession', right_on = 'id', how='left')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "temp/seq_db exists and will be overwritten\n",
      "createdb --dbtype 1 graphpart_api.fasta.tmp temp/seq_db \n",
      "\n",
      "MMseqs Version:       \t13.45111\n",
      "Database type         \t1\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[=\n",
      "Time for merging to seq_db_h: 0h 0m 0s 235ms\n",
      "Time for merging to seq_db: 0h 0m 1s 715ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 7s 151ms\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ln: temp/pref: File exists\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'full'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m/Users/Alvaro/Library/Mobile Documents/com~apple~CloudDocs/DTU/Autumn 2023/Phage project/phage/notebooks/06_model.ipynb Cell 6\u001b[0m line \u001b[0;36m4\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m labels \u001b[39m=\u001b[39m df[\u001b[39m'\u001b[39m\u001b[39mstaining\u001b[39m\u001b[39m'\u001b[39m]\u001b[39m.\u001b[39mto_numpy()\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39m# priority = df['experimental'].to_numpy()\u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m train_idx, test_idx, valid_idx \u001b[39m=\u001b[39m train_test_validation_split(sequences, \n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m                                                              labels\u001b[39m=\u001b[39;49mlabels,  \n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m                                                              alignment_mode\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mmmseqs2\u001b[39;49m\u001b[39m'\u001b[39;49m, \n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m                                                              threads \u001b[39m=\u001b[39;49m \u001b[39m2\u001b[39;49m,\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m \n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m                                                              test_size \u001b[39m=\u001b[39;49m \u001b[39m0.15\u001b[39;49m,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m                                                              valid_size \u001b[39m=\u001b[39;49m \u001b[39m0.05\u001b[39;49m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/Alvaro/Library/Mobile%20Documents/com~apple~CloudDocs/DTU/Autumn%202023/Phage%20project/phage/notebooks/06_model.ipynb#X26sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m                                                             )\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/graph_part/api.py:313\u001b[0m, in \u001b[0;36mtrain_test_validation_split\u001b[0;34m(sequences, labels, priority, test_size, valid_size, threshold, transformation, alignment_mode, initialization_mode, no_moving, remove_same, save_checkpoint_path, denominator, nucleotide, prefilter, triangular, threads, chunks, parallel_mode, gapopen, gapextend, endweight, endopen, endextend, matrix, edge_file, metric_column)\u001b[0m\n\u001b[1;32m    278\u001b[0m config \u001b[39m=\u001b[39m {\n\u001b[1;32m    279\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39malignment_mode\u001b[39m\u001b[39m\"\u001b[39m: alignment_mode,\n\u001b[1;32m    280\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mfasta_file\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39m\"\u001b[39m\u001b[39mgraphpart_api.fasta.tmp\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    309\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mremoval_type\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39mnot\u001b[39;00m remove_same,\n\u001b[1;32m    310\u001b[0m }\n\u001b[1;32m    312\u001b[0m \u001b[39m# 3. Partition\u001b[39;00m\n\u001b[0;32m--> 313\u001b[0m partition_assignment_df \u001b[39m=\u001b[39m run_partitioning(config, write_output_file\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m, write_json_report\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m, verbose\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m)\n\u001b[1;32m    314\u001b[0m os\u001b[39m.\u001b[39mremove(config[\u001b[39m'\u001b[39m\u001b[39mfasta_file\u001b[39m\u001b[39m'\u001b[39m])\n\u001b[1;32m    316\u001b[0m \u001b[39m# 4. Make output lists.\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/graph_part/graph_part.py:603\u001b[0m, in \u001b[0;36mrun_partitioning\u001b[0;34m(config, write_output_file, write_json_report, verbose)\u001b[0m\n\u001b[1;32m    597\u001b[0m json_dict[\u001b[39m'\u001b[39m\u001b[39mconfig\u001b[39m\u001b[39m'\u001b[39m][\u001b[39m'\u001b[39m\u001b[39mthreshold_transformed\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m threshold\n\u001b[1;32m    600\u001b[0m \u001b[39m## Processing starts here:\u001b[39;00m\n\u001b[1;32m    601\u001b[0m \n\u001b[1;32m    602\u001b[0m \u001b[39m## Load entities/samples as networkx graphs. labels contains label metadata.\u001b[39;00m\n\u001b[0;32m--> 603\u001b[0m full_graph, part_graph, labels \u001b[39m=\u001b[39m make_graphs_from_sequences(config, threshold, json_dict, verbose)\n\u001b[1;32m    606\u001b[0m \u001b[39m## Let's look at the number of edges\u001b[39;00m\n\u001b[1;32m    607\u001b[0m \u001b[39mprint\u001b[39m(\u001b[39m\"\u001b[39m\u001b[39mFull graph nr. of edges:\u001b[39m\u001b[39m\"\u001b[39m, full_graph\u001b[39m.\u001b[39mnumber_of_edges())\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/graph_part/graph_part.py:454\u001b[0m, in \u001b[0;36mmake_graphs_from_sequences\u001b[0;34m(config, threshold, json_dict, verbose)\u001b[0m\n\u001b[1;32m    452\u001b[0m \u001b[39melif\u001b[39;00m config[\u001b[39m'\u001b[39m\u001b[39malignment_mode\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m'\u001b[39m\u001b[39mmmseqs2\u001b[39m\u001b[39m'\u001b[39m:\n\u001b[1;32m    453\u001b[0m     \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mmmseqs_utils\u001b[39;00m \u001b[39mimport\u001b[39;00m generate_edges_mmseqs\n\u001b[0;32m--> 454\u001b[0m     generate_edges_mmseqs(config[\u001b[39m'\u001b[39;49m\u001b[39mfasta_file\u001b[39;49m\u001b[39m'\u001b[39;49m], full_graph, config[\u001b[39m'\u001b[39;49m\u001b[39mtransformation\u001b[39;49m\u001b[39m'\u001b[39;49m], threshold, config[\u001b[39m'\u001b[39;49m\u001b[39mthreshold\u001b[39;49m\u001b[39m'\u001b[39;49m], denominator\u001b[39m=\u001b[39;49mconfig[\u001b[39m'\u001b[39;49m\u001b[39mdenominator\u001b[39;49m\u001b[39m'\u001b[39;49m], delimiter\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m|\u001b[39;49m\u001b[39m'\u001b[39;49m, is_nucleotide\u001b[39m=\u001b[39;49mconfig[\u001b[39m'\u001b[39;49m\u001b[39mnucleotide\u001b[39;49m\u001b[39m'\u001b[39;49m], use_prefilter\u001b[39m=\u001b[39;49mconfig[\u001b[39m'\u001b[39;49m\u001b[39mprefilter\u001b[39;49m\u001b[39m'\u001b[39;49m])\n\u001b[1;32m    455\u001b[0m     elapsed_align \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mperf_counter() \u001b[39m-\u001b[39m json_dict[\u001b[39m'\u001b[39m\u001b[39mtime_script_start\u001b[39m\u001b[39m'\u001b[39m] \n\u001b[1;32m    456\u001b[0m     \u001b[39mif\u001b[39;00m verbose:\n",
      "File \u001b[0;32m~/miniconda3/envs/phage/lib/python3.8/site-packages/graph_part/mmseqs_utils.py:46\u001b[0m, in \u001b[0;36mgenerate_edges_mmseqs\u001b[0;34m(entity_fp, full_graph, tranformation, threshold_transformed, threshold_original, denominator, delimiter, is_nucleotide, use_prefilter)\u001b[0m\n\u001b[1;32m     43\u001b[0m     subprocess\u001b[39m.\u001b[39mrun([\u001b[39m'\u001b[39m\u001b[39mmmseqs_fake_prefilter.sh\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mtemp/seq_db\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mtemp/seq_db\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mtemp/pref\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mseq_db\u001b[39m\u001b[39m'\u001b[39m])\n\u001b[1;32m     45\u001b[0m \u001b[39m# 0: alignment length 1: shorter, 2: longer sequence\u001b[39;00m\n\u001b[0;32m---> 46\u001b[0m id_mode \u001b[39m=\u001b[39m {\u001b[39m'\u001b[39;49m\u001b[39mn_aligned\u001b[39;49m\u001b[39m'\u001b[39;49m:\u001b[39m'\u001b[39;49m\u001b[39m0\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mshortest\u001b[39;49m\u001b[39m'\u001b[39;49m:\u001b[39m'\u001b[39;49m\u001b[39m1\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m'\u001b[39;49m\u001b[39mlongest\u001b[39;49m\u001b[39m'\u001b[39;49m:\u001b[39m'\u001b[39;49m\u001b[39m2\u001b[39;49m\u001b[39m'\u001b[39;49m}[denominator]\n\u001b[1;32m     48\u001b[0m command \u001b[39m=\u001b[39m [\u001b[39m'\u001b[39m\u001b[39mmmseqs\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39malign\u001b[39m\u001b[39m'\u001b[39m,  \u001b[39m'\u001b[39m\u001b[39mtemp/seq_db\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mtemp/seq_db\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mtemp/pref\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39mtemp/align_db\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m--alignment-mode\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m3\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m-e\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39minf\u001b[39m\u001b[39m'\u001b[39m, \u001b[39m'\u001b[39m\u001b[39m--seq-id-mode\u001b[39m\u001b[39m'\u001b[39m, id_mode]\n\u001b[1;32m     49\u001b[0m \u001b[39mif\u001b[39;00m threshold_original \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "\u001b[0;31mKeyError\u001b[0m: 'full'"
     ]
    }
   ],
   "source": [
    "sequences = df['sequence'].to_numpy()\n",
    "labels = df['staining'].to_numpy()\n",
    "# priority = df['experimental'].to_numpy()\n",
    "train_idx, test_idx, valid_idx = train_test_validation_split(sequences, \n",
    "                                                             labels=labels,  \n",
    "                                                             alignment_mode='mmseqs2', \n",
    "                                                             threads = 2,\n",
    "                                                             threshold = 0.3,\n",
    "                                                             test_size = 0.15,\n",
    "                                                             valid_size = 0.05\n",
    "                                                            )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Homology partitioning (mmseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MN335248.1</td>\n",
       "      <td>MN335248.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MK250029.1</td>\n",
       "      <td>MK250029.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MK250028.1</td>\n",
       "      <td>MK250028.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MK250028.1</td>\n",
       "      <td>MK250024.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MK250028.1</td>\n",
       "      <td>MK250020.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Cluster   Sequences\n",
       "0  MN335248.1  MN335248.1\n",
       "1  MK250029.1  MK250029.1\n",
       "2  MK250028.1  MK250028.1\n",
       "3  MK250028.1  MK250024.1\n",
       "4  MK250028.1  MK250020.1"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_names = [\"Cluster\", \"Sequences\"]\n",
    "\n",
    "df = pd.read_csv(\"../data/interim/clustering/clusterRes_cluster.tsv\", sep=\"\\t\", header=None, names=column_names)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AB002632.1</td>\n",
       "      <td>[AB002632.1, NC_001956.1, KC357596.1, NC_02156...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AB009866.2</td>\n",
       "      <td>[AB009866.2, NC_002321.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AB012574.1</td>\n",
       "      <td>[AB012574.1, NC_005949.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AB043679.1</td>\n",
       "      <td>[AB043679.1, NC_002363.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AB044554.1</td>\n",
       "      <td>[AB044554.1, NC_002486.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5933</th>\n",
       "      <td>OR413342.1</td>\n",
       "      <td>[OR413342.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5934</th>\n",
       "      <td>OR413344.1</td>\n",
       "      <td>[OR413344.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5935</th>\n",
       "      <td>OR413347.1</td>\n",
       "      <td>[OR413347.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5936</th>\n",
       "      <td>OR437326.1</td>\n",
       "      <td>[OR437326.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5937</th>\n",
       "      <td>X98106.1</td>\n",
       "      <td>[X98106.1, NC_004305.1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5938 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Cluster                                          Sequences\n",
       "0     AB002632.1  [AB002632.1, NC_001956.1, KC357596.1, NC_02156...\n",
       "1     AB009866.2                          [AB009866.2, NC_002321.1]\n",
       "2     AB012574.1                          [AB012574.1, NC_005949.1]\n",
       "3     AB043679.1                          [AB043679.1, NC_002363.1]\n",
       "4     AB044554.1                          [AB044554.1, NC_002486.1]\n",
       "...          ...                                                ...\n",
       "5933  OR413342.1                                       [OR413342.1]\n",
       "5934  OR413344.1                                       [OR413344.1]\n",
       "5935  OR413347.1                                       [OR413347.1]\n",
       "5936  OR437326.1                                       [OR437326.1]\n",
       "5937    X98106.1                            [X98106.1, NC_004305.1]\n",
       "\n",
       "[5938 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped = df.groupby('Cluster')['Sequences'].agg(list).reset_index()\n",
    "grouped"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Traditional trainining and testing split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Features (independent variables)\n",
    "features = ['Genome Length (bp)', 'Jumbophage', 'molGC (%)', 'Number CDS',\n",
    "            'Positive Strand (%)', 'Negative Strand (%)', 'Coding Capacity (%)',\n",
    "            'tRNAs', 'Molecule_DNA', 'Molecule_RNA', 'Molecule_ss-DNA', 'Molecule_ss-RNA']\n",
    "\n",
    "features = ['genome_length', 'jumbophage', 'gc_%',\n",
    "       'trna_count', 'cds_number', 'coding_capacity', 'positive_strand_%',\n",
    "       'negative_strand_%', 'molecule_type_ss-DNA', 'molecule_type_DNA',\n",
    "       'molecule_type_RNA', 'molecule_type_ss-RNA', 'topology_circular','topology_linear']\n",
    "# Target variable (dependent variable)\n",
    "target = 'staining'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract features and target\n",
    "X = df[features]\n",
    "y = df[target]\n",
    "\n",
    "# Split the data into training and testing sets (80% train, 20% test)\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Logistic regression model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>LogisticRegression()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">LogisticRegression</label><div class=\"sk-toggleable__content\"><pre>LogisticRegression()</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "LogisticRegression()"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Train a logistic regression model\n",
    "logreg = LogisticRegression()\n",
    "logreg.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predict probabilities on the test set\n",
    "y_pred_proba = logreg.predict_proba(X_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Predict class labels for F1 score calculation\n",
    "y_pred = logreg.predict(X_test)\n",
    "\n",
    "# Encode labels for mean squared error calculation\n",
    "label_encoder = LabelEncoder()\n",
    "y_test_encoded = label_encoder.fit_transform(y_test)\n",
    "y_pred_encoded = label_encoder.transform(y_pred)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.7450716845878136\n",
      "Balanced Accuracy: 0.5717191836799154\n",
      "Log Loss: 0.5591185876984811\n",
      "Mean Squared Error: 0.25492831541218636\n",
      "\n",
      "F1 Score: 0.2751592356687898\n",
      "Matthews Correlation Coefficient: 0.2575572836055283\n",
      "Matthews Correlation Coefficient Details:  \n",
      "     P: 634 \n",
      "     Tp: 108 \n",
      "     Fp: 43 \n",
      "     N: 1_598 \n",
      "     Tn: 1_555 \n",
      "     Fn: 526\n",
      "\n",
      "Classification Report:\n",
      "               precision    recall  f1-score   support\n",
      "\n",
      "           0       0.75      0.97      0.85      1598\n",
      "           1       0.72      0.17      0.28       634\n",
      "\n",
      "    accuracy                           0.75      2232\n",
      "   macro avg       0.73      0.57      0.56      2232\n",
      "weighted avg       0.74      0.75      0.68      2232\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Calculate F1 score\n",
    "positive_label = 'positive'\n",
    "f1score = f1_score(y_test, y_pred, pos_label=positive_label)\n",
    "\n",
    "# Calculate log loss using predicted probabilities\n",
    "logloss = log_loss(y_test, y_pred_proba)\n",
    "\n",
    "# Calculate mean squared error using encoded labels\n",
    "mse = mean_squared_error(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Calculate accuracy\n",
    "accuracy = accuracy_score(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Calculate balanced accuracy\n",
    "balanced_accuracy = balanced_accuracy_score(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Calculate Matthews Correlation Coefficient (MCC)\n",
    "mcc, mcc_details = matthews(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Generate the classification report\n",
    "classification_report_str = classification_report(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Print the evaluation metrics\n",
    "print('Accuracy:', accuracy)\n",
    "print('Balanced Accuracy:', balanced_accuracy)\n",
    "print('Log Loss:', logloss)  # Use the calculated log loss\n",
    "print('Mean Squared Error:', mse)\n",
    "print()\n",
    "print('F1 Score:', f1score)\n",
    "print('Matthews Correlation Coefficient:', mcc)\n",
    "print('Matthews Correlation Coefficient Details:', mcc_details)\n",
    "print()\n",
    "print('Classification Report:\\n', classification_report_str)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Random Forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-2 {color: black;}#sk-container-id-2 pre{padding: 0;}#sk-container-id-2 div.sk-toggleable {background-color: white;}#sk-container-id-2 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-2 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-2 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-2 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-2 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-2 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-2 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-2 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-2 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-2 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-2 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-2 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-2 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-2 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-2 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-2 div.sk-item {position: relative;z-index: 1;}#sk-container-id-2 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-2 div.sk-item::before, #sk-container-id-2 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-2 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-2 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-2 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-2 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-2 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-2 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-2 div.sk-label-container {text-align: center;}#sk-container-id-2 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-2 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-2\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-2\" type=\"checkbox\" checked><label for=\"sk-estimator-id-2\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">RandomForestClassifier</label><div class=\"sk-toggleable__content\"><pre>RandomForestClassifier()</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "RandomForestClassifier()"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Train a logistic regression model\n",
    "logreg = RandomForestClassifier()\n",
    "logreg.fit(X_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predict probabilities on the test set\n",
    "y_pred_proba = logreg.predict_proba(X_test)[:, 1]  # Probabilities for the positive class\n",
    "\n",
    "# Predict class labels for F1 score calculation\n",
    "y_pred = logreg.predict(X_test)\n",
    "\n",
    "# Encode labels for mean squared error calculation\n",
    "label_encoder = LabelEncoder()\n",
    "y_test_encoded = label_encoder.fit_transform(y_test)\n",
    "y_pred_encoded = label_encoder.transform(y_pred)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.950694755714926\n",
      "Balanced Accuracy: 0.9219307992222207\n",
      "Log Loss: 0.15221563976229685\n",
      "Mean Squared Error: 0.049305244285073956\n",
      "\n",
      "F1 Score: 0.9064625850340136\n",
      "Matthews Correlation Coefficient: 0.8758364686130411\n",
      "Matthews Correlation Coefficient Details:  \n",
      "     P: 622 \n",
      "     Tp: 533 \n",
      "     Fp: 21 \n",
      "     N: 1_609 \n",
      "     Tn: 1_588 \n",
      "     Fn: 89\n",
      "\n",
      "Classification Report:\n",
      "               precision    recall  f1-score   support\n",
      "\n",
      "           0       0.95      0.99      0.97      1609\n",
      "           1       0.96      0.86      0.91       622\n",
      "\n",
      "    accuracy                           0.95      2231\n",
      "   macro avg       0.95      0.92      0.94      2231\n",
      "weighted avg       0.95      0.95      0.95      2231\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Calculate F1 score\n",
    "positive_label = 'positive'\n",
    "f1score = f1_score(y_test, y_pred, pos_label=positive_label)\n",
    "\n",
    "# Calculate log loss using predicted probabilities\n",
    "logloss = log_loss(y_test, y_pred_proba)\n",
    "\n",
    "# Calculate mean squared error using encoded labels\n",
    "mse = mean_squared_error(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Calculate accuracy\n",
    "accuracy = accuracy_score(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Calculate balanced accuracy\n",
    "balanced_accuracy = balanced_accuracy_score(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Calculate Matthews Correlation Coefficient (MCC)\n",
    "mcc, mcc_details = matthews(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Generate the classification report\n",
    "classification_report_str = classification_report(y_test_encoded, y_pred_encoded)\n",
    "\n",
    "# Print the evaluation metrics\n",
    "print('Accuracy:', accuracy)\n",
    "print('Balanced Accuracy:', balanced_accuracy)\n",
    "print('Log Loss:', logloss)  # Use the calculated log loss\n",
    "print('Mean Squared Error:', mse)\n",
    "print()\n",
    "print('F1 Score:', f1score)\n",
    "print('Matthews Correlation Coefficient:', mcc)\n",
    "print('Matthews Correlation Coefficient Details:', mcc_details)\n",
    "print()\n",
    "print('Classification Report:\\n', classification_report_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Random forest with weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy: 0.9551971326164874\n",
      "Balanced Accuracy: 0.9296991902338492\n",
      "Log Loss: 0.1677458489520838\n",
      "Mean Squared Error: 0.044802867383512544\n",
      "\n",
      "F1 Score: 0.9169435215946844\n",
      "Matthews Correlation Coefficient: 0.8864261377564321\n",
      "Matthews Correlation Coefficient Details:  \n",
      "     P: 634 \n",
      "     Tp: 552 \n",
      "     Fp: 20 \n",
      "     N: 1_598 \n",
      "     Tn: 1_578 \n",
      "     Fn: 82\n",
      "\n",
      "Classification Report:\n",
      "               precision    recall  f1-score   support\n",
      "\n",
      "           0       0.95      0.99      0.97      1598\n",
      "           1       0.97      0.87      0.92       634\n",
      "\n",
      "    accuracy                           0.96      2232\n",
      "   macro avg       0.96      0.93      0.94      2232\n",
      "weighted avg       0.96      0.96      0.95      2232\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Features (independent variables)\n",
    "features = ['Genome Length (bp)', 'Jumbophage', 'molGC (%)', 'Number CDS',\n",
    "            'Positive Strand (%)', 'Negative Strand (%)', 'Coding Capacity (%)',\n",
    "            'tRNAs', 'Molecule_DNA', 'Molecule_RNA', 'Molecule_ss-DNA', 'Molecule_ss-RNA']\n",
    "\n",
    "# Target variable (dependent variable)\n",
    "target = 'staining'\n",
    "X = df[features]\n",
    "y = df[target]\n",
    "\n",
    "# Ensure y only contains valid labels ('negative' and 'positive')\n",
    "y = y[(y == 'negative') | (y == 'positive')]\n",
    "\n",
    "# Encode the target variable 'staining' to numerical labels\n",
    "label_encoder = LabelEncoder()\n",
    "y_encoded = label_encoder.fit_transform(y)\n",
    "\n",
    "# Split the data into training and testing sets\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.2, random_state=42)\n",
    "\n",
    "# Calculate class weights\n",
    "class_weights = compute_class_weight('balanced', classes=np.unique(y_train), y=y_train)\n",
    "\n",
    "# Train a random forest model with class weighting\n",
    "logreg = RandomForestClassifier(class_weight={0: class_weights[0], 1: class_weights[1]})\n",
    "logreg.fit(X_train, y_train)\n",
    "\n",
    "# Predict probabilities on the test set\n",
    "y_pred_proba = logreg.predict_proba(X_test)[:, 1]\n",
    "\n",
    "# Predict class labels for F1 score calculation\n",
    "y_pred = logreg.predict(X_test)\n",
    "\n",
    "# Calculate evaluation metrics\n",
    "f1score = f1_score(y_test, y_pred, pos_label=1)  # Positive class is 1 after encoding\n",
    "logloss = log_loss(y_test, y_pred_proba)\n",
    "mse = mean_squared_error(y_test, y_pred)\n",
    "accuracy = accuracy_score(y_test, y_pred)\n",
    "balanced_accuracy = balanced_accuracy_score(y_test, y_pred)\n",
    "\n",
    "# Generate the classification report\n",
    "classification_report_str = classification_report(y_test, y_pred)\n",
    "\n",
    "# Print the evaluation metrics\n",
    "print('Accuracy:', accuracy)\n",
    "print('Balanced Accuracy:', balanced_accuracy)\n",
    "print('Log Loss:', logloss)  # Use the calculated log loss\n",
    "print('Mean Squared Error:', mse)\n",
    "print()\n",
    "print('F1 Score:', f1score)\n",
    "print('Matthews Correlation Coefficient:', mcc)\n",
    "print('Matthews Correlation Coefficient Details:', mcc_details)\n",
    "print()\n",
    "print('Classification Report:\\n', classification_report_str)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.feature_selection import SelectFromModel\n",
    "\n",
    "model = RandomForestClassifier()\n",
    "sfm = SelectFromModel(model, threshold='mean')\n",
    "fit = sfm.fit(X_train, y_train)\n",
    "selected_features_sfm = X_train.columns[fit.get_support()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['molGC (%)', 'Positive Strand (%)', 'Negative Strand (%)',\n",
       "       'Coding Capacity (%)', 'tRNAs'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "selected_features_sfm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "phage",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
